<HTML>
<HEAD>
<TITLE>
	CUBIC publication list sorted by subject
</TITLE>

<LINK rel="STYLESHEET" href="../style/default.css" type="text/css">

<LINK rel="Index" HREF="index.html">
<LINK rev="Prev"  HREF="list.html">
<LINK rel="Next"  HREF="list_year.html">
</HEAD>



<BODY>


<!-- ================================================================================ -->
<!-- beg: top navigate buttons  -->
<!--   -->
<A NAME="top">

<A HREF="#bottom">			Bottom</A> - 
<A HREF="list.html">			Previous</A> - 
<A HREF="list_year.html">			Next</A> - 
<A HREF="http://cubic.bioc.columbia.edu/">			CUBIC</A> -
<A HREF="http://dodo.bioc.columbia.edu/predictprotein/predictprotein.html">	PP home</A>
  <A HREF="index.html">CUBIC papers</A> -  <A HREF="list_year.html">Papers by year</A> -  <A HREF="list_ascii.txt">Papers as flat file</A>

</P></A>
<!-- end: top navigate buttons  -->
<!-- ================================================================================ -->



<H1>CUBIC publication list sorted by subject</H1>
<P>
<!-- ================================================================================ -->
<!-- beg: summary  -->
<P>
<!-- ================================================================================ -->
<!-- beg: summary  -->
<UL>
<LI><A HREF="#structure">Protein structure</A></LI>
<LI><A HREF="#function">Protein function</A></LI>
<LI><A HREF="#localization">Sub-cellular localization</A></LI>
<LI><A HREF="#interaction">Protein-protein interaction</A></LI>
<LI><A HREF="#secondary">Secondary structure</A></LI>
<LI><A HREF="#accessibility">Solvent accessibility</A></LI>
<LI><A HREF="#transmembrane">Transmembrane segments</A></LI>
<LI><A HREF="#threading">Threading (remote homology)</A></LI>
<LI><A HREF="#evolution">Evolution</A></LI>
<LI><A HREF="#evaluation">Evaluation of prediction methods</A></LI>
<LI><A HREF="#cluster">Clustering proteins</A></LI>
<LI><A HREF="#alignment">Alignment methods</A></LI>
<LI><A HREF="#structural gen">Structural genomics</A></LI>
<LI><A HREF="#proteomics">Protemics predictions</A></LI>
<LI><A HREF="#service">Prediction services</A></LI>
<LI><A HREF="#sequence ana">Sequence analysis</A></LI>
<LI><A HREF="#review">Reviews</A></LI>
<LI><A HREF="#database">Databases</A></LI>
</UL>
<!-- end: summary -->
<!-- ================================================================================ -->
<P>
<P>
<!-- ================================================================================ -->
<!-- beg: table -->
<TABLE BORDER=1 COLS=3 CELLPADDING=3 CELLSPACING=2 WIDTH="100%">
<TR><TH WIDTH=20><STRONG>Link</STRONG></TH>	<TH WIDTH=50><STRONG>Title</STRONG></TH>	<TH WIDTH=90><STRONG>Quote</STRONG></TH>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="structure">Protein structure</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_x_receptors/">www</A></TD>	<TD WIDTH=50>Transmembrane domains in the functions of Fc receptors</TD>	<TD WIDTH=90>R Zidovetzki, B Rost, DL Armstrong and I Pecht (2003) J Biophys Chem, 15, 555-575</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_norsp/">www</A></TD>	<TD WIDTH=50>NORSp: predictions of long regions without regular secondary structure</TD>	<TD WIDTH=90>J Liu and B Rost (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_loc3d/">www</A></TD>	<TD WIDTH=50>LOC3D: annotate sub-cellular localization for protein structures</TD>	<TD WIDTH=90>R Nair and B Rost (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>DSSPcont: continuous secondary structure assignments for proteins</TD>	<TD WIDTH=90>P Carter, CAF Andersen and B Rost (2003) Nucl Acids Res,  submitted</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>EVA: evaluation of protein structure prediction servers</TD>	<TD WIDTH=90>IYY Koh, VA Eyrich, MA Mart’-Renom, D Przybylski, MS Madhusudhan, A Fiser, F Pazos, A Valencia, A Sali and B Rost (2003) Nucl Acids Res,  submitted</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Better prediction of sub-cellular localization through evolution and structure</TD>	<TD WIDTH=90>R Nair and B Rost (2003) Proteins,  submitted</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>CAFASP3 in the spotlight of EVA</TD>	<TD WIDTH=90>VA Eyrich, IYY Koh, D Przybylski, O Grana, F Pazos, A Valencia and B Rost (2003) Proteins,  submitted</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_pp/">www</A></TD>	<TD WIDTH=50>The PredictProtein server</TD>	<TD WIDTH=90>B Rost and J Liu (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Improving the prediction of protein secondary structure in three and eight classes using recurrent neural networks and profiles</TD>	<TD WIDTH=90>G Pollastri, D Przybylski, B Rost and P Baldi (2002) Proteins, 47, 228-235</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_x_cm">www</A></TD>	<TD WIDTH=50>Reliability of assessment of protein structure prediction methods</TD>	<TD WIDTH=90>MA Marti-Renom, MS Madhusudhan, A Fiser, B Rost and A Sali (2002) Structure, 10, 435-440</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_psiblast/">www</A></TD>	<TD WIDTH=50>Alignments grow, secondary structure prediction improves</TD>	<TD WIDTH=90>D Przybylski and B Rost (2002) Proteins, 46, 197-205</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_dsspcont">www</A></TD>	<TD WIDTH=50>Continuum secondary structure captures protein flexibility</TD>	<TD WIDTH=90>CAF Andersen, AG Palmer, S Brunak and B Rost (2002) Structure, 10, 175-185</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Loopy proteins appear conserved in evolution</TD>	<TD WIDTH=90>J Liu, H Tan and B Rost (2002) J Mol Biol, 322, 53-64</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2001_x_cafasp2">www</A></TD>	<TD WIDTH=50>Cafasp2: The second critical assessment of fully automated structure prediction methods</TD>	<TD WIDTH=90>D Fischer, A Elofsson, L Rychlewski, F Pazos, A Valencia, B Rost, AR Ortiz and RLJ Dunbrack (2001) Proteins, 45 Suppl 5, S171-183</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2001_eva_bioinf/">www</A></TD>	<TD WIDTH=50>EVA: continuous automatic evaluation of protein structure prediction servers</TD>	<TD WIDTH=90>VA Eyrich, MA Marti-Renom, MS Madhusudhan, A Fiser, F Pazos, A Valencia, A Sali, and B. Rost (2001) Bioinformatics, 17, 1242-1243</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>A modified definition of SOV, a segment-based measure for protein secondary structure prediction assessment</TD>	<TD WIDTH=90>A Zemla, C Venclovas, K Fidelis and B Rost (1999) Proteins, 34, 220-223</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Adaptation of protein surfaces to subcellular location</TD>	<TD WIDTH=90>MA Andrade, SI O'Donoghue and B Rost (1998) Journal of Molecular Biology, 276, 517-525</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Midnight zone of protein structure evolution</TD>	<TD WIDTH=90>B Rost, S O'Donoghue and C Sander (1998) </TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Protein fold recognition by prediction-based threading</TD>	<TD WIDTH=90>B Rost, R Schneider and C Sander (1997) Journal of Molecular Biology, 270, 471-480</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Protein structures sustain evolutionary drift</TD>	<TD WIDTH=90>B Rost (1997) Folding & Design, 2, S19-S24</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Better 1D predictions by experts with machines</TD>	<TD WIDTH=90>B Rost (1997) Proteins, Suppl. 1, 192-197</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>PHD: predicting one-dimensional protein structure by profile based neural networks</TD>	<TD WIDTH=90>B Rost (1996) Methods in Enzymology, 266, 525-539</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Refining neural network predictions for helical transmembrane proteins by dynamic programming</TD>	<TD WIDTH=90>B Rost, R Casadio and P Fariselli (1996) Fourth International Conference on Intelligent Systems for Molecular Biology,  192-200</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Topology prediction for helical transmembrane proteins at 86% accuracy</TD>	<TD WIDTH=90>B Rost, R Casadio and P Fariselli (1996) Protein Science, 5, 1704-1718</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Progress of 1D protein structure prediction at last</TD>	<TD WIDTH=90>B Rost and C Sander (1995) Proteins, 23, 295-300</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Prediction of helical transmembrane segments at 95% accuracy</TD>	<TD WIDTH=90>B Rost, R Casadio, P Fariselli and C Sander (1995) Protein Science, 4, 521-533</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/1995_dk/">www</A></TD>	<TD WIDTH=50>Fitting 1-D predictions into 3-D structures</TD>	<TD WIDTH=90>B Rost (1995) in: 'Protein folds: a distance based approach' (eds. H Bohr and S Brunak),  132-151</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>TOPITS: Threading One-dimensional Predictions Into Three-dimensional Structures</TD>	<TD WIDTH=90>B Rost (1995) Third International Conference on Intelligent Systems for Molecular Biology,  314-321</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>PHD - an automatic server for protein secondary structure prediction</TD>	<TD WIDTH=90>B Rost, C Sander and R Schneider (1994) CABIOS, 10, 53-60</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Redefining the goals of protein secondary structure prediction</TD>	<TD WIDTH=90>B Rost, C Sander and R Schneider (1994) Journal of Molecular Biology, 235, 13-26</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>1D secondary structure prediction through evolutionary profiles</TD>	<TD WIDTH=90>B Rost and C Sander (1994) in: 'Protein Structure by Distance Analysis' (eds. H Bohr and S Brunak),  257-276</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Combining evolutionary information and neural networks to predict protein secondary structure</TD>	<TD WIDTH=90>B Rost and C Sander (1994) Proteins, 19, 55-72</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Conservation and prediction of solvent accessibility in protein families</TD>	<TD WIDTH=90>B Rost and C Sander (1994) Proteins, 20, 216-226</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Improved prediction of protein secondary structure by use of sequence profiles and neural networks</TD>	<TD WIDTH=90>B Rost and C Sander (1993) Proceedings of the National Academy of Sciences, 90, 7558-7562</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Prediction of protein secondary structure at better than 70% accuracy</TD>	<TD WIDTH=90>B Rost and C Sander (1993) Journal of Molecular Biology, 232, 584-599</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Secondary structure prediction of all-helical proteins in two states</TD>	<TD WIDTH=90>B Rost and C Sander (1993) Protein Engineering, 6, 831-836</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/1992_nature/">www</A></TD>	<TD WIDTH=50>Jury returns on structure prediction</TD>	<TD WIDTH=90>B Rost and C Sander (1992) Nature, 360, 540</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Exercising multi-layered networks on protein secondary structure</TD>	<TD WIDTH=90>B Rost and C Sander (1992) Neural Networks: From Biology to High Energy Physics, 3, 209-220</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="function">Protein function</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_x_receptors/">www</A></TD>	<TD WIDTH=50>Transmembrane domains in the functions of Fc receptors</TD>	<TD WIDTH=90>R Zidovetzki, B Rost, DL Armstrong and I Pecht (2003) J Biophys Chem, 15, 555-575</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_inter_an/">www</A></TD>	<TD WIDTH=50>Analysing six types of protein-protein interfaces</TD>	<TD WIDTH=90>Y Ofran and B Rost (2003) J Mol Biol, 325, 377-387</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_norsp/">www</A></TD>	<TD WIDTH=50>NORSp: predictions of long regions without regular secondary structure</TD>	<TD WIDTH=90>J Liu and B Rost (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Local sequence information can predict protein-protein interaction sites</TD>	<TD WIDTH=90>Y Ofran and B Rost (2003) FEBS Let,  submitted</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_dsspcont">www</A></TD>	<TD WIDTH=50>Continuum secondary structure captures protein flexibility</TD>	<TD WIDTH=90>CAF Andersen, AG Palmer, S Brunak and B Rost (2002) Structure, 10, 175-185</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_loci_text/">www</A></TD>	<TD WIDTH=50>Inferring sub-cellular localization through automated lexical analysis</TD>	<TD WIDTH=90>R Nair and B Rost (2002) Bioinformatics, 18:Suppl 1, S78-S86</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Loopy proteins appear conserved in evolution</TD>	<TD WIDTH=90>J Liu, H Tan and B Rost (2002) J Mol Biol, 322, 53-64</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_enzyme_cons">www</A></TD>	<TD WIDTH=50>Enzyme function less conserved than anticipated</TD>	<TD WIDTH=90>B Rost (2002) J Mol Biol, 318, 595-608</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2001_genomes/">www</A></TD>	<TD WIDTH=50>Comparing function and structure between entire proteomes</TD>	<TD WIDTH=90>J Liu and B Rost (2001) Protein Science, 10, 1970-1979</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2000_nls/">www</A></TD>	<TD WIDTH=50>Finding nuclear localization signals</TD>	<TD WIDTH=90>M Cokol, R Nair and B Rost (2000) EMBO Reports, 1, 411-415</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Adaptation of protein surfaces to subcellular location</TD>	<TD WIDTH=90>MA Andrade, SI O'Donoghue and B Rost (1998) Journal of Molecular Biology, 276, 517-525</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Prediction of helical transmembrane segments at 95% accuracy</TD>	<TD WIDTH=90>B Rost, R Casadio, P Fariselli and C Sander (1995) Protein Science, 4, 521-533</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Molecular modelling of the Norrie disease protein predicts a cysteine knot growth factor tertiary structure</TD>	<TD WIDTH=90>T Meitinger, A Meindl, P Bork, B Rost, C Sander, M Haasemann and J Murken (1993) Nature Genetics, 5, 376-380</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="localization">Sub-cellular localization</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_nar_nls/">www</A></TD>	<TD WIDTH=50>NLSdb: database of nuclear localization signals</TD>	<TD WIDTH=90>R Nair, P Carter and B Rost (2003) Nucl Acid Res, 322, 397-399</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_loc3d/">www</A></TD>	<TD WIDTH=50>LOC3D: annotate sub-cellular localization for protein structures</TD>	<TD WIDTH=90>R Nair and B Rost (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Better prediction of sub-cellular localization through evolution and structure</TD>	<TD WIDTH=90>R Nair and B Rost (2003) Proteins,  submitted</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_loci_text/">www</A></TD>	<TD WIDTH=50>Inferring sub-cellular localization through automated lexical analysis</TD>	<TD WIDTH=90>R Nair and B Rost (2002) Bioinformatics, 18:Suppl 1, S78-S86</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_loci_curve/">www</A></TD>	<TD WIDTH=50>Sequence conserved for sub-cellular localization</TD>	<TD WIDTH=90>R Nair and B Rost (2002) Protein Science, 11, 2836-2847</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="interaction">Protein-protein interaction</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_inter_an/">www</A></TD>	<TD WIDTH=50>Analysing six types of protein-protein interfaces</TD>	<TD WIDTH=90>Y Ofran and B Rost (2003) J Mol Biol, 325, 377-387</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Local sequence information can predict protein-protein interaction sites</TD>	<TD WIDTH=90>Y Ofran and B Rost (2003) FEBS Let,  submitted</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="secondary">Secondary structure</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_norsp/">www</A></TD>	<TD WIDTH=50>NORSp: predictions of long regions without regular secondary structure</TD>	<TD WIDTH=90>J Liu and B Rost (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>DSSPcont: continuous secondary structure assignments for proteins</TD>	<TD WIDTH=90>P Carter, CAF Andersen and B Rost (2003) Nucl Acids Res,  submitted</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_pp/">www</A></TD>	<TD WIDTH=50>The PredictProtein server</TD>	<TD WIDTH=90>B Rost and J Liu (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Improving the prediction of protein secondary structure in three and eight classes using recurrent neural networks and profiles</TD>	<TD WIDTH=90>G Pollastri, D Przybylski, B Rost and P Baldi (2002) Proteins, 47, 228-235</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_psiblast/">www</A></TD>	<TD WIDTH=50>Alignments grow, secondary structure prediction improves</TD>	<TD WIDTH=90>D Przybylski and B Rost (2002) Proteins, 46, 197-205</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_dsspcont">www</A></TD>	<TD WIDTH=50>Continuum secondary structure captures protein flexibility</TD>	<TD WIDTH=90>CAF Andersen, AG Palmer, S Brunak and B Rost (2002) Structure, 10, 175-185</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2001_x_cafasp2">www</A></TD>	<TD WIDTH=50>Cafasp2: The second critical assessment of fully automated structure prediction methods</TD>	<TD WIDTH=90>D Fischer, A Elofsson, L Rychlewski, F Pazos, A Valencia, B Rost, AR Ortiz and RLJ Dunbrack (2001) Proteins, 45 Suppl 5, S171-183</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>A modified definition of SOV, a segment-based measure for protein secondary structure prediction assessment</TD>	<TD WIDTH=90>A Zemla, C Venclovas, K Fidelis and B Rost (1999) Proteins, 34, 220-223</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Better 1D predictions by experts with machines</TD>	<TD WIDTH=90>B Rost (1997) Proteins, Suppl. 1, 192-197</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>PHD: predicting one-dimensional protein structure by profile based neural networks</TD>	<TD WIDTH=90>B Rost (1996) Methods in Enzymology, 266, 525-539</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Topology prediction for helical transmembrane proteins at 86% accuracy</TD>	<TD WIDTH=90>B Rost, R Casadio and P Fariselli (1996) Protein Science, 5, 1704-1718</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Progress of 1D protein structure prediction at last</TD>	<TD WIDTH=90>B Rost and C Sander (1995) Proteins, 23, 295-300</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>PHD - an automatic server for protein secondary structure prediction</TD>	<TD WIDTH=90>B Rost, C Sander and R Schneider (1994) CABIOS, 10, 53-60</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Redefining the goals of protein secondary structure prediction</TD>	<TD WIDTH=90>B Rost, C Sander and R Schneider (1994) Journal of Molecular Biology, 235, 13-26</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>1D secondary structure prediction through evolutionary profiles</TD>	<TD WIDTH=90>B Rost and C Sander (1994) in: 'Protein Structure by Distance Analysis' (eds. H Bohr and S Brunak),  257-276</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Combining evolutionary information and neural networks to predict protein secondary structure</TD>	<TD WIDTH=90>B Rost and C Sander (1994) Proteins, 19, 55-72</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Improved prediction of protein secondary structure by use of sequence profiles and neural networks</TD>	<TD WIDTH=90>B Rost and C Sander (1993) Proceedings of the National Academy of Sciences, 90, 7558-7562</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Prediction of protein secondary structure at better than 70% accuracy</TD>	<TD WIDTH=90>B Rost and C Sander (1993) Journal of Molecular Biology, 232, 584-599</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Secondary structure prediction of all-helical proteins in two states</TD>	<TD WIDTH=90>B Rost and C Sander (1993) Protein Engineering, 6, 831-836</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/1992_nature/">www</A></TD>	<TD WIDTH=50>Jury returns on structure prediction</TD>	<TD WIDTH=90>B Rost and C Sander (1992) Nature, 360, 540</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Exercising multi-layered networks on protein secondary structure</TD>	<TD WIDTH=90>B Rost and C Sander (1992) Neural Networks: From Biology to High Energy Physics, 3, 209-220</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="accessibility">Solvent accessibility</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_pp/">www</A></TD>	<TD WIDTH=50>The PredictProtein server</TD>	<TD WIDTH=90>B Rost and J Liu (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_psiblast/">www</A></TD>	<TD WIDTH=50>Alignments grow, secondary structure prediction improves</TD>	<TD WIDTH=90>D Przybylski and B Rost (2002) Proteins, 46, 197-205</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Adaptation of protein surfaces to subcellular location</TD>	<TD WIDTH=90>MA Andrade, SI O'Donoghue and B Rost (1998) Journal of Molecular Biology, 276, 517-525</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Better 1D predictions by experts with machines</TD>	<TD WIDTH=90>B Rost (1997) Proteins, Suppl. 1, 192-197</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>PHD: predicting one-dimensional protein structure by profile based neural networks</TD>	<TD WIDTH=90>B Rost (1996) Methods in Enzymology, 266, 525-539</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Progress of 1D protein structure prediction at last</TD>	<TD WIDTH=90>B Rost and C Sander (1995) Proteins, 23, 295-300</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Conservation and prediction of solvent accessibility in protein families</TD>	<TD WIDTH=90>B Rost and C Sander (1994) Proteins, 20, 216-226</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="transmembrane">Transmembrane segments</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_x_receptors/">www</A></TD>	<TD WIDTH=50>Transmembrane domains in the functions of Fc receptors</TD>	<TD WIDTH=90>R Zidovetzki, B Rost, DL Armstrong and I Pecht (2003) J Biophys Chem, 15, 555-575</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_htm/">www</A></TD>	<TD WIDTH=50>Static benchmarking of membrane helix predictions</TD>	<TD WIDTH=90>A Kernytsky and B Rost (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_pp/">www</A></TD>	<TD WIDTH=50>The PredictProtein server</TD>	<TD WIDTH=90>B Rost and J Liu (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2001_genomes/">www</A></TD>	<TD WIDTH=50>Comparing function and structure between entire proteomes</TD>	<TD WIDTH=90>J Liu and B Rost (2001) Protein Science, 10, 1970-1979</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>The role of transmembrane domains in the functions of B- and T-cell receptors</TD>	<TD WIDTH=90>R Zidovetzki, B Rost and I Pecht (1998) Immunology Letters, 64, 97-107</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>PHD: predicting one-dimensional protein structure by profile based neural networks</TD>	<TD WIDTH=90>B Rost (1996) Methods in Enzymology, 266, 525-539</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Refining neural network predictions for helical transmembrane proteins by dynamic programming</TD>	<TD WIDTH=90>B Rost, R Casadio and P Fariselli (1996) Fourth International Conference on Intelligent Systems for Molecular Biology,  192-200</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Topology prediction for helical transmembrane proteins at 86% accuracy</TD>	<TD WIDTH=90>B Rost, R Casadio and P Fariselli (1996) Protein Science, 5, 1704-1718</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Prediction of helical transmembrane segments at 95% accuracy</TD>	<TD WIDTH=90>B Rost, R Casadio, P Fariselli and C Sander (1995) Protein Science, 4, 521-533</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="threading">Threading (remote homology)</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>EVA: evaluation of protein structure prediction servers</TD>	<TD WIDTH=90>IYY Koh, VA Eyrich, MA Mart’-Renom, D Przybylski, MS Madhusudhan, A Fiser, F Pazos, A Valencia, A Sali and B Rost (2003) Nucl Acids Res,  submitted</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>CAFASP3 in the spotlight of EVA</TD>	<TD WIDTH=90>VA Eyrich, IYY Koh, D Przybylski, O Grana, F Pazos, A Valencia and B Rost (2003) Proteins,  submitted</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2001_eva_bioinf/">www</A></TD>	<TD WIDTH=50>EVA: continuous automatic evaluation of protein structure prediction servers</TD>	<TD WIDTH=90>VA Eyrich, MA Marti-Renom, MS Madhusudhan, A Fiser, F Pazos, A Valencia, A Sali, and B. Rost (2001) Bioinformatics, 17, 1242-1243</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http:/cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>A platform for integrating threading results with protein family analyses</TD>	<TD WIDTH=90>F Pazos, B Rost and A Valencia (1999) Bioinformatics, 15, 1062-1063</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Effective use of sequence correlation and conservation in fold recognition</TD>	<TD WIDTH=90>O Olmea, B Rost and A Valencia (1999) Journal of Molecular Biology, 293, 1221-1239</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Protein fold recognition by prediction-based threading</TD>	<TD WIDTH=90>B Rost, R Schneider and C Sander (1997) Journal of Molecular Biology, 270, 471-480</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Update on protein structure prediction: results of the 1995 IRBM workshop</TD>	<TD WIDTH=90>T Hubbard, A Tramontano, G Barton, D Jones, M Sippl, A Valencia, A Lesk, J Moult, B Rost, C Sander, R Schneider, A Lahm, R Leplae, C Buta , M Eisenstein, O Fjellström, H Floeckner, JG Grossmann, J Hansen, M Helmer-Citterich, FS Joergensen, A Marchler-Bauer, J Osuna, J Park, A Reinhardt, L Ribas de Pouplana, A Rojo-Dominguez, V Saudek, J Sinclair, S Sturrock, C Venclovas and C Vinals (1996) Folding & Design, 1, R55-R63</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Protein fold recognition by merging 1D structure prediction and sequence alignments</TD>	<TD WIDTH=90>B Rost (1996) </TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/1995_dk/">www</A></TD>	<TD WIDTH=50>Fitting 1-D predictions into 3-D structures</TD>	<TD WIDTH=90>B Rost (1995) in: 'Protein folds: a distance based approach' (eds. H Bohr and S Brunak),  132-151</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>TOPITS: Threading One-dimensional Predictions Into Three-dimensional Structures</TD>	<TD WIDTH=90>B Rost (1995) Third International Conference on Intelligent Systems for Molecular Biology,  314-321</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="evolution">Evolution</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_psiblast/">www</A></TD>	<TD WIDTH=50>Alignments grow, secondary structure prediction improves</TD>	<TD WIDTH=90>D Przybylski and B Rost (2002) Proteins, 46, 197-205</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_loci_curve/">www</A></TD>	<TD WIDTH=50>Sequence conserved for sub-cellular localization</TD>	<TD WIDTH=90>R Nair and B Rost (2002) Protein Science, 11, 2836-2847</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_enzyme_cons">www</A></TD>	<TD WIDTH=50>Enzyme function less conserved than anticipated</TD>	<TD WIDTH=90>B Rost (2002) J Mol Biol, 318, 595-608</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Midnight zone of protein structure evolution</TD>	<TD WIDTH=90>B Rost, S O'Donoghue and C Sander (1998) </TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Protein structures sustain evolutionary drift</TD>	<TD WIDTH=90>B Rost (1997) Folding & Design, 2, S19-S24</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Combining evolutionary information and neural networks to predict protein secondary structure</TD>	<TD WIDTH=90>B Rost and C Sander (1994) Proteins, 19, 55-72</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Prediction of protein secondary structure at better than 70% accuracy</TD>	<TD WIDTH=90>B Rost and C Sander (1993) Journal of Molecular Biology, 232, 584-599</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="evaluation">Evaluation of prediction methods</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_htm/">www</A></TD>	<TD WIDTH=50>Static benchmarking of membrane helix predictions</TD>	<TD WIDTH=90>A Kernytsky and B Rost (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>EVA: evaluation of protein structure prediction servers</TD>	<TD WIDTH=90>IYY Koh, VA Eyrich, MA Mart’-Renom, D Przybylski, MS Madhusudhan, A Fiser, F Pazos, A Valencia, A Sali and B Rost (2003) Nucl Acids Res,  submitted</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>CAFASP3 in the spotlight of EVA</TD>	<TD WIDTH=90>VA Eyrich, IYY Koh, D Przybylski, O Grana, F Pazos, A Valencia and B Rost (2003) Proteins,  submitted</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_x_cm">www</A></TD>	<TD WIDTH=50>Reliability of assessment of protein structure prediction methods</TD>	<TD WIDTH=90>MA Marti-Renom, MS Madhusudhan, A Fiser, B Rost and A Sali (2002) Structure, 10, 435-440</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_htm_eval/">www</A></TD>	<TD WIDTH=50>Transmembrane helix predictions revisited</TD>	<TD WIDTH=90>CP Chen, A Kernytsky and B Rost (2002) Protein Science, 11, 2774-2791</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_htm_length/">www</A></TD>	<TD WIDTH=50>Long membrane helices and short loops predicted less accurately</TD>	<TD WIDTH=90>CP Chen and B Rost (2002) Protein Science, volxx, 2766-2773</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_enzyme_cons">www</A></TD>	<TD WIDTH=50>Enzyme function less conserved than anticipated</TD>	<TD WIDTH=90>B Rost (2002) J Mol Biol, 318, 595-608</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>A modified definition of SOV, a segment-based measure for protein secondary structure prediction assessment</TD>	<TD WIDTH=90>A Zemla, C Venclovas, K Fidelis and B Rost (1999) Proteins, 34, 220-223</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Protein fold recognition by prediction-based threading</TD>	<TD WIDTH=90>B Rost, R Schneider and C Sander (1997) Journal of Molecular Biology, 270, 471-480</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Progress of 1D protein structure prediction at last</TD>	<TD WIDTH=90>B Rost and C Sander (1995) Proteins, 23, 295-300</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>TOPITS: Threading One-dimensional Predictions Into Three-dimensional Structures</TD>	<TD WIDTH=90>B Rost (1995) Third International Conference on Intelligent Systems for Molecular Biology,  314-321</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Redefining the goals of protein secondary structure prediction</TD>	<TD WIDTH=90>B Rost, C Sander and R Schneider (1994) Journal of Molecular Biology, 235, 13-26</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="cluster">Clustering proteins</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>CHOP proteins into structural domain-like fragments</TD>	<TD WIDTH=90>J Liu and B Rost (2003) J Mol Biol, volxx, submitted</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_target/">www</A></TD>	<TD WIDTH=50>Target space for structural genomics revisited</TD>	<TD WIDTH=90>J Liu and B Rost (2002) Bioinformatics, 18, 922-933</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="alignment">Alignment methods</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_psiblast/">www</A></TD>	<TD WIDTH=50>Alignments grow, secondary structure prediction improves</TD>	<TD WIDTH=90>D Przybylski and B Rost (2002) Proteins, 46, 197-205</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Twilight zone of protein sequence alignments</TD>	<TD WIDTH=90>B Rost (1999) Protein Engineering, 12, 85-94</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>TOPITS: Threading One-dimensional Predictions Into Three-dimensional Structures</TD>	<TD WIDTH=90>B Rost (1995) Third International Conference on Intelligent Systems for Molecular Biology,  314-321</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="structural gen">Structural genomics</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_nar_pep/">www</A></TD>	<TD WIDTH=50>PEP: Predictions for Entire Proteomes</TD>	<TD WIDTH=90>P Carter, J Liu and B Rost (2003) Nucl Acid Res, 31, 410-413</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>CHOP proteins into structural domain-like fragments</TD>	<TD WIDTH=90>J Liu and B Rost (2003) J Mol Biol, volxx, submitted</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_target/">www</A></TD>	<TD WIDTH=50>Target space for structural genomics revisited</TD>	<TD WIDTH=90>J Liu and B Rost (2002) Bioinformatics, 18, 922-933</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Loopy proteins appear conserved in evolution</TD>	<TD WIDTH=90>J Liu, H Tan and B Rost (2002) J Mol Biol, 322, 53-64</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_enzyme_cons">www</A></TD>	<TD WIDTH=50>Enzyme function less conserved than anticipated</TD>	<TD WIDTH=90>B Rost (2002) J Mol Biol, 318, 595-608</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers">www</A></TD>	<TD WIDTH=50>Bioinformatics in structural genomics</TD>	<TD WIDTH=90>B Rost, B Honig and A Valencia (2002) Bioinformatics, 18, 897</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="proteomics">Protemics predictions</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_nar_nls/">www</A></TD>	<TD WIDTH=50>NLSdb: database of nuclear localization signals</TD>	<TD WIDTH=90>R Nair, P Carter and B Rost (2003) Nucl Acid Res, 322, 397-399</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_nar_pep/">www</A></TD>	<TD WIDTH=50>PEP: Predictions for Entire Proteomes</TD>	<TD WIDTH=90>P Carter, J Liu and B Rost (2003) Nucl Acid Res, 31, 410-413</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_target/">www</A></TD>	<TD WIDTH=50>Target space for structural genomics revisited</TD>	<TD WIDTH=90>J Liu and B Rost (2002) Bioinformatics, 18, 922-933</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_loci_text/">www</A></TD>	<TD WIDTH=50>Inferring sub-cellular localization through automated lexical analysis</TD>	<TD WIDTH=90>R Nair and B Rost (2002) Bioinformatics, 18:Suppl 1, S78-S86</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Loopy proteins appear conserved in evolution</TD>	<TD WIDTH=90>J Liu, H Tan and B Rost (2002) J Mol Biol, 322, 53-64</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_loci_curve/">www</A></TD>	<TD WIDTH=50>Sequence conserved for sub-cellular localization</TD>	<TD WIDTH=90>R Nair and B Rost (2002) Protein Science, 11, 2836-2847</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2001_genomes/">www</A></TD>	<TD WIDTH=50>Comparing function and structure between entire proteomes</TD>	<TD WIDTH=90>J Liu and B Rost (2001) Protein Science, 10, 1970-1979</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="service">Prediction services</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_norsp/">www</A></TD>	<TD WIDTH=50>NORSp: predictions of long regions without regular secondary structure</TD>	<TD WIDTH=90>J Liu and B Rost (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_loc3d/">www</A></TD>	<TD WIDTH=50>LOC3D: annotate sub-cellular localization for protein structures</TD>	<TD WIDTH=90>R Nair and B Rost (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_htm/">www</A></TD>	<TD WIDTH=50>Static benchmarking of membrane helix predictions</TD>	<TD WIDTH=90>A Kernytsky and B Rost (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>UniqueProt: creating sequence-unique protein data sets</TD>	<TD WIDTH=90>S Mika and B Rost (2003) Nucl Acids Res,  submitted</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>META-PP: single crucial to prediction servers</TD>	<TD WIDTH=90>VA Eyrich and B Rost (2003) Nucl Acids Res,  submitted</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>DSSPcont: continuous secondary structure assignments for proteins</TD>	<TD WIDTH=90>P Carter, CAF Andersen and B Rost (2003) Nucl Acids Res,  submitted</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>EVA: evaluation of protein structure prediction servers</TD>	<TD WIDTH=90>IYY Koh, VA Eyrich, MA Mart’-Renom, D Przybylski, MS Madhusudhan, A Fiser, F Pazos, A Valencia, A Sali and B Rost (2003) Nucl Acids Res,  submitted</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>CAFASP3 in the spotlight of EVA</TD>	<TD WIDTH=90>VA Eyrich, IYY Koh, D Przybylski, O Grana, F Pazos, A Valencia and B Rost (2003) Proteins,  submitted</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_pp/">www</A></TD>	<TD WIDTH=50>The PredictProtein server</TD>	<TD WIDTH=90>B Rost and J Liu (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>PHD - an automatic server for protein secondary structure prediction</TD>	<TD WIDTH=90>B Rost, C Sander and R Schneider (1994) CABIOS, 10, 53-60</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="sequence ana">Sequence analysis</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_inter_an/">www</A></TD>	<TD WIDTH=50>Analysing six types of protein-protein interfaces</TD>	<TD WIDTH=90>Y Ofran and B Rost (2003) J Mol Biol, 325, 377-387</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>UniqueProt: creating sequence-unique protein data sets</TD>	<TD WIDTH=90>S Mika and B Rost (2003) Nucl Acids Res,  submitted</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>META-PP: single crucial to prediction servers</TD>	<TD WIDTH=90>VA Eyrich and B Rost (2003) Nucl Acids Res,  submitted</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_pp/">www</A></TD>	<TD WIDTH=50>The PredictProtein server</TD>	<TD WIDTH=90>B Rost and J Liu (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Loopy proteins appear conserved in evolution</TD>	<TD WIDTH=90>J Liu, H Tan and B Rost (2002) J Mol Biol, 322, 53-64</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_loci_curve/">www</A></TD>	<TD WIDTH=50>Sequence conserved for sub-cellular localization</TD>	<TD WIDTH=90>R Nair and B Rost (2002) Protein Science, 11, 2836-2847</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2001_genomes/">www</A></TD>	<TD WIDTH=50>Comparing function and structure between entire proteomes</TD>	<TD WIDTH=90>J Liu and B Rost (2001) Protein Science, 10, 1970-1979</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Twilight zone of protein sequence alignments</TD>	<TD WIDTH=90>B Rost (1999) Protein Engineering, 12, 85-94</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>The role of transmembrane domains in the functions of B- and T-cell receptors</TD>	<TD WIDTH=90>R Zidovetzki, B Rost and I Pecht (1998) Immunology Letters, 64, 97-107</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Molecular modelling of the Norrie disease protein predicts a cysteine knot growth factor tertiary structure</TD>	<TD WIDTH=90>T Meitinger, A Meindl, P Bork, B Rost, C Sander, M Haasemann and J Murken (1993) Nature Genetics, 5, 376-380</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="review">Reviews</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_rev_cluster/">www</A></TD>	<TD WIDTH=50>Domains, motifs, and clusters in the protein universe</TD>	<TD WIDTH=90>J Liu and B Rost (2003) Curr Opinion Chem Biol, 7, 5-11</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_rev_assign">www</A></TD>	<TD WIDTH=50>Secondary structure assignment</TD>	<TD WIDTH=90>CAF Andersen and B Rost (2003) in: 'Structural Bioinformatics' (eds. P Bourne and H Weissig), New York: Wiley, 339-361</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_cellcycle/">www</A></TD>	<TD WIDTH=50>Cataloguing proteins in cell cycle control</TD>	<TD WIDTH=90>KO Wrzeszczynski and B Rost (2003) in: 'Cell cycle checkpoint control protocols' (ed. H Liebermann), CITY: Humana Press, pages_xx</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Annotating protein function through lexical analysis</TD>	<TD WIDTH=90>R Nair and B Rost (2003) AAAI AI Magazine,  submitted</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_rev_1d">www</A></TD>	<TD WIDTH=50>Prediction In 1D: secondary structure, membrane helices, and accessibility</TD>	<TD WIDTH=90>B Rost (2003) in: 'Structural Bioinformatics' (eds. P Bourne and H Weissig), New York: Wiley, 557-585</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_rev_dekker/">www</A></TD>	<TD WIDTH=50>Rising accuracy of protein secondary structure prediction</TD>	<TD WIDTH=90>B Rost (2003) in: 'Protein structure determination, analysis, and modeling for drug discovery' (ed. D Chasman), New York: Dekker, 207-249</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_rev_nn">www</A></TD>	<TD WIDTH=50>Neural networks predict protein structure: hype or hit?</TD>	<TD WIDTH=90>B Rost (2003) in: 'Artificial intelligence and heuristic methods for bioinformatics' (ed. P Frasconi), CITY: ISO Press, pages_xx</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_rev_cheminfo/">www</A></TD>	<TD WIDTH=50>Predict protein structure through evolution</TD>	<TD WIDTH=90>B Rost, J Liu, D Przybylski, R Nair, KO Wrzeszczynski, H Bigelow and Y Ofran (2003) in: 'Chemoinformatics - From Data to Knowledge' (eds. J Gasteiger &amp; T Engel), CITY: Wiley, pages_xx</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Automatic prediction of protein function</TD>	<TD WIDTH=90>B Rost, J Liu, R Nair, KO Wrzeszczynski and Y Ofran (2003) CMSL, volxx, submitted 2003-03-25</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_rev_htm">www</A></TD>	<TD WIDTH=50>State-of-the-art in membrane protein prediction</TD>	<TD WIDTH=90>CP Chen and B Rost (2002) Applied Bioinformatics, 1, 21-35</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers">www</A></TD>	<TD WIDTH=50>Did evolution leap to create the protein universe?</TD>	<TD WIDTH=90>B Rost (2002) Curr Opin Struc Biol, 12, 409-416</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2002_rev_chem/">www</A></TD>	<TD WIDTH=50>Predict protein structure and function through evolutionary information</TD>	<TD WIDTH=90>B Rost, J Liu, D Przybylski, R Nair, KO Wrzeszczynski, H Bigelow and Y Ofran (2002) Preprint CUBIC_2003_02, New York: Columbia University, Dec 2002</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2001_rev_sec_jsb/">www</A></TD>	<TD WIDTH=50>Protein secondary structure prediction continues to rise</TD>	<TD WIDTH=90>B Rost (2001) J Structural Biology, 134, 204-218</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2001_casp4_eva">www</A></TD>	<TD WIDTH=50>EVA: large-scale analysis of secondary structure prediction</TD>	<TD WIDTH=90>B Rost and VE Eyrich (2001) Proteins, 45:S5, S192-199</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Third generation prediction of secondary structure</TD>	<TD WIDTH=90>B Rost and C Sander (2000) in: 'Protein structure prediction: Methods and Protocols' (ed. D. Webster), Totowa, NJ: Humana Press, 71-95</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/1999_rev_honnef/">www</A></TD>	<TD WIDTH=50>Evolution teaches neural networks</TD>	<TD WIDTH=90>B Rost (1999) in: 'Scientific applications of neural nets' (eds. JW Clark, T Lindenau and ML Ristig),  207-223</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Neural networks for protein structure prediction: hype or hit?</TD>	<TD WIDTH=90>B Rost (1999) </TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Pedestrian guide to analysing sequence databases</TD>	<TD WIDTH=90>B Rost and R Schneider (1999) in: 'Core techniques in biochemistry' (eds. K Ashman),  in press since 96..</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Protein structure prediction in 1D, 2D, and 3D</TD>	<TD WIDTH=90>B Rost (1998) in: 'The Encyclopaedia of Computational Chemistry' (eds. PvR Schleyer, NL Allinger, T Clark, J Gasteiger, PA Kollman, HF Schaefer III and PR Schreiner), 3, 2242-2255</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/98structure.html">www</A></TD>	<TD WIDTH=50>Marrying structure and genomics</TD>	<TD WIDTH=90>B Rost (1998) Structure, 6, 259-263</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>NN which predicts protein secondary structure</TD>	<TD WIDTH=90>B Rost (1997) in: 'Handbook of Neural Computation' (eds. E Fiesler and R Beale),  G4.1</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/ProcSweden97.html">www</A></TD>	<TD WIDTH=50>Learning from evolution to predict protein structure</TD>	<TD WIDTH=90>B Rost (1997) BCEC97: Bio-Computing and Emergent Computation,  87-101</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Sisyphus and prediction of protein structure</TD>	<TD WIDTH=90>B Rost and SI O'Donoghue (1997) Computer Applications in Biological Science, 13, 345-356</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Sisyphus and protein structure prediction</TD>	<TD WIDTH=90>B Rost (1997) 5th International Conference on Intelligent Systems for Molecular Biology,  Tutorial</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Bridging the protein sequence-structure gap by structure predictions</TD>	<TD WIDTH=90>B Rost and C Sander (1996) Annual Review of Biophysics and Biomolecular Structure, 25, 113-136</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Pitfalls of protein sequence analysis</TD>	<TD WIDTH=90>B Rost and A Valencia (1996) Current Opinion in Biotechnology, 7, 457-461</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Protein structure prediction by neural networks</TD>	<TD WIDTH=90>B Rost and C Sander (1995) in: 'The handbook of brain theory and neural networks' (eds. M Arbib),  772-775</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/1994_rev_cosb/">www</A></TD>	<TD WIDTH=50>Structure prediction of proteins - where are we now?</TD>	<TD WIDTH=90>B Rost and C Sander (1994) Current Opinion in Biotechnology, 5, 372-380</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/1993_rev_cda/">www</A></TD>	<TD WIDTH=50>Neural networks in chemistry</TD>	<TD WIDTH=90>B Rost and G Vriend (1993) CDA News, 8, 24-27</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/1993_rev_tibs">www</A></TD>	<TD WIDTH=50>Progress in protein structure prediction?</TD>	<TD WIDTH=90>B Rost, C Sander and R Schneider (1993) TIBS, 18, 120-123</TD>	</TR>



<TR><TD COLSPAN=4><STRONG><FONT SIZE="+2"><A NAME="database">Databases</A></FONT></STRONG></TD</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_nar_nls/">www</A></TD>	<TD WIDTH=50>NLSdb: database of nuclear localization signals</TD>	<TD WIDTH=90>R Nair, P Carter and B Rost (2003) Nucl Acid Res, 322, 397-399</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_nar_pep/">www</A></TD>	<TD WIDTH=50>PEP: Predictions for Entire Proteomes</TD>	<TD WIDTH=90>P Carter, J Liu and B Rost (2003) Nucl Acid Res, 31, 410-413</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2003_narweb_loc3d/">www</A></TD>	<TD WIDTH=50>LOC3D: annotate sub-cellular localization for protein structures</TD>	<TD WIDTH=90>R Nair and B Rost (2003) Nucl Acids Res, 31, pages_xx</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>EVA: evaluation of protein structure prediction servers</TD>	<TD WIDTH=90>IYY Koh, VA Eyrich, MA Mart’-Renom, D Przybylski, MS Madhusudhan, A Fiser, F Pazos, A Valencia, A Sali and B Rost (2003) Nucl Acids Res,  submitted</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>Better prediction of sub-cellular localization through evolution and structure</TD>	<TD WIDTH=90>R Nair and B Rost (2003) Proteins,  submitted</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>CAFASP3 in the spotlight of EVA</TD>	<TD WIDTH=90>VA Eyrich, IYY Koh, D Przybylski, O Grana, F Pazos, A Valencia and B Rost (2003) Proteins,  submitted</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/2000_nls/">www</A></TD>	<TD WIDTH=50>Finding nuclear localization signals</TD>	<TD WIDTH=90>M Cokol, R Nair and B Rost (2000) EMBO Reports, 1, 411-415</TD>	</TR>
<TR><TD WIDTH=20><A HREF="http://cubic.bioc.columbia.edu/papers/">www</A></TD>	<TD WIDTH=50>Twilight zone of protein sequence alignments</TD>	<TD WIDTH=90>B Rost (1999) Protein Engineering, 12, 85-94</TD>	</TR>
<TR><TD WIDTH=20>-</TD>	<TD WIDTH=50>TOPITS: Threading One-dimensional Predictions Into Three-dimensional Structures</TD>	<TD WIDTH=90>B Rost (1995) Third International Conference on Intelligent Systems for Molecular Biology,  314-321</TD>	</TR>
</TABLE>
<P>
<!-- end: table  -->
<!-- ================================================================================ -->


<!-- ================================================================================ -->
<!-- beg: page info  -->
<!--   -->
<TABLE CELLPADDING=3>
<TR VALIGN="BOTTOM">
<TD>	Last update: </TD>
<TD>	<DATE>Mar 25, 2003</DATE></TD>
<TD> 	Contact CUBIC WWW administrator: </TD>
<TD> 	<ADDRESS> <A HREF="mailto:cubic@cubic.bioc.columbia.edu">cubic@cubic.bioc.columbia.edu</A></ADDRESS> </TD></TR>
</TABLE>
<!--   -->
<!-- end: page info  -->
<!-- ================================================================================ -->


<P>

<!-- ================================================================================ -->
<!-- beg: bottom navigate buttons  -->
<!--   -->
<A NAME="bottom">

<A HREF="#top">				Top</A> - 
<A HREF="list.html">			Previous</A> - 
<A HREF="list_year.html">			Next</A> - 
<A HREF="http://cubic.bioc.columbia.edu/">			CUBIC</A> -
<A HREF="http://dodo.bioc.columbia.edu/predictprotein/predictprotein.html">	PP home</A>
  <A HREF="index.html">CUBIC papers</A> -  <A HREF="list_year.html">Papers by year</A> -  <A HREF="list_ascii.txt">Papers as flat file</A>

</P></A>
<!--   -->
<!-- end: bottom navigate buttons  -->
<!-- ================================================================================ -->



</BODY>
</HTML>
