<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Draft//EN">
<h2>Quotes for methods</h2>
<ol>

<li><a NAME="quote_pp"><strong><font SIZE="+1">PredictProtein</font>: </a> <cite>B Rost,G Yachdav and J Liu (2004) The PredictProtein Server. Nucleic Acids Research 32(Web Server issue):W321-W326.
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	help@predictprotein.org
<li> URL:		<a href="http://www.predictprotein.org">http://www.predictprotein.org</a>
<li> Version:	10.20.04
<li> Description:	PredictProtein is the acronym for all prediction programs run.
</ul>



<li><a NAME="quote_pp"><strong><font SIZE="+1">PROSITE</font>: </a> <cite>A Bairoch, P Bucher &and; K Hofmann (1997) Nucleic Acids Research, 25:217-221
</cite></strong>
<ul>
<li> Author:	A Bairoch, bairoch@cmu.unige.ch P Bucher &and; K Hofmann
<li> Contact:	bairoch@cmu.unige.ch
<li> URL:		<a href="http://www.expasy.ch/prosite">http://www.expasy.ch/prosite</a>

<li> Version:	99.07
<li> Description:	PROSITE is a database of functional motifs. ScanProsite, finds all functional motifs in your sequence that are annotated in the ProSite db.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">SEG</font>: </a> <cite>J C Wootton &and; S Federhen (1996) Methods in Enzymology, 266:554-571
</cite></strong>
<ul>
<li> Author:	J C Wootton &and; S Federhen, wootton@ncbi.nlm.nih.gov
<li> Contact:	help@predictprotein.org 
<li> URL:		<a href="wootton@ncbi.nlm.nih.gov">wootton@ncbi.nlm.nih.gov</a>

<li> Version:	1994
<li> Description:	SEG divides sequences into regions of low-, and high-complexity. Low-complexity regions typically correspond to 'simple sequences' or 'compositionally-biased' regions.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">ProDom</font>: </a> <cite>ELL Sonnhammer &and; D Kahn (1994) Protein Science, 3:482-492 
</cite></strong>
<ul>
<li> Author:	LL Sonnhammer; J Gouzy, F Corpet, F Servant, D Kahn, dkahn@zyx.toulouse.inra.fr
<li> Contact:	dkahn@zyx.toulouse.inra.fr
<li> URL:		<a href="http://protein.toulouse.inra.fr/prodom.html">http://protein.toulouse.inra.fr/prodom.html</a>

<li> Version:	2000.1
<li> Description:	ProDom is a database of putative protein domains. The database is searched with BLAST for domains corresponding to your protein.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">PHD</font>: </a> <cite>B Rost (1996) Methods in Enzymology, 266:525-539
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	help@predictprotein.org
<li> URL:		<a href="http://cubic.bioc.columbia.edu">http://cubic.bioc.columbia.edu</a>

<li> Version:	1.96
<li> Description:	PHD is a suite of programs predicting 1D structure (secondary structure, solvent accessibility) from multiple sequence alignments.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">PHDsec</font>: </a> <cite>B Rost &and; C Sander (1993) J. of Molecular Biology, 232:584-599
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	help@predictprotein.org
<li> URL:		<a href="http://cubic.bioc.columbia.edu">http://cubic.bioc.columbia.edu</a>

<li> Version:	1.96
<li> Description:	PHDsec predicts secondary structure from multiple sequence alignments.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">PHDacc</font>: </a> <cite>B Rost &and; C Sander (1994) Proteins, 20:216-226
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	help@predictprotein.org


<li> Version:	1.96
<li> Description:	PHDacc predicts per residue solvent accessibility from multiple sequence alignments.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">PHDhtm</font>: </a> <cite>B Rost, P Fariselli &and;  R Casadio (1996) Protein Science, 7:1704-1718 
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	help@predictprotein.org
<li> URL:		<a href="http://cubic.bioc.columbia.edu">http://cubic.bioc.columbia.edu</a>

<li> Version:	1.96
<li> Description:	PHDhtm predicts the location and topology of transmembrane helices from multiple sequence alignments.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">PROF</font>: </a> <cite>B Rost (2004) Meth. Mol. Biol., submitted.
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	help@predictprotein.org


<li> Version:	2000_04
<li> Description:	PROF is a suite of programs predicting 1D structure (secondary structure, solvent accessibility) from multiple sequence alignments.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">PROFsec</font>: </a> <cite>B Rost (2004) Meth. Mol. Biol., submitted.
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	help@predictprotein.org
<li> URL:		<a href="http://cubic.bioc.columbia.edu ">http://cubic.bioc.columbia.edu </a>

<li> Version:	2000_04
<li> Description:	PROFsec predicts secondary structure from multiple sequence alignments.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">PROFACC</font>: </a> <cite>B Rost (2004) Meth. Mol. Biol., submitted.
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	help@predictprotein.org
<li> URL:		<a href="http://cubic.bioc.columbia.edu">http://cubic.bioc.columbia.edu</a>

<li> Version:	2000_04
<li> Description:	PROFacc predicts per residue solvent accessibility from multiple sequence alignments.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">GLOBE</font>: </a> <cite>B Rost (1998) unpublished
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	help@predictprotein.org
<li> URL:       <a href="http://cubic.bioc.columbia.edu/papers/1999_globe/paper.html">http://cubic.bioc.columbia.edu/papers/1999_globe/paper.html</a>

<li> Version:	1.98.05
<li> Description:	GLOBE predicts the globularity of a protein

</ul>
<!--
<li><a NAME="quote_pp"><strong><font SIZE="+1">CYSPRED</font>: </a> <cite>Fariselli P, Riccobelli P &and; Casadio R (1999) PROTEINS 36:340-346
</cite></strong>
<ul>
<li> Author:	Fariselli P, Riccobelli P, Casadio R
<li> Contact:	piero@lipid.biocomp.unibo.it
<li> URL:		<a href="http://prion.biocomp.unibo.it/cyspred.html">http://prion.biocomp.unibo.it/cyspred.html</a>

<li> Version:	0
<li> Description:	CYSPRED finds whether the cys residue in your protein forms disulfide bridge.

</ul>
-->
<li><a NAME="quote_pp"><strong><font SIZE="+1">DISULFIND</font>: </a> <cite>A.Ceroni, P.Frasconi, A.Passerini and A.Vullo (2004) Bioinformatics, 20, 653-659, 2004
</cite></strong>
<ul>
<li> Author:	A.Ceroni, P.Frasconi, A.Passerini and A.Vullo
<li> Contact:	cystein@dsi.unifi.it
<li> URL:	<a href="http://cassandra.dsi.unifi.it/cysteines/index.html">http://cassandra.dsi.unifi.it/cysteines/index.html</a>

<li> Version:	2.0
<li> Description: DISULFIND is a disulphide bridges predictor based on a two steps process.

</ul>

<li><a NAME="quote_pp"><strong><font SIZE="+1">A conformational switch prediction program</font>: </a> <cite>Young et al. Protein Science(1999) 8:1752-64.
</cite></strong>
<ul>
<li> Author:	Young M, Kirshenbaum K, Dill KA and Highsmith S.
<li> Contact:	mmyoung@sandia.gov, kent@cheme.caltech.edu, shighsmith@sf.uop.edu


<li> Version:	1.0
<li> Description:	ASP finds regions that are most likely to behave as switches in proteins known to exhibit this behavior

</ul>
<li> <a NAME="quote_pp"><strong><font SIZE="+1">HMMPFAM:</font></a> <cite>Bateman et al. Nucleic Acids Research 2004 32:D138-D141.</cite></strong>
<ul>
<li> Author:	Bateman A, Coin L, Durbin R, Finn RD, Hollich V, Griffiths-Jones S, Khanna A, Marshall M, Moxon S, Sonnhammer EL, Studholme DJ, Yeats C, Eddy SR. 
<li> Contact:   agb@sanger.ac.uk

<li> Version:	2.2g
<li> Description:	Search one or more sequences against HMM database
</ul>


<li><a NAME="quote_pp"><strong><font SIZE="+1">DISULFIND</font>: </a> <cite>A.Ceroni, P.Frasconi, A.Passerini and A.Vullo (2004) Bioinformatics, 20, 653-659, 2004
</cite></strong>
<ul>
<li> Author:	A.Ceroni, P.Frasconi, A.Passerini and A.Vullo
<li> Contact:	cystein@dsi.unifi.it
<li> URL:	<a href="http://cassandra.dsi.unifi.it/cysteines/index.html">http://cassandra.dsi.unifi.it/cysteines/index.html</a>

<li> Version:	2.0
<li> Description: DISULFIND is a disulphide bridges predictor based on a two steps process.

</ul>


<li><a NAME="quote_pp"><strong><font SIZE="+1">NORS</font>: </a> <cite>Liu J, Rost B (2003) NORSp: predictions of long regions without regular secondary structure. Nucleic Acids Research 31(13):3833-3835
</cite></strong>
<ul>
<li> Author:	J. Liu
<li> Contact:	Jinfeng Liu <jinfeng.liu@gmail.com>
<li> URL:	<a href="http://cubic.bioc.columbia.edu/services/NORSp/">http://cubic.bioc.columbia.edu/services/NORSp/</a>

<li> Version:	1.0
<li> Description: NORSp is a predictor of NOn-Regular Secondary Structure.

</ul>

<li><a NAME="quote_pp"><strong><font SIZE="+1">CHOP</font>: </a> <cite>Liu J &and; Rost B (2004) CHOP proteins into structural domain-like fragments. Proteins, 55(3):678-688
</cite></strong>
<ul>
<li> Author:	J. Liu
<li> Contact:	Jinfeng Liu <jinfeng.liu@gmail.com>
<li> URL:	<a href="http://cubic.bioc.columbia.edu/services/chop/">http://cubic.bioc.columbia.edu/services/chop/</a>

<li> Version:	1.0
<li> Description: CHOP is a method of dissecting proteins into domain-like fragments based on sequence homology.

</ul>

<li><a NAME="quote_pp"><strong><font SIZE="+1">ISIS</font>: </a> <cite>Yanay Ofran and Burkhard Rost (2007). ISIS: Interaction Sites Identified from Sequence. Bioinformatics. 23  (2), e13-e16
</cite></strong>
<ul>
<li> Author:	Y. Ofran
<li> Contact:	Yanay Ofran <yanay@ofranlab.org>
<li> URL:	<a href="http://cubic.bioc.columbia.edu/services/isis/index.php">http://cubic.bioc.columbia.edu/services/isis/</a>

<li> Version:	1.0
<li> Description: Prediction of residues involved in external protein-protein interactions.
</ul>

<li><a NAME="quote_pp"><strong><font SIZE="+1">DISIS</font>: </a> <cite>Yanay Ofran and Venkatesh Mysore and Burkhard Rost (2007). Prediction of DNA-binding residues from sequence. Bioinformatics. 23  (13), i347-i353
</cite></strong>
<ul>
<li> Author:	Y. Ofran
<li> Contact:	Yanay Ofran <yanay@ofranlab.org>
<li> URL:	<a href="http://cubic.bioc.columbia.edu/services/disis/">http://cubic.bioc.columbia.edu/services/disis/</a>

<li> Version:	1.0
<li> Description: Prediction of residues involved in DNA binding.
</ul>

<li><a NAME="quote_pp"><strong><font SIZE="+1">NORSnet</font>: </a> <cite>Avner Schlessinger and Jinfeng Liu and Burkhard Rost (2007). Natively unstructured loops differ from other loops. PLoS Computational Biology. 3  (7), e140.
</cite></strong>
<ul>
<li> Author:	A. Schlessinger
<li> Contact:	Avner Schlessinger <avnersch@gmail.com>
<li> URL:	<a href="http://predictprotein.org/">http://predictprotein.org/</a>

<li> Version:	1.0.4
<li> Description: Identifies unstructured loops from sequence.
</ul>

<li><a NAME="quote_pp"><strong><font SIZE="+1">PROFbval</font>: </a> <cite>Avner Schlessinger and Guy Yachdav and Burkhard Rost (2006). PROFbval: predict flexible and rigid residues in proteins. Bioinformatics. 22  891-893.</cite></strong>
<ul>
<li> Author:	A. Schlessinger
<li> Contact:	Avner Schlessinger <avnersch@gmail.com>
<li> URL:	<a href="http://cubic.bioc.columbia.edu/newwebsite/services/Profbval">http://cubic.bioc.columbia.edu/newwebsite/services/Profbval</a>

<li> Version:	1.0.4
<li> Description: Prediction of protein flexibility and rigidity prediction from sequence.
</ul>

<li><a NAME="quote_pp"><strong><font SIZE="+1">METADISORDER</font>: </a> <cite>A. Schlessinger and M. Punta and G. Yachdav and L. Kajan and B. Rost (2009). Improved disorder prediction by combination of orthogonal approaches. PLoS One. 4  (2), e4433.</cite></strong>
<ul>
<li> Author:	A. Schlessinger
<li> Contact:	Avner Schlessinger <avnersch@gmail.com>
<li> URL:	<a href="http://cubic.bioc.columbia.edu/newwebsite/services/md/">http://cubic.bioc.columbia.edu/newwebsite/services/md/index.php</a>

<li> Version:	1.0.3
<li> Description: Protein disorder prediction based on orthogonal sources of information.
</ul>

<li><a NAME="quote_pp"><strong><font SIZE="+1">PROFcon</font>: </a> <cite>Marco Punta and Burkhard Rost (2005). PROFcon: novel prediction of long-range contacts. Bioinformatics. 21  (13), 2960-2968</cite></strong>
<ul>
<li> Author:	M. Punta
<li> Contact:	Marco Punta <punta@rostlab.org>
<li> URL:	<a href=http://cubic.bioc.columbia.edu/services/profcon/">http://cubic.bioc.columbia.edu/services/profcon/</a>

<li> Version:	1.0
<li> Description: contact prediction method.
</ul>

<li><a NAME="quote_pp"><strong><font SIZE="+1">PROFtmb</font>: </a> <cite>Henry Bigelow and Burkhard Rost (2006). PROFtmb: a web server for predicting bacterial transmembrane beta barrel proteins. Nucleic Acids Research. 34  (Web Server issue), W186-188.</cite></strong>
<ul>
<li> Author:	H. Bigelow
<li> Contact:	Henry Bigelow <hrbigelow@gmail.com>
<li> URL:	<a href=http://cubic.bioc.columbia.edu/services/proftmb/">http://cubic.bioc.columbia.edu/services/proftmb/</a>

<li> Version:	1.1.1
<li> Description: per-residue prediction of bacterial transmembrane beta barrels.
</ul>

<li><a NAME="quote_pp"><strong><font SIZE="+1">SNAP</font>: </a> <cite>Yana Bromberg and Guy Yachdav and Burkhard Rost (2008). SNAP predicts effect of mutations on protein function. Bioinformatics. in press.</cite></strong>
<ul>
<li> Author:	Y. Bromberg
<li> Contact:	Yana Bromberg <bromberg@rostlab.org>
<li> URL:	<a href=http://cubic.bioc.columbia.edu/services/snap/">http://cubic.bioc.columbia.edu/services/snap/</a>

<li> Version:	1.0.8
<li> Description: a method for evaluating effects of single amino acid substitutions on protein function.
</ul>

<li><a NAME="quote_pp"><strong><font SIZE="+1">LOCtree</font>: </a> <cite>Rajesh Nair and Burkhard Rost (2005). Mimicking cellular sorting improves prediction of subcellular localization. Journal of Molecular Biology. 348  (1), 85-100</cite></strong>
<ul>
<li> Author:	R. Nair 
<li> Contact:	Rajesh Nair <rajnair5@gmail.com>
<li> URL:	<a href=http://cubic.bioc.columbia.edu/services/loctree/">http://cubic.bioc.columbia.edu/services/loctree/</a>

<li> Version:	 1.0.3
<li> Description: predict the subcellular localization of proteins.
</ul>

</ol>
