
<HTML>
<HEAD>
<TITLE>MView: BLAST HSP processing</TITLE>
</HEAD>
<BODY BGCOLOR="white">


<h3>The rules used in MView for selecting and tiling HSPs</h3> 

<P>

A BLAST run comprises, among other sections, a 'ranking' of hits with
one-line summaries including scoring information, followed by the search
database 'hits' themselves. Each hit may contain one or more HSPs (ungapped
BLAST) or gapped alignments (gapped BLAST), not all of which may be
reported in the ranking.

<P>

Some control over which HSPs are to be processed by MView is provided by
the command line option <code>-hsp mode</code>. There are three choices:
<dl>
<dt>
<code>ranked</code>
<dd>
This is the default behaviour. Only those HSPs or alignments that
contributed to the ranked hit are selected. These are tiled into a single
alignment row per hit.

<dt>
<code>all</code>
<dd>
All HSPs or alignments are selected and tiled into a single alignment row
per hit.

<dt>
<code>discrete</code>
<dd>
All HSPs or alignments are selected but each is assigned its own row in the
resulting stacked alignment. 

</dl>

Precise application of these processing modes varies slightly with each
BLAST program because properties such as query and/or hit nucleotide strand
orientation or reading frame must be taken into consideration

<P>

A more detailed and specific description of the selection rules is given
for each BLAST program for each series recognised by MView:

<ul>
<li><a href="blast1.html">NCBI BLAST 1.4  (ungapped)</a>
<li><a href="blast1.html">WU-BLAST 2.0 (gapped)</a>
<li><a href="blast2.html">NCBI BLAST 2.0 and PSI-BLAST (gapped or ungapped)</a>
</ul>


<h3>General notes</h3>

<dl>

<dt>
Assemble (simple)
<dd>
Given a list of fragments each numbered from low to high along the query
strand (ie., conventional orientation): stack these at these positions
along the query template.

<dt>
Assemble (reverse)
<dd>
Given a list of fragments each numbered from high to low along the query
strand (ie., individually reversed): reverse each fragment, stack onto the
template, reverse the template, which recovers the original numbering high
to low and preserves intra- and inter-fragment order.

<dt>
alignmentnum
<dd>
The ordinal number (starting at 1) of a single HSP within a given hit
(starting afresh with each hit) as encountered in the raw BLAST output.

<dt>
bits
<dd>
NCBI BLAST 2.0 hit ranking rounds HSP scores which are floating point to
the nearset integer, so MView takes this into account when selecting HSPs.

<dt>
Filter
<dd>
Separate MView output must be produced for varous BLAST hit properties, such 
as PSI-BLAST search cycle, BLASTN query strand orientation, etc.

<dt>
hitnum
<dd>
The ordinal number (starting at 1) of a search hit as encountered in the
ranking section of the raw BLAST output.

<dt>
N
<dd>
A count of HSPs which is context dependent in MView. In raw BLAST output it
is the number of HSPs deemed by BLAST to belong in a coherent set of
alignment fragments, but the MView HSP processing options (ranked, all,
discrete) change the precise meaning.

<dt>
P(N)
<dd>
The BLAST alignment p-value associated with a set of N HSPs.

<dt>
p-value
<dd>
BLAST 1.4 series reports relatively large p-vales in floating point format
rather than scientific notation. The original HSP p-value may be rounded to
2 decimal places in the hit ranking, so comparison of these values by MView
when selecting HSPs takes this into account, which could lead to
over-selection of HSPs on occasion.

<dt>
Renumber
<dd>
Given a (start,stop) pair in nucleotide units, the corresponding nearest
values in amino acid units are given by:
<pre>
<code>  (start', stop') = ( int((start+2)/3), int(stop/3) )  </code>
</pre>
This renumbering is applied to each fragment's  positions along its parent query
strand to determine the query template positions in amino acid units.

<dt>
score
<dd>
The BLAST alignment score.

</dl>


<HR>
 <A HREF="http://www.nimr.mrc.ac.uk">[NIMR]</A>
 <A HREF="http://mathbio.nimr.mrc.ac.uk">[MathBio]</A>
<HR>
<ADDRESS><SMALL>Maintained by <A HREF="MAILTO:nbrown@nimr.mrc.ac.uk">Nigel Brown</A>.</SMALL> <SMALL>Last update Dec  1 1999
</SMALL>. <A HREF="http://mathbio.nimr.mrc.ac.uk/disclaimer.html"><SMALL>Disclaimer</SMALL></A></ADDRESS>
</BODY></HTML>
