<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta name="generator"
    content="HTML Tidy for Linux/x86 (vers 1st October 2002), see www.w3.org" />

    <title></title>
  </head>

  <body>
<pre>
                README for netblast (network-client BLAST)
                        Last updated 2/23/04

This README describes the BLAST client (blastcl3), which accesses the
NCBI BLAST search engine (version 2.2).  The software behind BLAST 
version 2.2 allows BLAST to handle new challenges posed by the 
sequence databases.  Updates to this software will continue in the 
coming years.

1. Introduction

Blastcl3 can search all the sequences in a FASTA file, produce 
one-to-many (master-slave) alignments in text or HTML format. It can 
also perform searches against multiple databases.  It also supports 
organism-specific searches via the -u option, which functions like the 
"Limit by Entrez Query Term" in the NCBI BLAST web pages. The old -l 
option is removed.  A full list of command-line options is given in 
Appendix B.

The setup for NCBI network clients has been greatly simplified. If you
are not behind a firewall no further action is required.  If you are 
behind a firewall, and already use Sequin or Entrez, or if your system
administrator has already performed the setup, then you should be able 
to start performing searches immediately.  Otherwise, please refer to
'firewall.html' that is included with this archive for instructions on 
firewall setups.

As an alternative to blastcl3, NCBI BLAST web server also supports
URL API, which uses url encoded command to interact with Blast.cgi to
"put" search requests and "get" search results. For details on the 
standard commands, please refer to the online document at:

        http://www.ncbi.nlm.nih.gov/BLAST/Doc/urlapi.pdf


2. Basic commandline for running blastcl3

Blastcl3 may be used to perform all five flavors of blast comparison. 
For large blastn searches one can choose to use MEGABLAST algorithm 
(-n T), which speeds up the search by concatenating the queries to 
reduce database scan time. Please refer to Appendix B for the complete
list of program options for blastcl3. The list can also be displayed 
by executing 'blastcl3 -' (the dash) command.  A typical use of 
blastcl3 is to perform a blastn search (nucleotide vs. nucleotide) of
a set of FASTA formatted sequences saved in a text file called QUERY:

        blastcl3 -p blastn -d nr -i QUERY -o out.QUERY

The output is placed in the output file out.QUERY and the search is 
performed against the 'nr' database. If a protein vs. protein search 
is desired, 'blastn' should be replaced with 'blastp' etc. The 
commandline below will search a query file named MY_QUERY against
protein nr database and place the result in a file called MY_QUEYR.out:

        blastcl3 -p blastp -d nr -i MY_QUEYR -o MY_QUERY.out

One should issue the blastcl3 command under the blastcl3 containing 

directory. For unix/linux system, ./ may need to be appended immediately
before the blastcl3 to instruct shell to look at the current directory.

One can also perform tther type of searches using different -p call
combined with correct -i/-d option value.


3. Commonly used options

The users need to use -i to specify the input file.  Other options not
specified will assume default value.  Often it is necessary to used 
custom parameter values to ensure optimal results. The commonly used 
blastcl3 options and their explanations are given below for your 
reference.

  -p  Program Name [String]
        Input should be one of "blastp", "blastn", "blastx", 
        "tblastn", or "tblastx".

  -d  Database [String]
        default = nr

        Multiple database names (bracketed by quotations) will be 
        accepted. An example would be:

        -d "nr est" 

        which searches both the nr and est databases, presenting the 
        results as if one 'virtual' database consisting of all the 
        entries from both were searched. The statistics are based on 
        the combined 'virtual' database of nr and est. For supported
        database list please refer to Appendix C.

  -i  Query File [File In]
        default = stdin

        The query sequences should be in FASTA format and the file 
        should be saved in plain text format.  If multiple FASTA 
        entries are in the input file, all queries will be searched.
        For query file format, see Appendix A.

  -e  Expectation value (E) [Real]
        default = 10.0
        
        Accepted formats are integer, decimal, fraction (1/6240), 
        and exponent (2e-64)

  -o  BLAST report Output File [File Out]  Optional
          default = stdout [screen]

  -F  Filter query sequence (DUST with blastn, SEG with others) [T/F]
        default = T

        BLAST 2.0 uses the dust low-complexity filter for blastn and 
        seg for the other programs. Both 'dust' and 'seg' are integral 
        parts of the NCBI toolkit and are accessed automatically.

        If one uses "-F T" then normal filtering by seg or dust 
        (for blastn) occurs (likewise "-F F" means no filtering 
        whatsoever).  The seg options can be changed by using:

        -F "S 10 1.0 1.5"

        which specifies a window of 10, locut of 1.0 and hicut of 1.5.
        A coiled-coiled filter, based on the work of Lupas et al. 
        (Science, vol 252, pp. 1162-4 (1991)) and written by John 
        Kuzio (Wilson et al., J Gen Virol, vol. 76, p2923-32 (1995)), 
        may be invoked by specifying:

        -F "C"

        There are three parameters for this: window, cutoff (prob of a 
        coil-coil), and linker (distance between two coiled-coiled 
        regions that should be linked together).  These are now set to

        window: 22
        cutoff: 40.0
        linker: 32

        One may also change the coiled-coiled parameters in a manner
        analogous to that of seg:

        -F "C 28 40.0 32" will change the window to 28.

        One may also run both seg and coiled-coiled together by using
        a ";":

        -F "C;S"

        Filtering by dust may also be specified by:

        -F "D"

        It is possible to specify that the masking should only be done
        during the process of building the initial words by starting 
        the filtering command with 'm':

        -F "m S"


        which specifies that seg (with default arguments) should be 
        used for masking, but that the masking should only be done when
        the words are being built.  This masking option is available 
        with all filters.

        -F "R"
        
        This would invoke the human repeat filter to mask the human
        repeat element present in the query sequences.

        -F "V"
        
        This would invoke the vector filter to mask the vector sequences 
        present in the query sequences.

  -S  Query strands to search against database (for blast[nx], and 
        tblastx).
        [Integer] 
         
        3 is both, 1 is top [input strand], 2 is bottom [reverse complement] 
        default = 3

  -T  Produce HTML output [T/F]
        default = F

  -u  Entrez query terms [Optional]

          Some examples for how to use -u option:
          -u mouse[organism]        
                This limits the search to sequences from mouse.
          -u biomol_mrna[properties]        
                This limits the search to mRNA sequences.
          -u "nucleotide_all[filter] NOT human[organism]"        
                This limits the search to all non-human entries.

For details on the terms, please refer to this web document:
http://www.ncbi.nlm.nih.gov:80/entrez/query/static/help/helpdoc.html


4. Firewall Settings

blastcl3 needs constant internet connection to NCBI blast servers to 
function properly.  There are extra setups if blastcl3 connects to 
NCBI from behind a firewall.  

First, one need to check with local network administrator to make sure 
that the following IP address/port combinations are open in the 
firewall configuration:

IP address         Port
-------------      ----    
130.14.29.112      5861
130.14.29.112      5862 
130.14.29.112      5863 

This ensures that the appropriate path is open in the firewall through 
which blastcl3 can make needed connections to NCBI server.
 
Second, one needs to set environment variables so that blastcl3 is 
aware of the firewall setting.  This can be achieved indirectly by 
creating .ncbirc file (or ncbi.ini for Window machines) containing the 
settings. This file should be placed in appropriate places, such as the 
home directory for unix or linux system, or WINNT directory for Windows 
machine.  This file instructs blastcl3 to use the settings in the file, 
rather than the default settings built into the program itself for 
connecting to NCBI server.

The settings needed are the followings:

[CONN]
FIREWALL=TRUE

If one specifies it by using environment variables, use this:

CONN_FIREWALL=TRUE


5. FASTA Query file format

Query files with multiple query sequences should be put in this format
with defline in a separate line and the sequences put in 70-80 
characters per line.  NOTE the file should be in text format.  If WORD 
or other word processing program is used, make sure the file is saved
as text.

&gt;query_sequence_1
MNTIRNSICLTIITMVLCGFLFPLAITLIGQIFFYQQANGSLITYDNRIVGSKLIGQHWTETRYFHGRPS
AVDYNMNPEKLYKNGVSSGGSNESNGNTELIARMKHHVKFGNSNVTIDAATSSGSGLDPHITVENALKQA
PRIADARHISTSRVADLIQHRKQRGVLTNDYVNVLELNIALDKMKD
&gt;query_sequence_2
MAQPGPAPQPDVSLQQRVAELEKINAEFLRAQQQLEQEFNQKRAKFKELYLAKEEDLKRQNAVLQAAQDD
LGHLRTQLWEAQAEMENIKAIATVSENTKQEAIDEVKRQWREEVASLQAIMKETVRDYEHQFHLRLEQER
AQWAQYRESAEREIADLRRRLSEGQEEENLENEMKKAQEDAEKLRSVVMPMEKEIAALKDKLTEAEDKIK
ELEASKVKELNHYLEAEKSCRTDLEMYVAVLNTQKSVLQEDAEKLRKELHEVCHLLEQERQQHNQLKHTW


6. Complete list of blastcl3 arguments

The complete program options for blastcl3 are listed below for reference
purposes.  Most of the parameters can be left out of the commandline.
In these cases, they will assume default settings, sufficient for normal
blast search purposes.

One can also generate this list by typing "blastcl3 -", without the quotes,
followed by a return key stroke. blastcl3 will print them on the screen.

  -p  Program Name [String]
  -d  Database [String]
    default = nr
  -i  Query File [File In]
    default = stdin
  -e  Expectation value (E) [Real]
    default = 10.0
  -m  alignment view options:
0 = pairwise,
1 = query-anchored showing identities,
2 = query-anchored no identities,
3 = flat query-anchored, show identities,
4 = flat query-anchored, no identities,
5 = query-anchored no identities and blunt ends,
6 = flat query-anchored, no identities and blunt ends,
7 = XML Blast output,
8 = tabular, 
9 = tabular with comment lines [Integer]
    default = 0
  -o  BLAST report Output File [File Out]  Optional
    default = stdout
  -F  Filter query sequence (DUST with blastn, SEG with others)[String]
    default = T
  -G  Cost to open a gap (zero invokes default behavior) [Integer]
    default = 0
  -E  Cost to extend a gap (zero invokes default behavior) [Integer]
    default = 0
  -X  X dropoff value for gapped alignment (in bits) (zero invokes 
      default behavior)
      blastn 30, megablast 20, tblastx 0, all others 15 [Integer]
    default = 0
  -I  Show GI's in deflines [T/F]
    default = F
  -q  Penalty for a nucleotide mismatch (blastn only) [Integer]
    default = -3
  -r  Reward for a nucleotide match (blastn only) [Integer]
    default = 1
  -v  Number of database sequences to show one-line descriptions 
      for (V) [Integer]
    default = 500
  -b  Number of database sequence to show alignments for (B) [Integer]
    default = 250
  -f  Threshold for extending hits, default if zero
      blastp 11, blastn 0, blastx 12, tblastn 13
      tblastx 13, megablast 0 [Integer]
    default = 0
  -g  Perfom gapped alignment (not available with tblastx) [T/F]
    default = T
  -Q  Query Genetic code to use [Integer]
    default = 1
  -D  DB Genetic code (for tblast[nx] only) [Integer]
    default = 1
  -a  Number of processors to use [Integer]
    default = 1
  -O  SeqAlign file [File Out]  Optional
  -J  Believe the query defline [T/F]
    default = F
  -M  Matrix [String]
    default = BLOSUM62
  -W  Word size, default if zero (blastn 11, megablast 28, all others 3)
      [Integer]
    default = 0
  -z  Effective length of the database (use zero for the real size) 
      [Real]
    default = 0
 -K  Number of best hits from a region to keep (off by default, if used
      an value of 100 is recommended) [Integer]
    default = 0
  -Y  Effective length of the search space (use zero for the real size)
      [Real]
    default = 0
  -S  Query strands to search against database (for blast[nx], and
      tblastx) 3 is both, 1 is top, 2 is bottom [Integer]
    default = 3
  -T  Produce HTML output [T/F]
    default = F
  -u  Restrict search of database to results of Entrez2 lookup 
      [String]  Optional
  -U  Use lower case filtering of FASTA sequence [T/F]  Optional
    default = F
  -y  X dropoff value for ungapped extensions in bits (0.0 invokes 
      default behavior) blastn 20, megablast 10, all others 7 [Real]
    default = 0.0
  -Z  X dropoff value for final gapped alignment in bits (0.0 invokes
      default behavior) blastn/megablast 50, tblastx 0, all others 25
      [Integer]
    default = 0
  -R  RPS Blast search [T/F]
    default = F
  -n  MegaBlast search [T/F]
    default = F
  -L  Location on query sequence [String]  Optional
  -A  Multiple Hits window size, default if zero (blastn/megablast 0,
      all others 40 [Integer]
    default = 0
  -w  Frame shift penalty (OOF algorithm for blastx) [Integer]
    default = 0
  -t  Length of the largest intron allowed in tblastn for linking HSPs
      (0 disables linking) [Integer]
    default = 0


Appendix. Currently supported databases 

NOTE: all the database should be called with -d NAME, where the
        NAME should be exactly as it appears in the lists given below!

A. General Nucleotide Sequence Databases

+--------+-----------------------------------------------------------+
|NAME    |CONTENT                                                    |
+--------|-----------------------------------------------------------+
 nr          All GenBank+EMBL+DDBJ sequences (but no EST, STS, 
             GSS, or phase 0, 1 or 2 HTGS sequences). WGS entries are
             also excluded. No longer "non-redundant".  

 est         Database for entries from EST division of 
             GenBank+EMBL+DDBJ.
 
 est_human   Human subset of est. 

 est_mouse   Mouse subset of est. 

 est_others  ESTs other than human or mouse.

 gss         Genome Survey Sequence, includes single-pass genomic data, 
             exon-trapped sequences, and Alu PCR sequences. 

 htgs        Unfinished High Throughput Genomic Sequences: phases 0, 1 
             and 2. Finished, phase 3 HTG sequences are in nr.
 
 pat         Nucleotides from the Patent division of GenBank. 

 month       All new or revised GenBank+EMBL+DDBJ+PDB sequences
             released updated in the last 30 days. 

 alu         Select Alu repeats from REPBASE, suitable for masking 
             Alu repeats from query sequences. 

 sts         Database of GenBank+EMBL+DDBJ sequences from STS Division. 

 chromosome  Complete genomes, complete chromosome, or concatenated 
             genomic contigs form the NCBI Reference Sequence project.

 UniVec      The UniVec non-redundant vector fragment sequences.

 wgs         whole genome shotgun sequence assembly          
+-----------+--------------------------------------------------------+

One can blast search a specific completed bacterial or archeal genomes 
by searching the chromosome database and limit the search to that 
microbe using -u option. For example, to search a query file QUERY 
against Aquifex aeolicus genome, one can use the following commandline: 

blastcl3 -p blastn -i QUERY -d chromosome -u "Aquifex aeolicus[orgn]" -o Out

For list of completed bacterial and archeal genomes, please visit:
  http://www.ncbi.nlm.nih.gov/PMGifs/Genomes/eub.html
  http://www.ncbi.nlm.nih.gov/PMGifs/Genomes/a_g.html
  

B. General peptide sequence databases 

+-----------+--------------------------------------------------------+
|NAME       |CONTENT                                                 |
+-----------+--------------------------------------------------------+
 nr          All non-redundant GenBank CDS translations + PDB 
             + SwissProt + PIR + PRF.  Non-redundant.

 swissprot   Last major release of the SWISS-PROT protein sequence 
             database (no updates) 

 pat         Patent protein sequence database.
 
 pdb         Sequences derived from the 3-dimensional structure
             records from PDB.

 month       All new or revised GenBank CDS translation + PDB + 
             SwissProt + PIR + PRF released in the last 30 days.         
+-----------+--------------------------------------------------------+

C. Specialized Human Sequence Databases

C.1 Human genome related databases
+--------------------------+-----------------------------------------+
|NAME                      |CONTENT                                  |
+--------------------------+-----------------------------------------+
 hs_genome/genome           default genome database
 hs_genome/gnomon_prot      Ab nitio (predicted) protein sequences
 hs_genome/gnomon_rna       Ab nitio (predicted) mRNA sequences
 hs_genome/hs_bes           human BAC end sequences
 hs_genome/hs_est           human EST sequences
 hs_genome/hs_htg           human HTG sequences
 hs_genome/hs_refm          RefSeq mRNA sequences
 hs_genome/hs_refp          RefSeq protein sequences
 hs_genome/protein          Build RNA
 hs_genome/rna              Build protein
 Trace/Homo_Sapiens_EST     EST Traces
 Trace/Homo_Sapiens_Others  Other Traces
+--------------------------+-----------------------------------------+

C.2 Human SNP databases
+--------------------------+-----------------------------------------+
|NAME                      |CONTENT                                  |
+--------------------------+-----------------------------------------+
 snp/snpch1.fas             human snps mapped to Chromosome 1
 snp/snpch2.fas             human snps mapped to Chromosome 2
 snp/snpch3.fas             human snps mapped to Chromosome 3
 snp/snpch4.fas             human snps mapped to Chromosome 4
 snp/snpch5.fas             human snps mapped to Chromosome 5
 snp/snpch6.fas             human snps mapped to Chromosome 6
 snp/snpch7.fas             human snps mapped to Chromosome 7
 snp/snpch8.fas             human snps mapped to Chromosome 8
 snp/snpch9.fas             human snps mapped to Chromosome 9
 snp/snpch10.fas            human snps mapped to Chromosome 10
 snp/snpch11.fas            human snps mapped to Chromosome 11
 snp/snpch12.fas            human snps mapped to Chromosome 12
 snp/snpch13.fas            human snps mapped to Chromosome 13
 snp/snpch14.fas            human snps mapped to Chromosome 14
 snp/snpch15.fas            human snps mapped to Chromosome 15
 snp/snpch16.fas            human snps mapped to Chromosome 16
 snp/snpch17.fas            human snps mapped to Chromosome 17
 snp/snpch18.fas            human snps mapped to Chromosome 18
 snp/snpch19.fas            human snps mapped to Chromosome 19
 snp/snpch20.fas            human snps mapped to Chromosome 20
 snp/snpch21.fas            human snps mapped to Chromosome 21
 snp/snpch22.fas            human snps mapped to Chromosome 22
 snp/snpchX.fas.            human snps mapped to Chromosome X
 snp/snpchY.fas.nsq         human snps mapped to Chromosome Y
 snp/snpchMulti.fas         human snps mapped to multiple chromosomes
 snp/snpchNotOn.fas         human snps not yet mapped to chromosome
 snp/snpchMasked.fas        repeat masked human snps
 snp/snp                    All the above
+--------------------------+-----------------------------------------+

D. Specialized Mouse Sequence databases
+--------------------------+----------------------------------------+
|NAME                      |CONTENT                                 |
+--------------------------+----------------------------------------+
 mouse_contig/genome        current mouse genome assembly
 mouse_contig/gnomon_prot   predicted proteins from the above genome
 mouse_contig/gnomon_rna    predicted mRNAs from the above genome
 mouse_contig/mm_bes        mouse bac end sequences (from gss)
 mouse_contig/mm_est        mouse est (est_mouse)
 mouse_contig/mm_htg        mouse htgs sequeneces, including phase 3
 mouse_contig/mm_refm       mouse refseq mRNA (updated daily)
 mouse_contig/mm_refp       mouse refseq protin (updated daily)
 mouse_contig/protein       mouse build mRNAs
 mouse_contig/rna           mouse proteins from build mRNAs
 Trace/Mus_Musculus_EST     EST Traces
 Trace/Mus_Musculus_others  Other Traces
+--------------------------+-----------------------------------------+

E. Specialized Rat Sequence databases
+---------------------------+----------------------------------------+
|NAME                       |CONTENT                                 |
+---------------------------+----------------------------------------+
 rn_genome/genome            latest public rat genome build
 rn_genome/gnomon_prot       proteins from predicted mRNAs
 rn_genome/gnomon_rna        mRNAs predicted off the genome build
 rn_genome/protein           proteins from build mRNA 
 rn_genome/rn_bes            rat BAC end sequences
 rn_genome/rn_est            rat EST sequences
 rn_genome/rn_htg            rat htgs, including phase 3
 rn_genome/rn_refm           rat refseq mRNA, updated daily
 rn_genome/rn_refp           rat refseq protein, updated daily
 rn_genome/rna               rat build mRNAs
 rn_genome/wgs_contigs       rat wgs contigs
+--------------------------+-----------------------------------------+

F. Specialized Zebra fish and Puffer fish Sequence databases
+---------------------------+----------------------------------------+
|NAME                       |CONTENT                                 |
+---------------------------+----------------------------------------+
 dr_genome/dr_mrna           Zebra fish mRNA
 dr_genome/dr_htg            HTGS phase 0, 1, 2, or 3 sequences.
 dr_genome/dr_est            subset of Zebra fish EST from the EST database
 dr_genome/dr_refm           Reference mRNAs from RefSeq project
 dr_genome/dr_refp           Reference Proteins from RefSeq project
 Trace/Danio_Rerio_WGS       Zebra fish whole genome shotgun Traces
 Trace/Danio_Rerio_EST       Zebra fish EST Traces

 genomes/fugu                Fugu genome
+---------------------------+----------------------------------------+


G. Specialized Insects and Worm Sequence databases
+---------------------------+----------------------------------------+
|NAME                       |CONTENT                                 |
+---------------------------+----------------------------------------+
 genomes/ag_mrna              Mosquito mRNA predicted off the genome
 genomes/agambiae             Mosquito genome scaffold
 genomes/agambiae             Mosquito proteins from predicted mRNA
 genomes/dm                   Drosophila melanogaster mRNAs
 genomes/dm                   Drosophila melanogaster proteins
 genomes/dm_genomic           Drosophila melanogaster genomic contigs
 genomes/dpWGS                Drosophila pseudoobscura WGS contigs

 genomes/hbWGS                honeybee wgs contigs
 genomes/insects_genomic      insects' genomic sequences

 genomes/c_briggsae           C. briggsae genomic sequences
 genomes/c_elegans            C. elegans predicted mRNA
 genomes/c_elegans            C. elegans protein from predicted mRNA
 genomes/c_elegans_chr        C. elegans chromosomes
+---------------------------+----------------------------------------+

H. Specialized Fungi, Plants genome database
+---------------------------+----------------------------------------+
|NAME                       |CONTENT                                 |
+---------------------------+----------------------------------------+
 genomes/Giardia             Giardia lamblia genomic sequences
 genomes/Giardia             Giardia lamblia protein sequences
 genomes/anidulans           Aspergillus nidulans DNA sequences
 genomes/anidulans           Aspergillus nidulans protein sequences
 genomes/cionaWGS            Ciona intestinalis wgs sequences

 genomes/p_falciparum        P. falciparum genome
 genomes/p_falciparum        P. falciparum protein
 genomes/p_yoelii            P. yoelii genome
 genomes/p_yoelii            P. yoelii protein

 genomes/s_bayanusMIT        S. bayanus genome(MIT) 
 genomes/s_bayanusWashU      S. bayanus genome(WUSTL)
 genomes/s_castellii         S. castellii genome
 genomes/s_kluyveri          S. kluuyveri genome
 genomes/s_kudriavzevii      S. kudriavzevii genome
 genomes/s_mikataeMIT        S. mikatae genome (MIT)
 genomes/s_mikataeWashU      S. mikatae genome (WUSTL)
 genomes/s_paradoxus         S. paradoxus genome
 genomes/scerevis            S. cerevisiae genome
 genomes/scerevis            S. cerevisiae proteins
 genomes/spombe              S. pombe genome
 genomes/spombe              S. pombe proteins
 genomes/nc_comp             Neurospora genome
 genomes/nc_mrna             Neurospora mRNAs
 genomes/nc_prot             Neurospora proteins
 genomes/fungi               all fungal genomes
 genomes/fungi               all predicted fungal proteins
 genomes/fungi_mrna          all predicted fungal mRNAs

 genomes/ara                 Arabidopsis mRNAs
 genomes/ara.pal             Arabidopsis proteins
 genomes/ara_chr             Arabidopsis chromosome
 genomes/ara_clone           Arabidopsis genomic clones
 genomes/barley              barley mRNA sequences              
 genomes/barley_clone        barley genomic clones
 genomes/barley_est          barley est entries
 genomes/barley_sts          barley sts entries
 genomes/corn                corn mRNA sequences
 genomes/corn_clone          corn genomic clones
 genomes/corn_est            corn est entries
 genomes/corn_gss            corn gss entries
 genomes/corn_patnt          corn patent nucleotides
 genomes/corn_sts            corn sts entries
 genomes/oat                 oat mRNAs
 genomes/oat_clone           oat genomic clones
 genomes/oat_est             oat est entries
 genomes/oat_sts             oat sts entries
 genomes/rice            ??  rice mRNA sequences
 genomes/rice                rice protein sequences
 genomes/riceChWGS           rice wgs assemblies
 genomes/rice_clone          rice genomic contigs
 genomes/rice_est            rice est entries
 genomes/rice_gss            rice gss entries
 genomes/rice_htgs           rice htgs entries
 genomes/rice_mrna           rice mRNA sequences
 genomes/rice_sts            rice sts entries
 genomes/soybean             Glycine max mRNA sequences
 genomes/soybean_clone       Glycine max genomic clones
 genomes/soybean_est         Glycine max est entries
 genomes/tomato              Tomato mRNA sequences
 genomes/tomato_est          Tomato est entries
 genomes/wheat               Wheat mRNA sequences
 genomes/wheat_clone         Wheat genomic clones
 genomes/wheat_est           Wheat est entries
 genomes/wheat_gss           Wheat gss entries
 genomes/wheat_sts           Wheat sts entries
 genomes/plant               genomic sequences for plants
                               listed here
+---------------------------+------------------------------------+

I. Microbial, archaeal, and lower eukaryotic genome databases
+------------------+---------------------------------------------+
| NAME             |           CONTENT                           |
+------------------+---------------------------------------------+
 Microbial/100226        Streptomyces coelicolor A3(2)
 Microbial/100379        Onion yellows phytoplasma
 Microbial/103690        Nostoc sp. PCC 7120
 Microbial/1063          Rhodobacter sphaeroides
 Microbial/1076          Rhodopseudomonas palustris
 Microbial/107806        Buchnera aphidicola str. APS            
                           (Acyrthosiphon pisum)
 Microbial/1085          Rhodospirillum rubrum
 Microbial/1108          Chloroflexus aurantiacus
 Microbial/111955        Sulfolobus tokodaii
 Microbial/1148          Synechocystis sp. PCC 6803
 Microbial/115711        Chlamydophila pneumoniae AR39
 Microbial/115713        Chlamydophila pneumoniae CWL029
 Microbial/117           Pirellula sp.
 Microbial/119072        Thermoanaerobacter tengcongensis
 Microbial/119219        Ralstonia metallidurans
 Microbial/122586        Neisseria meningitidis MC58
 Microbial/122587        Neisseria meningitidis Z2491
 Microbial/1282          Staphylococcus epidermidis
 Microbial/1299          Deinococcus radiodurans
 Microbial/1309          Streptococcus mutans
 Microbial/1336          Streptococcus equi
 Microbial/134537        Burkholderia fungorum
 Microbial/1352          Enterococcus faecium
 Microbial/135720        Neisseria meningitidis serogroup C
 Microbial/1360          Lactococcus lactis subsp. lactis
 Microbial/13773         Pyrobaculum aerophilum
 Microbial/138677        Chlamydophila pneumoniae J138
 Microbial/139           Borrelia burgdorferi
 Microbial/1422          Geobacillus stearothermophilus
 Microbial/1423          Bacillus subtilis
 Microbial/145262        Methanothermobacter thermautotrophicus
 Microbial/1488          Clostridium acetobutylicum
 Microbial/1496          Clostridium difficile
 Microbial/1502          Clostridium perfringens
 Microbial/155864        Escherichia coli O157:H7 EDL933
 Microbial/155919        Xylella fastidiosa Dixon
 Microbial/155920        Xylella fastidiosa Ann-1
 Microbial/156889        Magnetococcus sp. MC-1
 Microbial/158878        Staphylococcus aureus subsp. aureus Mu50
 Microbial/158879        Staphylococcus aureus subsp. aureus N315
 Microbial/159087        Dechloromonas aromatica RCB
 Microbial/159288        Staphylococcus aureus subsp. aureus 252
 Microbial/159289        Staphylococcus aureus subsp. aureus 476
 Microbial/1596          Lactobacillus gasseri
 Microbial/160           Treponema pallidum
 Microbial/160488        Pseudomonas putida KT2440
 Microbial/160489        Pseudomonas putida PRS1
 Microbial/160490        Streptococcus pyogenes M1 GAS
 Microbial/160491        Streptococcus pyogenes Manfredo
 Microbial/160492        Xylella fastidiosa 9a5c
 Microbial/163164        Wolbachia endosymbiont of Drosophila
                           melanogaster
 Microbial/1642          Listeria innocua
 Microbial/164513        Mycobacterium tuberculosis 210
 Microbial/164609        Wigglesworthia glossinidia endosymbiont
                           of Glossina brevipalpis
 Microbial/167539        Prochlorococcus marinus subsp. 
                           marinus str. CCMP1375
 Microbial/167879        Colwellia psychrerythraea 34H
 Microbial/1679          Bifidobacterium longum biovar Longum
 Microbial/169963        Listeria monocytogenes EGD-e
 Microbial/170187        Streptococcus pneumoniae TIGR4
 Microbial/171101        Streptococcus pneumoniae R6
 Microbial/1717          Corynebacterium diphtheriae
 Microbial/1718          Corynebacterium glutamicum
 Microbial/176279        Staphylococcus epidermidis RP62A
 Microbial/176280        Staphylococcus epidermidis ATCC 12228
 Microbial/176281        Listeria monocytogenes ATCC 19115
 Microbial/1765          Mycobacterium bovis
 Microbial/1769          Mycobacterium leprae
 Microbial/178306        Pyrobaculum aerophilum str. IM2
 Microbial/178328        Salmonella typhimurium TR7095
 Microbial/180454        Anopheles gambiae str. PEST
 Microbial/180835        Agrobacterium tumefaciens str. C58 
                           (U. Washington)
 Microbial/181661        Agrobacterium tumefaciens str. C58
                           (Cereon)
 Microbial/182082        Chlamydophila pneumoniae TW-183
 Microbial/182710        Oceanobacillus iheyensis
 Microbial/183190        Xylella fastidiosa Temecula1
 Microbial/184922        Giardia lamblia ATCC 50803
 Microbial/186103        Streptococcus pyogenes MGAS8232
 Microbial/186497        Pyrococcus furiosus DSM 3638
 Microbial/187410        Yersinia pestis KIM
 Microbial/187420        Methanothermobacter thermautotrophicus
                           str. Delta H
 Microbial/188           Magnetospirillum magnetotacticum
 Microbial/188937        Methanosarcina acetivorans C2A
 Microbial/189423        Streptococcus pneumoniae 670
 Microbial/189518        Leptospira interrogans serovar lai 
                           str. 56601
 Microbial/190192        Methanopyrus kandleri AV19
 Microbial/190304        Fusobacterium nucleatum subsp. 
                           nucleatum ATCC 25586
 Microbial/190485        Xanthomonas campestris pv. Campestris
                           str. ATCC 33913
 Microbial/190486        Xanthomonas axonopodis pv. citri str. 306
 Microbial/190650        Caulobacter crescentus CB15
 Microbial/191218        Bacillus anthracis str. A2012
 Microbial/192222        Campylobacter jejuni subsp. jejuni 
                           NCTC 11168
 Microbial/192952        Methanosarcina mazei Goe1
 Microbial/193567        Streptococcus pyogenes SSI-1
 Microbial/194439        Chlorobium tepidum TLS
 Microbial/195099        Campylobacter jejuni RM1221
 Microbial/195102        Clostridium perfringens str. 13
 Microbial/195103        Clostridium perfringens ATCC 13124
 Microbial/196164        Corynebacterium efficiens YS-314
 Microbial/196600        Vibrio vulnificus YJ016
 Microbial/196620        Staphylococcus aureus subsp. aureus MW2
 Microbial/196627        Corynebacterium glutamicum ATCC 13032
 Microbial/197           Campylobacter jejuni
 Microbial/197221        Thermosynechococcus elongatus BP-1
 Microbial/198094        Bacillus anthracis str. Ames
 Microbial/198214        Shigella flexneri 2a str. 301
 Microbial/198215        Shigella flexneri 2a str. 2457T
 Microbial/198466        Streptococcus pyogenes MGAS315
 Microbial/198628        Erwinia chrysanthemi str. 3937
 Microbial/198804        Buchnera aphidicola str. Sg 
                           (Schizaphis graminum)
 Microbial/199310        Escherichia coli CFT073
 Microbial/2021          Thermobifida fusca
 Microbial/203119        Clostridium thermocellum ATCC 27405
 Microbial/203120        Leuconostoc mesenteroides subsp. 
                           mesenteroides ATCC 8293
 Microbial/203122        Microbulbifer degradans 2-40
 Microbial/203123        Oenococcus oeni MCW
 Microbial/203124        Trichodesmium erythraeum IMS101
 Microbial/203267        Tropheryma whipplei str. Twist
 Microbial/203275        Tannerella forsythensis ATCC 43037
 Microbial/203907        Candidatus Blochmannia floridanus
 Microbial/204722        Brucella suis 1330
 Microbial/205913        Bifidobacterium longum DJO10A
 Microbial/205914        Haemophilus somnus 129PT
 Microbial/205918        Pseudomonas syringae pv. syringae B728a
 Microbial/205919        Bacillus anthracis str. KrugerB
 Microbial/205920        Ehrlichia chaffeensis str. Arkansas
 Microbial/205921        Streptococcus agalactiae A909
 Microbial/205922        Pseudomonas fluorescens PfO-1
 Microbial/206672        Bifidobacterium longum NCC2705
 Microbial/207559        Desulfovibrio desulfuricans G20
 Microbial/208435        Streptococcus agalactiae 2603V/R
 Microbial/208963        Pseudomonas aeruginosa UCBPP-PA14
 Microbial/208964        Pseudomonas aeruginosa PAO1
 Microbial/209261        Salmonella enterica subsp. 
                           enterica serovar Typhi Ty2
 Microbial/2097          Mycoplasma genitalium
 Microbial/209841        Actinobacillus pleuropneumoniae serovar 7
 Microbial/209882        Fusobacterium nucleatum subsp. 
                           vincentii ATCC 49256
 Microbial/210007        Streptococcus mutans UA159
 Microbial/2104          Mycoplasma pneumoniae
 Microbial/2107          Mycoplasma pulmonis
 Microbial/2111          Mycoplasma arthritidis
 Microbial/211110        Streptococcus agalactiae NEM316
 Microbial/211586        Shewanella oneidensis MR-1
 Microbial/212042        Anaplasma phagocytophilum HZ
 Microbial/212045        Bacillus anthracis str. WesternNA
 Microbial/212717        Clostridium tetani E88
 Microbial/2130          Ureaplasma urealyticum
 Microbial/214092        Yersinia pestis CO92
 Microbial/214684        Cryptococcus neoformans var. 
                           neoformans JEC21
 Microbial/214688        Gemmata obscuriglobus UQM 2246
 Microbial/216591        Burkholderia cenocepacia J2315
 Microbial/216592        Escherichia coli 042
 Microbial/216593        Escherichia coli E2348
 Microbial/216594        Mycobacterium marinum M
 Microbial/216595        Pseudomonas fluorescens SBW25
 Microbial/216596        Rhizobium leguminosarum bv. viciae 3841
 Microbial/216597        Salmonella typhimurium SL1344
 Microbial/216598        Shigella dysenteriae M131649
 Microbial/216599        Shigella sonnei 53G
 Microbial/216600        Streptococcus pneumoniae 23F
 Microbial/216895        Vibrio vulnificus CMCP6
 Microbial/218491        Erwinia carotovora subsp. 
                           atroseptica SCRI1043
 Microbial/218493        Salmonella bongori 12149
 Microbial/218494        Streptococcus suis P1/7
 Microbial/218495        Streptococcus uberis 0140J
 Microbial/218496        Tropheryma whipplei TW08/27
 Microbial/218497        Chlamydophila abortus S26/3
 Microbial/2190          Methanocaldococcus jannaschii
 Microbial/220664        Pseudomonas fluorescens Pf-5
 Microbial/220668        Lactobacillus plantarum WCFS1
 Microbial/2208          Methanosarcina barkeri
 Microbial/221109        Oceanobacillus iheyensis HTE831
 Microbial/222523        Bacillus cereus ATCC 10987
 Microbial/222891        Neorickettsia sennetsu str. Miyayama
 Microbial/222929        Coccidioides posadasii C735
 Microbial/223283        Pseudomonas syringae pv. 
                           tomato str. DC3000
 Microbial/2234          Archaeoglobus fulgidus
 Microbial/223926        Vibrio parahaemolyticus RIMD 2210633
 Microbial/224308        Bacillus subtilis subsp. 
                           subtilis str. 168
 Microbial/224324        Aquifex aeolicus VF5
 Microbial/224325        Archaeoglobus fulgidus DSM 4304
 Microbial/224326        Borrelia burgdorferi B31
 Microbial/224911        Bradyrhizobium japonicum USDA 110
 Microbial/224914        Brucella melitensis 16M
 Microbial/224915        Buchnera aphidicola str. Bp 
                           (Baizongia pistaciae)
 Microbial/226125        Saccharomyces paradoxus NRRL Y-17217
 Microbial/226126        Saccharomyces mikatae IFO 1815
 Microbial/226127        Saccharomyces bayanus MCYC 623
 Microbial/226185        Enterococcus faecalis V583
 Microbial/226186        Bacteroides thetaiotaomicron VPI-5482
 Microbial/226230        Saccharomyces kudriavzevii IFO 1802
 Microbial/226231        Saccharomyces bayanus 623-6C
 Microbial/226301        Saccharomyces castellii NRRL Y-12630
 Microbial/226302        Saccharomyces kluyveri NRRL Y-12651
 Microbial/226900        Bacillus cereus ATCC 14579
 Microbial/227321        Aspergillus nidulans FGSC A4
 Microbial/227377        Coxiella burnetii RSA 493
 Microbial/227882        Streptomyces avermitilis MA-4680
 Microbial/227941        Chlamydophila caviae GPIC
 Microbial/228399        Actinobacillus pleuropneumoniae serovar 1
                           str. 4074
 Microbial/228400        Haemophilus somnus 2336
 Microbial/228405        Hyphomonas neptunium ATCC 15444
 Microbial/228410        Nitrosomonas europaea ATCC 19718
 Microbial/2287          Sulfolobus solfataricus
 Microbial/228908        Nanoarchaeum equitans Kin4-M
 Microbial/229533        Gibberella zeae PH-1
 Microbial/2303          Thermoplasma acidophilum
 Microbial/233150        Mycoplasma gallisepticum R
 Microbial/233412        Haemophilus ducreyi 35000HP
 Microbial/233413        Mycobacterium bovis AF2122/97
 Microbial/2336          Thermotoga maritima
 Microbial/234826        Anaplasma marginale str. St. Maries
 Microbial/235279        Helicobacter hepaticus ATCC 51449
 Microbial/235443        Cryptococcus neoformans var. grubii H99
 Microbial/237631        Ustilago maydis 521
 Microbial/238030        Pasteuria nishizawae str. North American
 Microbial/24            Shewanella putrefaciens
 Microbial/240176        Coprinopsis cinerea okayama7#130
 Microbial/242507        Magnaporthe grisea 70-15
 Microbial/242619        Porphyromonas gingivalis W83
 Microbial/243090        Pirellula sp. 1
 Microbial/243159        Acidithiobacillus ferrooxidans ATCC 23270
 Microbial/243160        Burkholderia mallei ATCC 23344
 Microbial/243164        Dehalococcoides ethenogenes 195
 Microbial/243230        Deinococcus radiodurans R1
 Microbial/243231        Geobacter sulfurreducens PCA
 Microbial/243233        Methylococcus capsulatus str. Bath
 Microbial/243243        Mycobacterium avium 104
 Microbial/243265        Photorhabdus luminescens subsp. 
                           laumondii TTO1
 Microbial/243272        Mycoplasma arthritidis 158L3-1
 Microbial/243275        Treponema denticola ATCC 35405
 Microbial/243276        Treponema pallidum subsp. 
                           pallidum str. Nichols
 Microbial/243277        Vibrio cholerae O1 biovar eltor 
                           str. N16961
 Microbial/243365        Chromobacterium violaceum ATCC 12472
 Microbial/246194        Carboxydothermus hydrogenoformans Z-2901
 Microbial/246195        Dichelobacter nodosus VCS1703A
 Microbial/246196        Mycobacterium smegmatis MC2
 Microbial/246197        Myxococcus xanthus DK 1622
 Microbial/246198        Prevotella intermedia 17
 Microbial/246199        Ruminococcus albus 8
 Microbial/246200        Silicibacter pomeroyi DSS-3
 Microbial/246201        Streptococcus mitis NCTC 12261
 Microbial/246202        Streptococcus sobrinus 6715
 Microbial/254945        Ehrlichia ruminantium str. Welgevonden
 Microbial/257310        Bordetella bronchiseptica RB50
 Microbial/257313        Bordetella pertussis Tohama I
 Microbial/257314        Lactobacillus johnsonii NCC 533
 Microbial/258594        Rhodopseudomonas palustris CGA009
 Microbial/259536        Psychrobacter sp. 273-4
 Microbial/262316        Mycobacterium avium subsp.
                           paratuberculosis str. k10
 Microbial/262727        Haemophilus influenzae R2846
 Microbial/262728        Haemophilus influenzae R2866
 Microbial/28227         Mycoplasma penetrans
 Microbial/28232         Geobacter metallireducens
 Microbial/28450         Burkholderia pseudomallei
 Microbial/287           Pseudomonas aeruginosa
 Microbial/29292         Pyrococcus abyssi
 Microbial/29390         Streptococcus gordonii str. Challis
 Microbial/294           Pseudomonas fluorescens
 Microbial/29459         Brucella melitensis
 Microbial/29461         Brucella melitensis biovar Suis
 Microbial/305           Ralstonia solanacearum
 Microbial/31033         Takifugu rubripes
 Microbial/32049         Synechococcus sp. PCC 7002
 Microbial/33072         Gloeobacter violaceus
 Microbial/33178         Aspergillus terreus
 Microbial/34054         Yersinia enterocolitica (type 0:8)
 Microbial/354           Azotobacter vinelandii
 Microbial/35793         Rickettsia sibirica
 Microbial/36329         Plasmodium falciparum 3D7
 Microbial/36826         Clostridium botulinum A
 Microbial/36870         Wigglesworthia glossinidia 
                           endosymbiont of Glossina brevipalpis
 Microbial/3702          Arabidopsis thaliana
 Microbial/375           Bradyrhizobium japonicum
 Microbial/38081         Pneumocystis carinii
 Microbial/381           Mesorhizobium loti
 Microbial/382           Sinorhizobium meliloti
 Microbial/40325         Actinobacillus pleuropneumoniae serovar 1
 Microbial/42068         Pneumocystis jirovecii
 Microbial/42897         Shigella flexneri 2a
 Microbial/44101         Mycoplasma mycoides subsp. mycoides SC
 Microbial/44294         Actinobacillus pleuropneumoniae serovar 5
 Microbial/446           Legionella pneumophila
 Microbial/47834         Spiroplasma kunkelii
 Microbial/485           Neisseria gonorrhoeae
 Microbial/48935         Novosphingobium aromaticivorans
 Microbial/4896          Schizosaccharomyces pombe
 Microbial/4932          Saccharomyces cerevisiae
 Microbial/49338         Desulfitobacterium hafniense
 Microbial/50339         Thermoplasma volcanium
 Microbial/5067          Aspergillus parasiticus
 Microbial/5085          Aspergillus fumigatus
 Microbial/5141          Neurospora crassa
 Microbial/51511         Ciona savignyi
 Microbial/518           Bordetella bronchiseptica
 Microbial/519           Bordetella parapertussis
 Microbial/520           Bordetella pertussis
 Microbial/5207          Filobasidiella neoformans
 Microbial/5306          Phanerochaete chrysosporium
 Microbial/53953         Pyrococcus horikoshii
 Microbial/54388         Salmonella paratyphi
 Microbial/5476          Candida albicans
 Microbial/5514          Fusarium sporotrichioides
 Microbial/556           Erwinia chrysanthemi
 Microbial/56636         Aeropyrum pernix
 Microbial/5691          Trypanosoma brucei
 Microbial/5693          Trypanosoma cruzi
 Microbial/573           Klebsiella pneumoniae
 Microbial/5759          Entamoeba histolytica
 Microbial/5807          Cryptosporidium parvum
 Microbial/5811          Toxoplasma gondii
 Microbial/5823          Plasmodium berghei strain ANKA
 Microbial/5825          Plasmodium chabaudi
 Microbial/5855          Plasmodium vivax
 Microbial/5861          Plasmodium yoelii
 Microbial/5865          Babesia bovis
 Microbial/5874          Theileria annulata
 Microbial/5875          Theileria parva
 Microbial/592           Salmonella enteritidis
 Microbial/59374         Fibrobacter succinogenes subsp. 
                           succinogenes S85
 Microbial/59919         Prochlorococcus marinus subsp. 
                           pastoris str. CCMP1986
 Microbial/6035          Encephalitozoon cuniculi
 Microbial/6182          Schistosoma japonicum
 Microbial/6183          Schistosoma mansoni
 Microbial/6238          Caenorhabditis briggsae
 Microbial/6239          Caenorhabditis elegans
 Microbial/6279          Brugia malayi
 Microbial/632           Yersinia pestis
 Microbial/63363         Aquifex aeolicus
 Microbial/63737         Nostoc punctiforme
 Microbial/64091         Halobacterium sp. NRC-1
 Microbial/666           Vibrio cholerae
 Microbial/672           Vibrio vulnificus
 Microbial/6943          Amblyomma americanum
 Microbial/714           Actinobacillus actinomycetemcomitans
 Microbial/71421         Haemophilus influenzae Rd KW20
 Microbial/7159          Aedes aegypti
 Microbial/7160          Aedes albopictus
 Microbial/7162          Ochlerotatus triseriatus
 Microbial/7227          Drosophila melanogaster
 Microbial/7237          Drosophila pseudoobscura
 Microbial/730           Haemophilus ducreyi
 Microbial/73239         Plasmodium yoelii yoelii
 Microbial/74547         Prochlorococcus marinus str. MIT 9313
 Microbial/7460          Apis mellifera
 Microbial/747           Pasteurella multocida
 Microbial/7719          Ciona intestinalis
 Microbial/781           Rickettsia conorii
 Microbial/782           Rickettsia prowazekii
 Microbial/783           Rickettsia rickettsii
 Microbial/813           Chlamydia trachomatis
 Microbial/817           Bacteroides fragilis
 Microbial/83331         Mycobacterium tuberculosis CDC1551
 Microbial/83332         Mycobacterium tuberculosis H37Rv
 Microbial/83333         Escherichia coli K12
 Microbial/83334         Escherichia coli O157:H7
 Microbial/83554         Chlamydophila psittaci
 Microbial/83560         Chlamydia muridarum
 Microbial/837           Porphyromonas gingivalis
 Microbial/844           Wolinella succinogenes
 Microbial/84588         Synechococcus sp. WH 8102
 Microbial/85569         Salmonella typhimurium DT104
 Microbial/85962         Helicobacter pylori 26695
 Microbial/85963         Helicobacter pylori J99
 Microbial/86665         Bacillus halodurans
 Microbial/882           Desulfovibrio vulgaris subsp. 
                           vulgaris str. Hildenborough
 Microbial/90370         Salmonella enterica subsp. 
                           enterica serovar Typhi
 Microbial/915           Nitrosomonas europaea
 Microbial/93061         Staphylococcus aureus subsp. 
                           aureus NCTC 8325
 Microbial/93062         Staphylococcus aureus subsp. aureus COL
 Microbial/959           Bdellovibrio bacteriovorus
 Microbial/97393         Ferroplasma acidarmanus
 Microbial/98360         Salmonella enterica subsp. 
                           enterica serovar Dublin
 Microbial/985           Cytophaga hutchinsonii
 Microbial/99287         Salmonella typhimurium LT2
+------------------+-------------------------------------------+


J.Trace Databases
+-----------------------------------------------------+------------+
|NAME                                                 | CONTENT*   |
+-----------------------------------------------------+------------+
 Trace/Acidithiobacillus_Ferrooxidans_Atcc_23270_WGS
 Trace/Aedes_Aegypti_OTHER
 Trace/Anopheles_Gambiae_OTHER
 Trace/Anopheles_Gambiae_WGS
 Trace/Apis_Mellifera_OTHER
 Trace/Apis_Mellifera_WGS
 Trace/Archaeoglobus_Fulgidus_Dsm_4304_WGS
 Trace/Artibeus_Jamaicensis_OTHER
 Trace/Aspergillus_Flavus_OTHER
 Trace/Aspergillus_Fumigatus_WGS
 Trace/Aspergillus_Nidulans_WGS
 Trace/Aspergillus_Parasiticus_OTHER
 Trace/Aspergillus_Terreus_WGS
 Trace/Atelerix_Albiventris_OTHER
 Trace/Bacillus_Anthracis_Str__A0039_OTHER
 Trace/Bacillus_Anthracis_Str__A01055_OTHER
 Trace/Bacillus_Anthracis_Str__A2012_WGS
 Trace/Bacillus_Anthracis_Str__Ames_0581_OTHER
 Trace/Bacillus_Anthracis_Str__Ames_WGS
 Trace/Bacillus_Anthracis_Str__France_OTHER
 Trace/Bacillus_Anthracis_Str__Kruger_B_OTHER
 Trace/Bacillus_Anthracis_Str__Vollum_OTHER
 Trace/Bacillus_Anthracis_Str__Western_North_America_OTHER
 Trace/Bacillus_Cereus_Atcc_10987_WGS
 Trace/Borrelia_Burgdorferi_B31_WGS
 Trace/Bos_Taurus_EST
 Trace/Bos_Taurus_OTHER
 Trace/Botryllus_Schlosseri_OTHER
 Trace/Branchiostoma_Floridae_OTHER
 Trace/Brassica_Oleracea_WGS
 Trace/Brucella_Suis_1330_OTHER
 Trace/Burkholderia_Mallei_Atcc_23344_WGS
 Trace/Caenorhabditis_Briggsae_WGS
 Trace/Callicebus_Moloch_OTHER
 Trace/Callithrix_Jacchus_OTHER
 Trace/Campylobacter_Jejuni_Rm1221_WGS
 Trace/Canis_Familiaris_OTHER
 Trace/Canis_Familiaris_WGS
 Trace/Carollia_Perspicillata_OTHER
 Trace/Cercopithecus_Aethiops_OTHER
 Trace/Chlamydia_Muridarum_Nigg_WGS
 Trace/Chlamydomonas_Reinhardtii_OTHER
 Trace/Chlamydomonas_Reinhardtii_WGS
 Trace/Chlamydophila_Caviae_Gpic_WGS
 Trace/Chlamydophila_Pneumoniae_Ar39_WGS
 Trace/Chlorobium_Tepidum_Tls_WGS
 Trace/Cicer_Arietinum_OTHER
 Trace/Ciona_Intestinalis_OTHER
 Trace/Ciona_Intestinalis_WGS
 Trace/Ciona_Savignyi_WGS.00
 Trace/Ciona_Savignyi_WGS.01
 Trace/Clostridium_Perfringens_Atcc_13124_WGS
 Trace/Coccidioides_Immitis_WGS
 Trace/Colwellia_Psychrerythraea_34h_WGS
 Trace/Coprinus_Cinereus_WGS
 Trace/Coxiella_Burnetii_Rsa_493_WGS
 Trace/Cryptococcus_Neoformans_Serotype_A_WGS
 Trace/Danio_Rerio_EST
 Trace/Danio_Rerio_OTHER
 Trace/Dasypus_Novemcinctus_OTHER
 Trace/Dehalococcoides_Ethenogenes_195_WGS
 Trace/Didelphis_Virginiana_OTHER
 Trace/Drosophila_Melanogaster_OTHER
 Trace/Drosophila_Pseudoobscura_EST
 Trace/Drosophila_Pseudoobscura_OTHER
 Trace/Drosophila_Pseudoobscura_WGS
 Trace/Drosophila_Simulans_WGS
 Trace/Drosophila_Virilis_OTHER
 Trace/Drosophila_Yakuba_WGS
 Trace/Ehrlichia_Chaffeensis_Str__Arkansas_WGS
 Trace/Emericella_Nidulans_OTHER
 Trace/Enterococcus_Faecalis_V583_WGS
 Trace/Equus_Caballus_OTHER
 Trace/Euprymna_Scolopes_EST
 Trace/Felis_Catus_OTHER
 Trace/Gallus_Gallus_OTHER
 Trace/Gallus_Gallus_WGS
 Trace/Geobacter_Sulfurreducens_Pca_WGS
 Trace/Gibberella_Zeae_WGS
 Trace/Glycine_Max_EST
 Trace/Gopherus_Agassizii_OTHER
 Trace/Gorilla_Gorilla_OTHER
 Trace/Homo_Sapiens_EST
 Trace/Homo_Sapiens_OTHER
 Trace/Homo_Sapiens_WGS
 Trace/Lemur_Catta_OTHER
 Trace/Macaca_Mulatta_EST
 Trace/Macaca_Mulatta_OTHER
 Trace/Macaca_Mulatta_WGS
 Trace/Macropus_Eugenii_OTHER
 Trace/Magnaportha_Grisea_WGS
 Trace/Medicago_Truncatula_OTHER
 Trace/Medicago_Truncatula_WGS
 Trace/Mesoplasma_Florum_WGS
 Trace/Methanosarcina_Acetivorans_WGS
 Trace/Methylococcus_Capsulatus_Str__Bath_WGS
 Trace/Microcebus_Murinus_OTHER
 Trace/Monodelphis_Domestica_OTHER
 Trace/Monodelphis_Domestica_WGS
 Trace/Muntiacus_Muntjak_Vaginalis_OTHER
 Trace/Mus_Musculus_EST
 Trace/Mus_Musculus_OTHER
 Trace/Mus_Musculus_WGS
 Trace/Mycobacterium_Tuberculosis_Cdc1551_WGS
 Trace/Mycoplasma_Alligatoris_WGS
 Trace/Mycoplasma_Arthritidis_158l3_1_WGS
 Trace/Myxococcus_Xanthus_OTHER
 Trace/Neorickettsia_Sennetsu_Str__Miyayama_WGS
 Trace/Ornithorhynchus_Anatinus_OTHER
 Trace/Oryctolagus_Cuniculus_OTHER
 Trace/Oryza_Sativa__indica_Cultivar_group__OTHER
 Trace/Oryza_Sativa__japonica_Cultivar_group__OTHER
 Trace/Otolemur_Garnettii_OTHER
 Trace/Ovis_Aries_OTHER
 Trace/Pan_Troglodytes_OTHER
 Trace/Pan_Troglodytes_WGS
 Trace/Papio_Anubis_OTHER
 Trace/Papio_Cynocephalus_OTHER
 Trace/Paramecium_Tetraurelia_OTHER
 Trace/Phanerochaete_Chrysosporium_WGS
 Trace/Phytophthora_Infestans_OTHER
 Trace/Phytophthora_Infestans_WGS
 Trace/Phytophthora_Ramorum_WGS
 Trace/Phytophthora_Sojae_WGS
 Trace/Plasmodium_Falciparum_OTHER
 Trace/Podospora_Anserina_OTHER
 Trace/Polypterus_Bichir_OTHER
 Trace/Pongo_Pygmaeus_OTHER
 Trace/Populus_Balsamifera_WGS
 Trace/Porphyromonas_Gingivalis_W83_WGS
 Trace/Pseudomonas_Aeruginosa_OTHER
 Trace/Pseudomonas_Aeruginosa_WGS
 Trace/Pseudomonas_Putida_Kt2440_WGS
 Trace/Pseudomonas_Syringae_Pv__Tomato_Str__Dc3000_WGS
 Trace/Rattus_Norvegicus_EST
 Trace/Rattus_Norvegicus_OTHER
 Trace/Rattus_Norvegicus_WGS
 Trace/Rhizopus_Oryzae_WGS
 Trace/Saccharomyces_Bayanus_WGS
 Trace/Saccharomyces_Mikatae_WGS
 Trace/Saccharomyces_Paradoxus_WGS
 Trace/Saccoglossus_Kowalevskii_WGS
 Trace/Saimiri_Boliviensis_Boliviensis_OTHER
 Trace/Schistosoma_Mansoni_EST
 Trace/Schistosoma_Mansoni_WGS
 Trace/Schmidtea_Mediterranea_WGS
 Trace/Shewanella_Oneidensis_Mr_1_OTHER
 Trace/Silurana_Tropicalis_EST
 Trace/Silurana_Tropicalis_OTHER
 Trace/Silurana_Tropicalis_WGS
 Trace/Sminthopsis_Macroura_OTHER
 Trace/Staphylococcus_Aureus_Subsp__Aureus_Col_OTHER
 Trace/Staphylococcus_Aureus_Subsp__Aureus_Col_WGS
 Trace/Staphylococcus_Epidermidis_Rp62a_OTHER
 Trace/Streptococcus_Agalactiae_2603v_r_WGS
 Trace/Streptococcus_Pneumoniae_Tigr4_WGS
 Trace/Streptococcus_Pyogenes_M1_Gas_OTHER
 Trace/Strongylocentrotus_Purpuratus_OTHER
 Trace/Strongylocentrotus_Purpuratus_WGS
 Trace/Sus_Scrofa_OTHER
 Trace/Takifugu_Rubripes_OTHER
 Trace/Takifugu_Rubripes_WGS
 Trace/Tetrahymena_Thermophila_OTHER
 Trace/Tetraodon_Nigroviridis_OTHER
 Trace/Tetraodon_Nigroviridis_WGS
 Trace/Thalassiosira_Pseudonana_WGS
 Trace/Treponema_Denticola_Atcc_35405_WGS
 Trace/Triphysaria_Versicolor_EST
 Trace/Trypanosoma_Brucei_OTHER
 Trace/Ustilago_Hordei_OTHER
 Trace/Ustilago_Maydis_WGS
 Trace/Vibrio_Cholerae_O1_Biovar_Eltor_Str__N16961_WGS
 Trace/Wolbachia_Endosymbiont_Of_Drosophila_Melanogaster_OTHER
 Trace/Xenopus_Laevis_EST
 Trace/Zea_Mays_EST
 Trace/Zea_Mays_OTHER
 Trace/Zea_Mays_WGS
+-------------------------------------------------------------+------------+
* The content of each dataset is self-explanatory since they are named using
  "/Trace/organism_name_DataType" convention, where the DataType can be EST, 
  OTHERS, or WGS.


K. Technical Support

Questions and comments on this document and blast in general should be 
addressed to:
         blast-help@ncbi.nlm.nih.gov

Quesiton and comments on NCBI resources other than BLAST, should be 
addressed to:
         info@ncbi.nlm.nih.gov
</pre>
  </body>
</html>

