<HTML>
<HEAD>
<TITLE>
	Explanations for terms (META PredictProtein)
</TITLE>
</HEAD>

<BODY bgcolor=#FFFFFF>
<H1>Brief explanation of META PredictProtein</H1>

<!-- -------------------------------------------------------------------------------- -->
<!-- beg toc  -->
<DIV style="padding: 1em; margin: 1em 0; border: thick inset; background: silver"> 
<P>
<STRONG>Contents</STRONG>
<P>
<UL>
<LI><A HREF="#user">User information</A>
<LI><A HREF="#list">List of all available services</A>
<LI><A HREF="#services">Services available</A>
	<OL>
	<LI><A HREF="#services_var">      Various analyses</A>
	<LI><A HREF="#services_sec">      Secondary structure prediction</A>
	<LI><A HREF="#services_htm">      Membrane regions</A>
	<LI><A HREF="#services_threading">Threading (detecting remote homologues in PDB)</A>
	<LI><A HREF="#services_homo">     Homology modelling</A>
	</OL>
<LI><A HREF="#formats">Formats for submitting sequence or alignment</A>
</UL>
</DIV>
<!-- end toc  -->
<!-- -------------------------------------------------------------------------------- -->

<P><BR><HR><BR><P>


<!-- -------------------------------------------------------------------------------- -->
<!-- beg user  -->

<TABLE WIDTH="100%">
<TR><TD VALIGN=TOP ALIGN=CENTER BGCOLOR="#000080">
	<STRONG><FONT COLOR="#FFFFFF" SIZE="+3">
	<A NAME="user">User information</A>
	</FONT></STRONG>
	</TD></TR>
</TABLE>

<OL>
<LI><A NAME="PX_usr_email">Your email</A>
		<I>Example: rost@embl-heidelberg.de</I>
	<BR>Your entire (and entirely correct) email address (e.g. rost@embl-heidelberg.de).
	<BR>Note: typos will result in that we shall not be ablet to return the results.

<!-- 
<LI><A NAME="PX_usr_password">Password</A>
		<I>Example: </I> (i.e. leave field empty!)
	<BR>Using PredictProtein is free for academical users.  Only companies have to fill in their password, here!
 -->

<LI><A NAME="PX_prot_name">One-line name of protein</A>
		<I>Example: Cytochrome C oxidase</I>

<LI><A NAME="PX_prot_seq">Paste, or type your sequence </A>
	<BR>Example:
        <PRE><I>	MSAQISDSIEEKRGFFTRWFMSTNHKDIGVLYLFTAGLAGLISVTLTVYMRMELQHPGVQ
	YMCLEGMRLVADAAAECTPNAHL</I></PRE>
	Please use only one-letter code amino acids. In particular, avoid numbers or '*', or '.'.
	<BR>
	For other possible input formats <A HREF="#PX_formats">click here</A>!

<LI><A NAME="PX_submit">SUBMIT or CLEAR</A>
	<BR>Click on the button <STRONG>SUBMIT</STRONG> to request a prediction
	<BR>Click on the button <STRONG>CLEAR </STRONG> to clear all data you filled in (e.g. to restart, or to send a new request).

</OL>
<!-- end user  -->
<!-- -------------------------------------------------------------------------------- -->

<P><BR><HR><BR><P>

<!-- -------------------------------------------------------------------------------- -->
<!-- beg list  -->
<TABLE WIDTH="100%">
<TR><TD VALIGN=TOP ALIGN=CENTER BGCOLOR="#000080">
	<STRONG><FONT COLOR="#FFFFFF" SIZE="+3">
	<A NAME="list">List of services available</A>
	</FONT></STRONG>
	</TD></TR>
</TABLE>

<TABLE CELLPADDING=2 WIDTH="100%">
<TR><TD VALIGN=TOP BGCOLOR="#0000FF" ALIGN=RIGHT><STRONG><FONT COLOR="#FFFFFF">various</FONT></STRONG></TD><TD> &nbsp;</TD><TD><A HREF="#PX_about_signalp">SignalP</A></TD><TD><A HREF="#PX_about_netoglyc">NetOglyc</A></TD><TD><A HREF="#PX_about_netpico">NetPico</A></TD><TD><A HREF="#PX_about_chlorop">ChloroP</A></TD>
<TR><TD VALIGN=TOP BGCOLOR="#0000FF" ALIGN=RIGHT><STRONG><FONT COLOR="#FFFFFF">homology modelling</FONT></STRONG></TD><TD> &nbsp;</TD><TD><A HREF="#PX_about_swissmodel">SWISS-MODEL</A></TD><TD><A HREF="#PX_about_cphmodels">CPHmodels</A></TD>
<TR><TD VALIGN=TOP BGCOLOR="#0000FF" ALIGN=RIGHT><STRONG><FONT COLOR="#FFFFFF">threading</FONT></STRONG></TD><TD> &nbsp;</TD><TD><A HREF="#PX_about_frsvr">FRSVR</A></TD><TD><A HREF="#PX_about_samt98">SAMt98</A></TD>
<TR><TD VALIGN=TOP BGCOLOR="#0000FF" ALIGN=RIGHT><STRONG><FONT COLOR="#FFFFFF">secondary structure</FONT></STRONG></TD><TD> &nbsp;</TD><TD><A HREF="#PX_about_jpred">JPRED</A></TD>
<TR><TD VALIGN=TOP BGCOLOR="#0000FF" ALIGN=RIGHT><STRONG><FONT COLOR="#FFFFFF">transmembrane helices</FONT></STRONG></TD><TD> &nbsp;</TD><TD><A HREF="#PX_about_tmhmm">TMHMM</A></TD><TD><A HREF="#PX_about_toppred">TopPred</A></TD><TD><A HREF="#PX_about_das">DAS</A></TD>
</TABLE>

<!-- end list  -->
<!-- -------------------------------------------------------------------------------- -->

<P><BR><HR><BR><P

<!-- -------------------------------------------------------------------------------- -->
<!-- beg services  -->

<TABLE WIDTH="100%">
<TR><TD VALIGN=TOP ALIGN=CENTER BGCOLOR="#000080">
	<STRONG><FONT COLOR="#FFFFFF" SIZE="+3">
	<A NAME="services">Information about services available</A>
	</FONT></STRONG>
	</TD></TR>
</TABLE>


<!-- various			-------------------------------------------------- -->
<P><HR><P>
<H3><A NAME="services_var">Various analyses</A></H3>
Various services for sequence analysis.  

<TABLE COLS=2 CELLPADDING=2 WIDTH="100%">
<!-- signalp                       ....................-->
<TR VALIGN=TOP><TD VALIGN=TOP WIDTH="15%"> </TD><TD VALIGN=TOP WIDTH="85%"> &nbsp;</TD>
<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Server</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
<A NAME="PX_about_signalp">
     <STRONG>SignalP</STRONG>
</A>
     </TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Site (URL)</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     <A HREF="http://www.cbs.dtu.dk/services/SignalP/">http://www.cbs.dtu.dk/services/SignalP/</A></TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">About</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        The SignalP World Wide Web server predicts the presence and location of signal peptide cleavage sites in amino acid sequences from different organisms: Gram-positive prokaryotes, Gram-negative prokaryotes, and eukaryotes. The method incorporates a prediction of cleavage sites and a signal peptide/non-signal peptide prediction based on a combination of several artificial neural networks.
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Options</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        If you do not check the organism (gram-positive, gram-negative prokaryotes, or eukaryotes), predictions for all three will be returned.
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Quote</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        <CITE>
   H Nielsen, J Engelbrecht, S Brunak, and G von Heijne: Identification of prokaryotic and eukaryotic signal peptides and prediction of their cleavage sites. Protein Engineering 10, 1-6, 1997
</CITE>
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Authors</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     Henrik Nielsen, Soeren Brunak (both CBS Copenhagen), and Gunnar von Heijne (Univ Stockholm, Sweden)
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Contact</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        Kristoffer Rapacki (<A HREF="mailto:rapacki@cbs.dtu.dk">rapacki@cbs.dtu.dk</A>)
</TD></TR>


<!-- netoglyc                      ....................-->
<TR VALIGN=TOP><TD VALIGN=TOP WIDTH="15%"> </TD><TD VALIGN=TOP WIDTH="85%"> &nbsp;</TD>
<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Server</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
<A NAME="PX_about_netoglyc">
     <STRONG>NetOglyc</STRONG>
</A>
     </TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Site (URL)</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     <A HREF="http://www.cbs.dtu.dk/services/NetOGlyc/">http://www.cbs.dtu.dk/services/NetOGlyc/</A></TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">About</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        The NetOglyc WWW server produces neural network predictions of mucin type GalNAc O-glycosylation sites in mammalian proteins.
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Quote</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        <CITE>
   <OL type="i">
   <LI>JE Hansen, O Lund, N Tolstrup, AA Gooley, KL Williams, and S Brunak: NetOglyc: Prediction of mucin type O-glycosylation sites based on sequence context and surface accessibility. Glycoconjugate Journal, 15, 115-130, 1998</LI>
   <LI>JE Hansen, O Lund, K Rapacki, and S Brunak: O-glycbase version 2.0 - A revised database of O-glycosylated proteins. Nucleic Acids Research, 25, 278-282, 1997</LI>
   <LI>JE Hansen, O Lund, K Rapacki J Engelbrecht, H Bohr, JO Nielsen, J-E S Hansen, and S Brunak: Prediction of O-glycosylation of mammalian proteins: Specificity patterns of UDP-GalNAc:-polypeptide N-acetylgalactosaminyltransferase. Biochemical Journal, 308, 801-813, 1995</LI>
</OL>
</CITE>
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Authors</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     Jan Hansen (CBS, Copenhagen, Denmark)
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Contact</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        Kristoffer Rapacki (<A HREF="mailto:rapacki@cbs.dtu.dk">rapacki@cbs.dtu.dk</A>)
</TD></TR>


<!-- netpico                       ....................-->
<TR VALIGN=TOP><TD VALIGN=TOP WIDTH="15%"> </TD><TD VALIGN=TOP WIDTH="85%"> &nbsp;</TD>
<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Server</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
<A NAME="PX_about_netpico">
     <STRONG>NetPico</STRONG>
</A>
     </TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Site (URL)</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     <A HREF="http://www.cbs.dtu.dk/services/NetPicoRNA/">http://www.cbs.dtu.dk/services/NetPicoRNA/</A></TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">About</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        The NetPicoRNA World Wide Web server produces neural network predictions of cleavage sites of picornaviral proteases.
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Quote</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        <CITE>
   N Blom, J Hansen, D Blaas, and S Brunak: Cleavage site analysis in picornaviral polyproteins: Discovering cellular targets by neural networks. Protein Science, 5, 2203-2216, 1996
</CITE>
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Authors</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     Nikolaj Blom (CBS, Copenhagen, Denmark)
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Contact</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        Kristoffer Rapacki (<A HREF="mailto:rapacki@cbs.dtu.dk">rapacki@cbs.dtu.dk</A>)
</TD></TR>


<!-- chlorop                       ....................-->
<TR VALIGN=TOP><TD VALIGN=TOP WIDTH="15%"> </TD><TD VALIGN=TOP WIDTH="85%"> &nbsp;</TD>
<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Server</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
<A NAME="PX_about_chlorop">
     <STRONG>ChloroP</STRONG>
</A>
     </TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Site (URL)</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     <A HREF="http://www.cbs.dtu.dk/services/ChloroP/">http://www.cbs.dtu.dk/services/ChloroP/</A></TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">About</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        The ChloroP www-server is able to predict two things: <UL><LI>1. cTP or no cTP
<BR>     Whether or not an amino acid sequence contains an N-terminal chloroplast transit peptide, cTP.<LI>2. Cleavage site
<BR>     The probable site for cleavage of the transit peptide (if it was predicted to exist in the first step).</UL>
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Help</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        <UL>
   <LI>Include the N-terminus<BR>
     It is strongly recommended to include the N-terminus of the submitted sequence. The further from the  N-terminal residue the submitted sequence starts, the more difficult and unreliable will the prediction be.</LI>
   <LI>Submit preferably 100-150 residues<BR>
     Submit if possible at least 100 and no more than 150 N-terminal residues. The lower boundary is due to the fact  that the "cTP"/"no cTP" predictor was trained with input sequences of length 100 residues. However, shorter sequences may also be satisfactory predicted (it is more important that the N-terminal part is intact). The  cleavage site prediction is in itself not influenced by sequence length, but restricting the submitted length to approximately 150 residues prevents the prediction of too long cTP's.</LI>
</UL>
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Quote</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        <CITE>
   O Emanuelsson, H Nielsen, and G von Heijne:ChloroP, a neural network-based method for predicting chloroplast transit peptides and their cleavage sites. Protein Science, 8, 978-984, 1999
</CITE>
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Authors</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     Olof Emanuelsson (CBS Copenhagen, Denmark)
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Contact</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        Kristoffer Rapacki (<A HREF="mailto:rapacki@cbs.dtu.dk">rapacki@cbs.dtu.dk</A>)
</TD></TR>


</TABLE>
<P><BR><P>


<!-- secondary str		-------------------------------------------------- -->
<P><HR><P>
<H3><A NAME="services_sec">Secondary structure prediction</A></H3>
Servers returning predictions of secondary structure based on single sequences, or sequence alignments.

<TABLE COLS=2 CELLPADDING=2 WIDTH="100%">
<!-- jpred                         ....................-->
<TR VALIGN=TOP><TD VALIGN=TOP WIDTH="15%"> </TD><TD VALIGN=TOP WIDTH="85%"> &nbsp;</TD>
<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Server</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
<A NAME="PX_about_jpred">
     <STRONG>JPRED</STRONG>
</A>
     </TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Site (URL)</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     <A HREF="http://circinus.ebi.ac.uk:8081/">http://circinus.ebi.ac.uk:8081/</A></TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">About</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        A consensus method for protein secondary structure prediction.
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Quote</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        <CITE>
   J A Cuff, and G J Barton: Evaluation and improvement of multiple sequence methods for protein secondary structure prediction. PROTEINS, 34, 508-519, 1999
</CITE>
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Authors</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     James Cuff, and Geoff Barton (EBI Hinxton, England)
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Contact</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        James Cuff (<A HREF="mailto:james@ebi.ac.uk">james@ebi.ac.uk</A>)
</TD></TR>


</TABLE>
<P><BR><P>


<!-- transmembrane helices	-------------------------------------------------- -->
<P><HR><P>
<H3><A NAME="services_htm">Prediction of transmembrane helices</A></H3>
Servers returning predictions of transmembrane helix location and topology based on single sequences, or sequence alignments. Note: there is no general method for detecting porin-like membrane proteins, yet.

<TABLE COLS=2 CELLPADDING=2 WIDTH="100%">
<!-- tmhmm                         ....................-->
<TR VALIGN=TOP><TD VALIGN=TOP WIDTH="15%"> </TD><TD VALIGN=TOP WIDTH="85%"> &nbsp;</TD>
<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Server</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
<A NAME="PX_about_tmhmm">
     <STRONG>TMHMM</STRONG>
</A>
     </TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Site (URL)</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     <A HREF="http://www.cbs.dtu.dk/services/TMHMM-1.0/">http://www.cbs.dtu.dk/services/TMHMM-1.0/</A></TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">About</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        Hidden Markov Model predicting the location of transmembrane helices and their topology.  (Remark Burkhard Rost: find it to be the most original method for this purpose!)
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Quote</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        <CITE>
   ELL Sonnhammer, G von Heijne, and A Krogh: A hidden Markov model for predicting transmembrane helices in protein sequences. Proc of the Sixth Intern Conf on Intelligent Systems for Molecular Biology (ISMB98), 175-182, 1998
</CITE>
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Authors</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     Anders Krogh (CBS, Copenhagen, Denmark)
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Contact</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        Anders Krogh (<A HREF="mailto:krogh@cbs.dtu.dk ">krogh@cbs.dtu.dk </A>)
</TD></TR>


<!-- toppred                       ....................-->
<TR VALIGN=TOP><TD VALIGN=TOP WIDTH="15%"> </TD><TD VALIGN=TOP WIDTH="85%"> &nbsp;</TD>
<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Server</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
<A NAME="PX_about_toppred">
     <STRONG>TopPred</STRONG>
</A>
     </TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Site (URL)</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     <A HREF="http://www.biokemi.su.se/~server/toppred2/">http://www.biokemi.su.se/~server/toppred2/</A></TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">About</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        Prediction of location and orientation of transmembrane helices through an advanced use of hydrophobicity patterns, and by applying the 'positive-inside' rule.
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Quote</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        <CITE>
   <OL type="i">
   <LI>G von Heijne: Membrane Protein Structure Prediction, Hydrophobicity Analysis and the Positive-inside Rule. J. Molecular Biology, 225, 487-494, 1992</LI>
   <LI>M Cserzo, E Wallin, I Simon, G von Heijne, and A Elofsson: Prediction of transmembrane alpha-helices in prokaryotic membrane proteins: the dense alignment surface method. Protein Engineering, 10, 673-676, 1997</LI>
</OL>
</CITE>
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Authors</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     Erik Wallin, and Gunnar von Heijne (Stockholm Univ, Sweden)
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Contact</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        Erik Wallin (<A HREF="mailto:erikw@biokemi.su.se">erikw@biokemi.su.se</A>)
</TD></TR>


<!-- das                           ....................-->
<TR VALIGN=TOP><TD VALIGN=TOP WIDTH="15%"> </TD><TD VALIGN=TOP WIDTH="85%"> &nbsp;</TD>
<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Server</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
<A NAME="PX_about_das">
     <STRONG>DAS</STRONG>
</A>
     </TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Site (URL)</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     <A HREF="http://www.biokemi.su.se/~server/DAS/">http://www.biokemi.su.se/~server/DAS/</A></TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">About</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        Prediction of location of transmembrane helices through an advanced use of hydrophobicity patterns.
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Quote</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        <CITE>
   M Cserzo, E Wallin, I Simon, G von Heijne, and A Elofsson: Prediction of transmembrane alpha-helices in procariotic membrane proteins: the Dense Alignment Surface method. Protein Engineering, 10, 673-676, 1997
</CITE>
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Authors</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     Miklos Cserzo, Istvan Simon (both Academy of Sciences, Budapest, Hungary), Erik Wallin, Gunnar von Heijne, Arne Elofsson (Stockholm Univ, Sweden)
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Contact</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        Miklos Cserzo (<A HREF="mailto:miklos@pugh.bip.bham.ac.uk">miklos@pugh.bip.bham.ac.uk</A>)
</TD></TR>


</TABLE>
<P><BR><P>


<!-- threading			-------------------------------------------------- -->
<P><HR><P>
<H3><A NAME="services_threading">Threading (detecting remote homologues in PDB)</A></H3>
Threading servers search through databases of proteins of known structures (subsets of PDB), and detect similiarities between proteins which are too weak to be inferred from simple sequence alignment techniques.  (The 'trick' that allows to intrude below the twilight zone of 'simple' similarity detection is using the information contained in the known structures of the proteins compared to your sequence.)
<BR>Note: detection of remote homologues is more of an art than of a solved problem.  Most of the results returned are supposedly wrong!  Thus, you have to gather independent evidence to be able to believe in a particular result.

<TABLE COLS=2 CELLPADDING=2 WIDTH="100%">
<!-- frsvr                         ....................-->
<TR VALIGN=TOP><TD VALIGN=TOP WIDTH="15%"> </TD><TD VALIGN=TOP WIDTH="85%"> &nbsp;</TD>
<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Server</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
<A NAME="PX_about_frsvr">
     <STRONG>FRSVR</STRONG>
</A>
     </TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Site (URL)</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     <A HREF="http://www.doe-mbi.ucla.edu/people/fischer/TEST/getsequence.html">http://www.doe-mbi.ucla.edu/people/fischer/TEST/getsequence.html</A></TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">About</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        Prediction-based threading, also incorporating purely sequence-based database searches.
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Options</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        <UL>
   <LI>Include H3P2: requests to include the results of the H3P2 prediction.</LI>
   <LI>Include PROFILESEARCH: requests to include the results of the PROFILESEARCH program (Smith-Waterman using a profile of the multiple alignment).</LI>
</UL>
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Quote</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        <CITE>
   <OL type="i">
   <LI>D Fischer, and DA Eisenberg: Fold Recognition Using Sequence-Derived Properties. Protein Science, 5, 947-955, 1996</LI>
   <LI>A Elofsson, D Fischer, DW Rice, S LeGrand, and DA Eisenberg: Study of Combined Structure-sequence Profiles. Folding and Design, 1, 451-461, 1996</LI>
</OL>
</CITE>
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Authors</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     Daniel Fischer (Ben Gurion Univ of the Negev, Israel)
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Contact</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        Daniel Fischer (<A HREF="mailto:dfischer@cs.bgu.ac.il ">dfischer@cs.bgu.ac.il </A>)
</TD></TR>


<!-- samt98                        ....................-->
<TR VALIGN=TOP><TD VALIGN=TOP WIDTH="15%"> </TD><TD VALIGN=TOP WIDTH="85%"> &nbsp;</TD>
<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Server</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
<A NAME="PX_about_samt98">
     <STRONG>SAMt98</STRONG>
</A>
     </TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Site (URL)</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     <A HREF="http://www.cse.ucsc.edu/research/compbio/HMM-apps/model-library-search.html">http://www.cse.ucsc.edu/research/compbio/HMM-apps/model-library-search.html</A></TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">About</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        A new hidden Markov model method (SAM-T98) for finding remote homologs of protein sequences is described and evaluated. The method begins with a single target sequence and iteratively builds a hidden Markov model (HMM) from the sequence and homologs found using the HMM for database search. SAM-T98 is also used to construct model libraries automatically from sequences in structural databases.
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Quote</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        <CITE>
   Kevin Karplus, Christian Barrett, and Richard Hughey:Hidden Markov Models for Detecting Remote Protein HomologiesBioinformatics, 14, 846-856, 1998
</CITE>
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Authors</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     Kevin Karplus, Christian Barrett, and Richard Hughey (UCSD, Santa Cruz, USA)
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Contact</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        SAM-INFO (<A HREF="mailto:sam-info@cse.ucsc.edu">sam-info@cse.ucsc.edu</A>)
</TD></TR>


</TABLE>
<P><BR><P>


<!-- homology model		-------------------------------------------------- -->
<P><HR><P>
<H3><A NAME="services_homo">Homology modelling</A></H3>
Homology modelling predicts the three-dimensional structure for your sequence based on the similarity of that sequence to a protein of experimentally determined structure. Since the assumption of the method is that the backbone of your protein and the similar one of known structure (referred to as 'the template') are identical, the sequence similarity between the two has to be significant!  <BR>
Note: if there is no similar structure in PDB, 3D structure can NOT (repeat NOT) be predicted by ANY method at this moment!

<TABLE COLS=2 CELLPADDING=2 WIDTH="100%">
<!-- swissmodel                    ....................-->
<TR VALIGN=TOP><TD VALIGN=TOP WIDTH="15%"> </TD><TD VALIGN=TOP WIDTH="85%"> &nbsp;</TD>
<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Server</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
<A NAME="PX_about_swissmodel">
     <STRONG>SWISS-MODEL</STRONG>
</A>
     </TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Site (URL)</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     <A HREF="http://www.expasy.ch/swissmod/SWISS-MODEL.html">http://www.expasy.ch/swissmod/SWISS-MODEL.html</A></TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">About</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        SWISS-MODEL is an Automated Protein Modelling Server running at the GlaxoWellcome Experimental Research in Geneva, Switzerland (<A HREF="http://www.expasy.ch/swissmod/SWISS-MODEL.html">click here</A> for more details on how the method works).
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Quote</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        <CITE>
   <OL type="i">
   <LI>M C Peitsch: Protein Modelling by E-mail.Bio/Technology, 13, 658-660, 1995.</LI>
   <LI>M C Peitsch: ProMod and Swiss-Model: Internet-based tools for automatedcomparative protein modelling.Biochem Soc Trans, 24, 274-279, 1996.</LI>
   <LI>N Guex, and M C Peitsch:SWISS-MODEL and the Swiss-PdbViewer:An environment for comparative protein modelling.Electrophoresis, 18, 2714-2723, 1997.</LI>
</OL>
</CITE>
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Authors</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     Manuel Peitsch, , Torsten Schwede, and Nicolas Guex (Glaxo, Geneva, Switzerland)
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Contact</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        Nicolas Guex (<A HREF="mailto:ng45767@GlaxoWellcome.co.uk">ng45767@GlaxoWellcome.co.uk</A>)
</TD></TR>


<!-- cphmodels                     ....................-->
<TR VALIGN=TOP><TD VALIGN=TOP WIDTH="15%"> </TD><TD VALIGN=TOP WIDTH="85%"> &nbsp;</TD>
<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Server</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
<A NAME="PX_about_cphmodels">
     <STRONG>CPHmodels</STRONG>
</A>
     </TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Site (URL)</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     <A HREF="http://www.cbs.dtu.dk/services/CPHmodels/">http://www.cbs.dtu.dk/services/CPHmodels/</A></TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">About</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        CPHmodels is a collection of methods and databases developed to predict protein structures.  It currently consists of the following tools: Sowhat: A neural network based method to predict contacts between C-alpha atoms from the amino acid sequence. RedHom: A tool to find a subset with low sequence similarity in a database.  Databases: Subsets of the Brookhaven Protein Data Bank (PDB) database with low sequence similarity produced using the RedHom tool.
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Quote</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        <CITE>
   O Lund, K Frimand, J Gorodkin, H Bohr, J Bohr, J Hansen, and S Brunak:Protein distance constraints predicted by neural networks and probability density functions. Protein Engineering, 10, 1241-1248, 1997
</CITE>
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Authors</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
     Ole Lund (CBS, Copenhagen, Denmark)
</TD></TR>

<TR VALIGN=TOP>
<TD VALIGN=TOP BGCOLOR="#0000FF" WIDTH="15%">
     <STRONG><FONT COLOR="#FFFFFF">Contact</FONT></STRONG></TD>
<TD VALIGN=TOP WIDTH="85%">
        Kristoffer Rapacki (<A HREF="mailto:rapacki@cbs.dtu.dk">rapacki@cbs.dtu.dk</A>)
</TD></TR>


</TABLE>
<P><BR><P>


<!-- end services  -->
<!-- -------------------------------------------------------------------------------- -->

<P><BR><HR><BR><P>

<!-- -------------------------------------------------------------------------------- -->
<!-- beg formats  -->
<A NAME="PX_formats">
<TABLE WIDTH="100%">
<TR><TD VALIGN=TOP ALIGN=CENTER BGCOLOR="#000080">
	<STRONG><FONT COLOR="#FFFFFF" SIZE="+3">
	<A NAME="formats">Formats for submitting sequence or alignment</A>
	</FONT></STRONG>
	</TD></TR>
</TABLE>
</A>
The following formats are valid for submitting your sequence, or alignment.<BR>
<B>Note: we recommend to use either of the formats in bold print! </B>
<UL>
<LI><B>simple sequence in one-letter amino acid code 
	(<A HREF="../Dexa/format_simple.html">Example</A>)</B>
<LI><B>sequence in FASTA format 
	(<A HREF="../Dexa/format_fasta.html">Example</A>)</B>
<LI>sequence in SWISS-PROT format 
	(<A HREF="../Dexa/format_swiss.html">Example</A>)
<LI>sequence in GCG format 
	(<A HREF="../Dexa/format_gcg.html">Example</A>)
<LI>sequence in PIR format 
	(<A HREF="../Dexa/format_pir.html">Example</A>)
<LI>sequence in DSSP format
	(<A HREF="../Dexa/format_dssp.html">Example</A>)

<LI><B>alignment in FASTA format 
	(<A HREF="../Dexa/format_fastamul.html">Example</A>)</B>
<LI><B>alignment in SAF format 
	(<A HREF="../Dexa/format_saf.html">Example</A>)</B>
<LI>alignment in MSF format 
	(<A HREF="../Dexa/format_msf.html">Example</A>)
<LI>alignment in HSSP format
	(<A HREF="../Dexa/format_hssp.html">Example</A>)
</UL>

<!-- end formats  -->
<!-- -------------------------------------------------------------------------------- -->

</BODY>
</HTML>
