<h2>Quotes for methods</h2>
<ol>

<li><a NAME="quote_pp"><strong><font SIZE="+1">PredictProtein</font>: </a> <cite>B Rost,G Yachdav and J Liu (2004) The PredictProtein Server. Nucleic Acids Research 32(Web Server issue):W321-W326.
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	yachdav@maple.bioc.columbia.edu
<li> URL:		<a href="http://cubic.bioc.columbia.edu">http://cubic.bioc.columbia.edu</a>

<li> Version:	10.20.04
<li> Description:	PredictProtein is the acronym for all prediction programs run.

</ul>


<!--
<li><a NAME="quote_pp"><strong><font SIZE="+1">PredictProtein</font>: </a> <cite>B Rost and J Liu (2003) The PredictProtein Server. Nucleic Acids Research 31(13): 3300-3304.
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	liu@maple.bioc.columbia.edu
<li> URL:		<a href="http://cubic.bioc.columbia.edu">http://cubic.bioc.columbia.edu</a>

<li> Version:	1.99.08
<li> Description:	PredictProtein is the acronym for all prediction programs run.

</ul>
-->

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2 Draft//EN">
<li><a NAME="quote_pp"><strong><font SIZE="+1">PROSITE</font>: </a> <cite>A Bairoch, P Bucher & K Hofmann (1997) Nucleic Acids Research, 25:217-221
</cite></strong>
<ul>
<li> Author:	A Bairoch, bairoch@cmu.unige.ch P Bucher & K Hofmann
<li> Contact:	bairoch@cmu.unige.ch
<li> URL:		<a href="http://www.expasy.ch/prosite">http://www.expasy.ch/prosite</a>

<li> Version:	99.07
<li> Description:	PROSITE is a database of functional motifs. ScanProsite, finds all functional motifs in your sequence that are annotated in the ProSite db.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">SEG</font>: </a> <cite>J C Wootton & S Federhen (1996) Methods in Enzymology, 266:554-571
</cite></strong>
<ul>
<li> Author:	J C Wootton & S Federhen, wootton@ncbi.nlm.nih.gov
<li> Contact:	rost@columbia.edu 
<li> URL:		<a href="wootton@ncbi.nlm.nih.gov">wootton@ncbi.nlm.nih.gov</a>

<li> Version:	1994
<li> Description:	SEG divides sequences into regions of low-, and high-complexity. Low-complexity regions typically correspond to 'simple sequences' or 'compositionally-biased' regions.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">ProDom</font>: </a> <cite>ELL Sonnhammer & D Kahn (1994) Protein Science, 3:482-492 
</cite></strong>
<ul>
<li> Author:	LL Sonnhammer; J Gouzy, F Corpet, F Servant, D Kahn, dkahn@zyx.toulouse.inra.fr
<li> Contact:	dkahn@zyx.toulouse.inra.fr
<li> URL:		<a href="http://protein.toulouse.inra.fr/prodom.html">http://protein.toulouse.inra.fr/prodom.html</a>

<li> Version:	2000.1
<li> Description:	ProDom is a database of putative protein domains. The database is searched with BLAST for domains corresponding to your protein.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">MaxHom</font>: </a> <cite>MaxHom: C Sander R Schneider (1991) Proteins, 9:56-68 
</cite></strong>
<ul>
<li> Author:	C Sander & R Schneider, schneider@lion-ag.de
<li> Contact:	schneider@lion-ag.de


<li> Version:	1.99.04
<li> Description:	MaxHom is a dynamic multiple sequence alignment program which finds similar sequences in a database.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">MView</font>: </a> <cite>MView: N P Brown, C Leroy & C Sander (1998) Bioinformatics, 14:380-381 
</cite></strong>
<ul>
<li> Author:	N Brown, nbrown@nimr.mrc.ac.uk

<li> URL:		<a href="http://www.predictprotein.org/mview/mathbio.nimr.mrc.ac.uk/_nbrown/MView/FAQ.html">http://mathbio.nimr.mrc.ac.uk/~nbrown/mview/</a>
<li> Copyright:	Copyright (C) Nigel P. Brown, 1997-1998. All rights reserved.
<li> Version:	1.40.2
<li> Description:	MView is a program converting multiple sequence alignments into fancy HTML formatted output.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">PHD</font>: </a> <cite>B Rost (1996) Methods in Enzymology, 266:525-539
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	rost@columbia.edu
<li> URL:		<a href="http://cubic.bioc.columbia.edu">http://cubic.bioc.columbia.edu</a>

<li> Version:	1.96
<li> Description:	PHD is a suite of programs predicting 1D structure (secondary structure, solvent accessibility) from multiple sequence alignments.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">PHDsec</font>: </a> <cite>B Rost & C Sander (1993) J. of Molecular Biology, 232:584-599
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	rost@columbia.edu
<li> URL:		<a href="http://cubic.bioc.columbia.edu">http://cubic.bioc.columbia.edu</a>

<li> Version:	1.96
<li> Description:	PHDsec predicts secondary structure from multiple sequence alignments.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">PHDacc</font>: </a> <cite>B Rost & C Sander (1994) Proteins, 20:216-226
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	rost@columbia.edu


<li> Version:	1.96
<li> Description:	PHDacc predicts per residue solvent accessibility from multiple sequence alignments.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">PHDhtm</font>: </a> <cite>B Rost, P Fariselli &  R Casadio (1996) Protein Science, 7:1704-1718 
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	rost@columbia.edu
<li> URL:		<a href="http://cubic.bioc.columbia.edu">http://cubic.bioc.columbia.edu</a>

<li> Version:	1.96
<li> Description:	PHDhtm predicts the location and topology of transmembrane helices from multiple sequence alignments.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">PROF</font>: </a> <cite>B Rost (2004) Meth. Mol. Biol., submitted.
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	rost@columbia.edu


<li> Version:	2000_04
<li> Description:	PROF is a suite of programs predicting 1D structure (secondary structure, solvent accessibility) from multiple sequence alignments.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">PROFsec</font>: </a> <cite>B Rost (2004) Meth. Mol. Biol., submitted.
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	rost@columbia.edu
<li> URL:		<a href="http://cubic.bioc.columbia.edu ">http://cubic.bioc.columbia.edu </a>

<li> Version:	2000_04
<li> Description:	PROFsec predicts secondary structure from multiple sequence alignments.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">PROFACC</font>: </a> <cite>B Rost (2004) Meth. Mol. Biol., submitted.
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	rost@columbia.edu
<li> URL:		<a href="http://cubic.bioc.columbia.edu">http://cubic.bioc.columbia.edu</a>

<li> Version:	2000_04
<li> Description:	PROFacc predicts per residue solvent accessibility from multiple sequence alignments.

</ul>
<li><a NAME="quote_pp"><strong><font SIZE="+1">GLOBE</font>: </a> <cite>B Rost (1998) unpublished
</cite></strong>
<ul>
<li> Author:	B Rost
<li> Contact:	rost@columbia.edu
<li> URL:       <a href="http://cubic.bioc.columbia.edu/papers/1999_globe/paper.html">http://cubic.bioc.columbia.edu/papers/1999_globe/paper.html</a>

<li> Version:	1.98.05
<li> Description:	GLOBE predicts the globularity of a protein

</ul>
<!--
<li><a NAME="quote_pp"><strong><font SIZE="+1">CYSPRED</font>: </a> <cite>Fariselli P, Riccobelli P & Casadio R (1999) PROTEINS 36:340-346
</cite></strong>
<ul>
<li> Author:	Fariselli P, Riccobelli P, Casadio R
<li> Contact:	piero@lipid.biocomp.unibo.it
<li> URL:		<a href="http://prion.biocomp.unibo.it/cyspred.html">http://prion.biocomp.unibo.it/cyspred.html</a>

<li> Version:	0
<li> Description:	CYSPRED finds whether the cys residue in your protein forms disulfide bridge.

</ul>
-->
<li><a NAME="quote_pp"><strong><font SIZE="+1">DISULFIND</font>: </a> <cite>A.Ceroni, P.Frasconi, A.Passerini and A.Vullo (2004) Bioinformatics, 20, 653-659, 2004
</cite></strong>
<ul>
<li> Author:	A.Ceroni, P.Frasconi, A.Passerini and A.Vullo
<li> Contact:	cystein@dsi.unifi.it
<li> URL:	<a href="http://cassandra.dsi.unifi.it/cysteines/index.html">http://cassandra.dsi.unifi.it/cysteines/index.html</a>

<li> Version:	2.0
<li> Description: DISULFIND is a disulphide bridges predictor based on a two steps process.

</ul>

<li><a NAME="quote_pp"><strong><font SIZE="+1">A conformational switch prediction program</font>: </a> <cite>Young et al. Protein Science(1999) 8:1752-64.
</cite></strong>
<ul>
<li> Author:	Young M, Kirshenbaum K, Dill KA and Highsmith S.
<li> Contact:	mmyoung@sandia.gov, kent@cheme.caltech.edu, shighsmith@sf.uop.edu


<li> Version:	1.0
<li> Description:	ASP finds regions that are most likely to behave as switches in proteins known to exhibit this behavior

</ul>
<li> <a NAME="quote_pp"><strong><font SIZE="+1">HMMPFAM:</font></a> <cite>Bateman et al. Nucleic Acids Research 2004 32:D138-D141.</cite></strong>
<ul>
<li> Author:	Bateman A, Coin L, Durbin R, Finn RD, Hollich V, Griffiths-Jones S, Khanna A, Marshall M, Moxon S, Sonnhammer EL, Studholme DJ, Yeats C, Eddy SR. 
<li> Contact:   agb@sanger.ac.uk

<li> Version:	2.2g
<li> Description:	Search one or more sequences against HMM database
</ul>

</ol>