
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
<HEAD>
<TITLE>MView: FAQ</TITLE>
<META NAME="description" CONTENT="multiple alignment viewer">
<META NAME="keywords"   CONTENT="faq, sequence alignment, sequence database search">
</HEAD>
<BODY BGCOLOR="white">

<H1>MView: <FONT COLOR="red">Frequently Asked Questions</FONT></H1>

<UL>
<LI><A HREF="#what">What is <STRONG>MView</STRONG>?</A>
<LI><A HREF="#examples">Examples</A>
<LI><A HREF="#informats">What input formats are recognised?</A>
<LI><A HREF="#outformats">What output formats are there?</A>
<LI><A HREF="#informats_web">Can <strong>MView</strong> process data from a Web page?</A>
<LI><A HREF="#options">Command line options</A>
<ul>
<li><A HREF="#basic_usage">Basic usage</A>
<li><A HREF="#basic_usage_html">Basic usage - adding HTML</A>
<li><A HREF="#rulers">Rulers</A>
<li><A HREF="#alignment_colouring_modes">Alignment colouring modes</A>
<li><A HREF="#palettes">Colour palettes</A>
<li><A HREF="#consensus_patterns">Consensus patterns</A>
<li><A HREF="#alignment_consensus_colouring">Alignment consensus colouring</A>
<li><A HREF="#reference">Sequence numbering or ranking</A>
<li><A HREF="#labelling_pagination">Labelling and pagination</A>
<li><A HREF="#filtering">Filtering alignments</A>
  <IMG SRC="gifs/updated.gif" ALT="[updated]">
<li><A HREF="#SRS">SRS (Sequence Retrieval System) links</A>
<li><A HREF="#css">Using Cascading Style Sheets</A>
</ul>
<LI><A HREF="#printing">How can I print?</A>
<LI><A HREF="#nohtml">Can I switch off HTML markup?</A>
<LI><A HREF="#percents">What do the percent identities mean?</A>
<LI><A HREF="#funnyseqs">Some sequences are incomplete or look corrupted?</A>
<ul>
<LI><A HREF="#incompletequery">Why is the BLAST query sequence incomplete?</A>
<LI><A HREF="#blaststack">How are overlapping BLAST HSPs processed?</A>
<LI><A HREF="#fastastack">What happens to sequence gaps?</A>
</ul>
<LI><A HREF="#memory">Memory usage</A>
  <IMG SRC="gifs/new.gif" ALT="[new]">
<LI><A HREF="#system">System requirements</A>
<LI><A HREF="#getting">Obtaining the software</A>
<LI><A HREF="#installing">Installation</A>
<LI><A HREF="#contact">Contacts and acknowledgments</A>
<LI><A HREF="#copying">Copyright and Licensing information</A>
</UL>


<DL>

<P><A NAME="what">
<DT><FONT COLOR="green">What is <STRONG>MView</STRONG>?</FONT></A>
<DD>

<STRONG>MView</STRONG> is a tool for converting the results of a sequence
database search into the form of a coloured multiple alignment of hits
stacked against the query. Alternatively, an existing multiple alignment
can be processed. In either case, the output is simply HTML, so the result
is platform independent and does not require a separate application or
applet to be loaded. <STRONG>MView</STRONG> is <STRONG><FONT
COLOR="red">NOT</FONT></STRONG> a multiple alignment program, nor is it a
general purpose alignment editor.

<P>If you use <STRONG>MView</STRONG> in your work, please cite:

<P> Brown, N.P., Leroy C., Sander C. (1998). MView: A Web compatible
database search or multiple alignment viewer. 
<I>Bioinformatics</I>. <B>14</B>(4):380-381.
<A HREF="http://www.oup.co.uk/bioinformatics/hdb/Volume_14/Issue_04/html/btb054_gml.html"><SMALL>HTML</SMALL></A>
<A HREF="http://www.ncbi.nlm.nih.gov/htbin-post/Entrez/query?uid=9632837&form=6&db=m&Dopt=b"><SMALL>MEDLINE</SMALL></A>


<P><A NAME="examples">
<DT><FONT COLOR="green">Examples</FONT></A>
<DD>

Some <A HREF="examples/index.html">examples</A> illustrate use of various
command line options.

<P>The EBI operates a web service for <A
HREF="http://www2.ebi.ac.uk/blast2/">BLAST2</A> and <A
HREF="http://www2.ebi.ac.uk/fasta3/">FASTA3</A>, which uses
<STRONG>MView</STRONG>, as do <A
HREF="http://www.sander.ebi.ac.uk/genequiz/">GeneQuiz</A> and the GeneQuiz
<A HREF="http://www.sander.ebi.ac.uk/gqsrv/submit">sequence submission</A>
services.


<P><A NAME="informats">
<DT><FONT COLOR="green">What input formats are recognised?</FONT></A>
<DD>

The code has been tested for the following formats and versions for protein
and nucleotide sequences:
<P>
<CENTER>
<TABLE WIDTH="95%" BORDER CELLPADDING=3>
<TR ALIGN="left"><TH COLSPAN=4><FONT COLOR="red">BLAST (NCBI series 2.0)</FONT></TR>
<TR ALIGN="left"><TH>format          </TH><TH>tested versions                </TH><TH>status      </TH><TH>MView option</TH></TR>
<TR ALIGN="left"><TD>blastp          </TD><TD>2.0.4, 2.0.5, 2.0.9, 2.0.10    </TD><TD>ok          </TD><TD><CODE>-in blast  </CODE></TD></TR>
<TR ALIGN="left"><TD>blastn          </TD><TD>2.0.4, 2.0.5, 2.0.9            </TD><TD>ok          </TD><TD><CODE>-in blast  </CODE></TD></TR>
<TR ALIGN="left"><TD>blastx          </TD><TD>2.0.5, 2.0.9                   </TD><TD>ok          </TD><TD><CODE>-in blast  </CODE></TD></TR>
<TR ALIGN="left"><TD>tblastn         </TD><TD>2.0.5, 2.0.10                  </TD><TD>ok          </TD><TD><CODE>-in blast  </CODE></TD></TR>
<TR ALIGN="left"><TD>tblastx         </TD><TD>2.0.5                          </TD><TD>ok          </TD><TD><CODE>-in blast  </CODE></TD></TR>
<TR ALIGN="left"><TD>psi-blast       </TD><TD>2.0.2, 2.0.4, 2.0.5, 2.0.6, 2.0.10   </TD><TD>ok          </TD><TD><CODE>-in blast  </CODE></TD></TR>
<TR ALIGN="left"><TD>phi-blast       </TD><TD>2.0.9                          </TD><TD>experimental</TD><TD><CODE>-in blast  </CODE></TD></TR>

<TR ALIGN="left"><TH COLSPAN=4><FONT COLOR="red">BLAST (NCBI series 1.4)</FONT></TR>
<TR ALIGN="left"><TH>format          </TH><TH>tested versions                </TH><TH>status      </TH><TH>MView option</TH></TR>
<TR ALIGN="left"><TD>blastp          </TD><TD>1.4.7, 1.4.9                   </TD><TD>ok          </TD><TD><CODE>-in blast  </CODE></TD></TR>
<TR ALIGN="left"><TD>blastn          </TD><TD>1.4.9                          </TD><TD>ok          </TD><TD><CODE>-in blast  </CODE></TD></TR>
<TR ALIGN="left"><TD>blastx          </TD><TD>1.4.9                          </TD><TD>ok          </TD><TD><CODE>-in blast  </CODE></TD></TR>
<TR ALIGN="left"><TD>tblastn         </TD><TD>1.4.9                          </TD><TD>ok          </TD><TD><CODE>-in blast  </CODE></TD></TR>
<TR ALIGN="left"><TD>tblastx         </TD><TD>1.4.9                          </TD><TD>ok          </TD><TD><CODE>-in blast  </CODE></TD></TR>

<TR ALIGN="left"><TH COLSPAN=4><FONT COLOR="red">BLAST (WashU series 2.0)</FONT></TR>
<TR ALIGN="left"><TH>format          </TH><TH>tested versions                </TH><TH>status      </TH><TH>MView option</TH></TR>
<TR ALIGN="left"><TD>blastp          </TD><TD>2.0a13, 2.0a19                 </TD><TD>ok          </TD><TD><CODE>-in blast  </CODE></TD></TR>
<TR ALIGN="left"><TD>blastn          </TD><TD>2.0a19                         </TD><TD>ok          </TD><TD><CODE>-in blast  </CODE></TD></TR>
<TR ALIGN="left"><TD>blastx          </TD><TD>2.0a19                         </TD><TD>ok          </TD><TD><CODE>-in blast  </CODE></TD></TR>
<TR ALIGN="left"><TD>tblastn         </TD><TD>2.0a19                         </TD><TD>ok          </TD><TD><CODE>-in blast  </CODE></TD></TR>
<TR ALIGN="left"><TD>tblastx         </TD><TD>2.0a19                         </TD><TD>ok          </TD><TD><CODE>-in blast  </CODE></TD></TR>

<TR ALIGN="left"><TH COLSPAN=4><FONT COLOR="red">FASTA (series 3)</FONT></TR>
<TR ALIGN="left"><TH>format          </TH><TH>tested versions                </TH><TH>status      </TH><TH>MView option</TH></TR>
<TR ALIGN="left"><TD>fasta3          </TD><TD>3.0t76, 3.2t01, 3.2t07, 3.3t01 </TD><TD>ok          </TD><TD><CODE>-in fasta   </CODE></TD></TR>
<TR ALIGN="left"><TD>tfastx3         </TD><TD>3.0t82, 3.1t07                 </TD><TD>ok          </TD><TD><CODE>-in fasta   </CODE></TD></TR>

<TR ALIGN="left"><TH COLSPAN=4><FONT COLOR="red">FASTA (series 2)</FONT></TR>
<TR ALIGN="left"><TH>format          </TH><TH>tested versions                </TH><TH>status      </TH><TH>MView option</TH></TR>
<TR ALIGN="left"><TD>fasta           </TD><TD>2.0u                           </TD><TD>ok          </TD><TD><CODE>-in fasta   </CODE></TD></TR>
<TR ALIGN="left"><TD>tfastx          </TD><TD>2.0u63                         </TD><TD>ok          </TD><TD><CODE>-in fasta   </CODE></TD></TR>

<TR ALIGN="left"><TH COLSPAN=4><FONT COLOR="red">FASTA (series 1)</FONT></TR>
<TR ALIGN="left"><TH>format          </TH><TH>tested versions                </TH><TH>status      </TH><TH>MView option</TH></TR>
<TR ALIGN="left"><TD>fasta           </TD><TD>1.6c24                         </TD><TD>ok          </TD><TD><CODE>-in fasta   </CODE></TD></TR>

<TR ALIGN="left"><TH COLSPAN=4><FONT COLOR="red">multiple alignment formats</FONT></TR> 
<TR ALIGN="left"><TH>format          </TH><TH>versions                       </TH><TH>status      </TH><TH>MView option</TH></TR>
<TR ALIGN="left"><TD>plain           </TD><TD>&nbsp;-                        </TD><TD>ok          </TD><TD><CODE>-in plain   </CODE></TD></TR>
<TR ALIGN="left"><TD>Pearson/FASTA   </TD><TD>&nbsp;-                        </TD><TD>ok          </TD><TD><CODE>-in pearson </CODE></TD></TR>
<TR ALIGN="left"><TD>PIR             </TD><TD>&nbsp;-                        </TD><TD>ok          </TD><TD><CODE>-in pir     </CODE></TD></TR>
<TR ALIGN="left"><TD>MSF             </TD><TD>&nbsp;-                        </TD><TD>ok          </TD><TD><CODE>-in msf     </CODE></TD></TR>
<TR ALIGN="left"><TD>CLUSTAL W       </TD><TD>1.60, 1.70                     </TD><TD>ok          </TD><TD><CODE>-in clustal </CODE></TD></TR>
<TR ALIGN="left"><TD>MaxHom/HSSP     </TD><TD>1.0 1991                       </TD><TD>ok          </TD><TD><CODE>-in hssp    </CODE></TD></TR>
<TR ALIGN="left"><TD>MULTAS/MULTAL   </TD><TD>&nbsp;-                        </TD><TD>experimental</TD><TD><CODE>-in multas  </CODE></TD></TR>
<TR ALIGN="left"><TD>MIPS-ALN        </TD><TD>&nbsp;-                        </TD><TD>experimental</TD><TD><CODE>-in mips    </CODE></TD></TR>
</TABLE>
</CENTER>

<P>The "plain" multiple alignment format is a trivial format comprising a
column of identifiers and an adjacent column of aligned sequences. If you
can convert some strange alignment to this you can always read it into
<STRONG>MView</STRONG>. More formats can be expected to follow. 

<P>Note that as of version 1.37 <STRONG>MView</STRONG> automatically
selects an appropriate parser for the particular BLAST or FASTA
program/version once it knows it is dealing with input from either program
suite.


<P><A NAME="outformats">
<DT><FONT COLOR="green">What output formats are there?</FONT></A>
<DD>

The default behaviour is to (re)produce a multiple alignment from the input
data as plain ASCII. HTML markup will be added if any of the HTML-specific
or colouring options are set. Other useful output formats are dumps of the
input in Pearson/FASTA format (<CODE>-out pearson</CODE>), PIR format
(<CODE>-out pir</CODE>), or MSF format (<CODE>-out msf</CODE>) for
processing by another program , or as an RDB table for storage/manipulation
in relational database form (<CODE>-out rdb</CODE>).

<P>Historically, the default mode can also be explicitly selected
(<CODE>-out new</CODE>) in contrast to an obsolete mode (<CODE>-out
old</CODE>) which interdigitates extra rows into the alignment containing
just the sequence identities as in a sequence comparison.


<P><A NAME="informats_web">
<DT><FONT COLOR="green">Can <strong>MView</strong> process data from a Web page?</FONT></A>
<DD>

Basically, no, unless you are lucky or prepared to edit the Web
page. The <STRONG>MView</STRONG> parsers are all built to recognise the raw
text output produced by the respective programs (BLAST, FASTA, etc.) or to
recognise particular flat-file formats (MSF, PIR, etc.). When a site adds
HTML markup to this to make a Web page, arbitrary parts are
changed/deleted/added polluting the text, so that even dumping the page in
text-only format still leaves traces. If you feel that
<STRONG>MView</STRONG> is sufficiently useful, you could ask the Web site
maintainer to add an <STRONG>MView</STRONG> output option to their service
as has been done for the EBI <A
HREF="http://www2.ebi.ac.uk/blast2/">BLAST2</A> and <A
HREF="http://www2.ebi.ac.uk/fasta3/">FASTA3</A> services. Afterall, the
program is stable, free, portable, and intended to be used as widely as
possible.


<P><A NAME="options">
<DT><FONT COLOR="green">Command line options</FONT></A>
<DD>

Probably the first place to start is to invoke <STRONG>MView</STRONG> with:

<PRE>    mview -help</PRE>

<P>There are a lot of options, but the commonest ones are detailed
here. The basic action of the program is to generate a plain text dump of
the input data with <A HREF="#percents">percent sequence identities</A>
computed with respect to the first sequence in the output. 

<P>There are more command line options that I haven't documented below -
some were added for locally used features. Expect changes and new options
as the software evolves.

<DL>

<P><A NAME="basic_usage">
<DT><FONT COLOR="orange">Basic usage</FONT></A>
<DD>

Given an existing alignment in a file "data" in "plain" format, the minimal
use might be:<BR>

<PRE>    mview -in plain data > data.out</PRE>

Or you might attach <strong>MView</strong> on the end of a pipeline:

<PRE>    some_process | mview -in plain > data.out</PRE>

<P>To change the input format to scan a FASTA run, also in "data", use:

<PRE>    mview -in fasta data > data.out</PRE>


<P><A NAME="basic_usage_html">
<DT><FONT COLOR="orange">Basic usage - adding HTML</FONT></A>
<DD>

To add some HTML markup a few extra options are needed, for example:

<PRE>    mview -in fasta -html body data > data.html</PRE>

<P>produces a page of HTML wrapped inside <small>&lt;BODY&gt;
&lt;/BODY&gt;</small> tags with a coloured background, and you can load
this into your Web browser with a URL like
"<CODE>file://your_path/data.html</CODE>". 

<P>If you want a complete Web page, you can use <CODE>-html full</CODE>
(gives MIME-type, <small>&lt;HTML&gt;</small>, <small>&lt;BODY&gt;</small>
tags) or <CODE>-html head</CODE> (gives <small>&lt;HTML&gt;</small>,
<small>&lt;BODY&gt;</small> tags).

<P>To get just the alignment block without these tags use <CODE>-html
data</CODE>.

<P>Adding some colour is simple. To colour all the residues:

<PRE>    mview -in fasta -html head -coloring any data > data.html</PRE>

<P>and this looks better in my Netscape if the residues are emboldened, so

<PRE>    mview -in fasta -html head -coloring any -bold data > data.html</PRE>

<P>Now try colouring by identity to the first sequence:

<PRE>    mview -in fasta -html head -coloring identity -bold data > data.html</PRE>

<P>and then make the non-identical residues and gaps grey, instead of
black:

<PRE>    mview -in fasta -html head -coloring identity -bold -symcolor gray 
               -gapcolor gray data > data.html
</PRE>

<P>Now try using an internal <A HREF="#css">style sheet</A> to get blocked
colouring. The <CODE>-bold</CODE> option is no longer needed:

<PRE>    mview -in fasta -html head -css on -coloring identity -symcolor gray 
               -gapcolor gray data > data.html
</PRE>

<P>The <code>-in</code> option isn't always necessary. If the filename
extension, or the filename itself minus any directory path begins with or
contains the first few letters of the valid <code>-in</code> options (eg.,
<CODE>mydata.msf</CODE> or <CODE>mydata.fasta</CODE> or
<CODE>tfastx_run1.dat</CODE>), <STRONG>MView</STRONG> tries to choose a
sensible input format, allowing multiple files in mixed formats to be
supplied on the command line. The <code>-in</code> option will always
override this mechanism but requires that all input files be of the same
format.


<P><A NAME="rulers">
<DT><FONT COLOR="orange">Rulers</FONT></A>
<DD>

Add a ruler along the top, with <CODE>-ruler on</CODE>. Only one kind of
ruler is currently provided, numbering the columns of the final alignment
from M to N (incrementing) or N to M (decrementing) based on the input
sequence numbering, if any. This defaults to 1 to the length of the
alignment for multiple alignments. TBLASTX rulers differ slightly in that
the native query numbering is given in nucleotide units, but
<STRONG>MView</STRONG> reports amino acid units instead (using modulo 3
arithmetic).


<P><A NAME="alignment_colouring_modes">
<DT><FONT COLOR="orange">Alignment colouring modes</FONT></A>
<DD>

There are several ways to colour the alignment: 

<P><CODE>-coloring any</CODE>, will colour every residue according to the
currently selected <A HREF="#palettes">palette</A>.

<P><CODE>-coloring identity</CODE>, will colour only those residues that are
identical to some <A HREF="#reference">reference</A> sequence (usually the
query or first row).

<P><CODE>-coloring consensus</CODE>, will colour only those residues that
belong to a specified physicochemical class that is conserved in at least
a specified percentage of all rows for a given column. This defaults to 70%
and and may be set to another threshold, eg., <CODE>-coloring
consensus -threshold 80</CODE> would specify 80%. Note that the physicochemical
classes in question can be confined to individual residues.

<P><CODE>-coloring group</CODE>, is like <CODE>-coloring consensus</CODE>,
but colours residues by the colour of the class to which they belong.

<P>By default, the consensus computation counts gap characters, so that
sections of the alignment may be uncolored where the presence of gaps
prevents the non-gap count from reaching the threshold. Setting
<CODE>-con_gaps off</CODE> prevents this, allowing sequence-only based
consensus thresholding.

<P>The default palette assumes the input alignment is of protein sequences
and sets their colours according to amino acid physicochemical properties:
another palette should be selected for DNA or RNA alignments.

<P>Consensus colouring is complicated and some understanding of <A
HREF="#palettes">palettes</A> and <A HREF="#consensus_patterns">consensus
patterns</A> is required first before trying to explain <A
HREF="#alignment_consensus_colouring">alignment consensus colouring</A>.


<P><A NAME="palettes">
<DT><FONT COLOR="orange">Colour palettes</FONT></A>
<DD>

Palettes have (arbitrary) names, eg., <strong>MView</strong> assumes a
protein alignment and defaults to the palette <CODE>P1</CODE> for proteins
or <CODE>D1</CODE> for nucleotides. To change default molecule type use
<CODE>-dna</CODE>. Different palettes are explicitly selected using the
<CODE>-colormap</CODE> option. For example, to select one of the built-in
palettes for viewing nucleotide sequences, use <CODE>-colormap D1</CODE>.

<P>There are default palettes for protein and nucleotide sequences. The
latter can be selected with the <CODE>-dna</CODE> option.

<P>The built-in palettes can be listed from the command line with
<CODE>-listcolors</CODE>, and new colour schemes can be loaded from a file
using <CODE>-colorfile</CODE> in exactly the same format as produced by
<CODE>-listcolors</CODE>. Palette names are case-insensitive, while symbols
to be coloured are case-sensitive. Lines can contain comments beginning
with a hash '#' character. Colours are specified as hexadecimal RGB codes
prefixed with hash '#', exactly as used in HTML markup (named colours may
not be supported equally by all browsers). Here are the default palettes:

<PRE><small>
[P1]
&#35;protein: highlight amino acid physicochemical properties
&#35;symbols =>  color           [#comment]
<FONT COLOR="#666666">*      </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#mismatch          (dark grey)</FONT>
<FONT COLOR="#999999">?      </FONT>  =>  <FONT COLOR="#999999">#999999        </FONT> <FONT COLOR="#999999">#unknown           (light grey)</FONT>
<FONT COLOR="#33cc00">Aa     </FONT>  =>  <FONT COLOR="#33cc00">#33cc00        </FONT> <FONT COLOR="#33cc00">#hydrophobic       (bright green)</FONT>
<FONT COLOR="#666666">Bb     </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#D or N            (dark grey)</FONT>
<FONT COLOR="#ffff00">Cc     </FONT>  =>  <FONT COLOR="#ffff00">#ffff00        </FONT> <FONT COLOR="#ffff00">#cysteine          (yellow)</FONT>
<FONT COLOR="#0033ff">Dd     </FONT>  =>  <FONT COLOR="#0033ff">#0033ff        </FONT> <FONT COLOR="#0033ff">#negative charge   (bright blue)</FONT>
<FONT COLOR="#0033ff">Ee     </FONT>  =>  <FONT COLOR="#0033ff">#0033ff        </FONT> <FONT COLOR="#0033ff">#negative charge   (bright blue)</FONT>
<FONT COLOR="#009900">Ff     </FONT>  =>  <FONT COLOR="#009900">#009900        </FONT> <FONT COLOR="#009900">#large hydrophobic (dark green)</FONT>
<FONT COLOR="#33cc00">Gg     </FONT>  =>  <FONT COLOR="#33cc00">#33cc00        </FONT> <FONT COLOR="#33cc00">#hydrophobic       (bright green)</FONT>
<FONT COLOR="#009900">Hh     </FONT>  =>  <FONT COLOR="#009900">#009900        </FONT> <FONT COLOR="#009900">#large hydrophobic (dark green)</FONT>
<FONT COLOR="#33cc00">Ii     </FONT>  =>  <FONT COLOR="#33cc00">#33cc00        </FONT> <FONT COLOR="#33cc00">#hydrophobic       (bright green)</FONT>
<FONT COLOR="#cc0000">Kk     </FONT>  =>  <FONT COLOR="#cc0000">#cc0000        </FONT> <FONT COLOR="#cc0000">#positive charge   (bright red)</FONT>
<FONT COLOR="#33cc00">Ll     </FONT>  =>  <FONT COLOR="#33cc00">#33cc00        </FONT> <FONT COLOR="#33cc00">#hydrophobic       (bright green)</FONT>
<FONT COLOR="#33cc00">Mm     </FONT>  =>  <FONT COLOR="#33cc00">#33cc00        </FONT> <FONT COLOR="#33cc00">#hydrophobic       (bright green)</FONT>
<FONT COLOR="#6600cc">Nn     </FONT>  =>  <FONT COLOR="#6600cc">#6600cc        </FONT> <FONT COLOR="#6600cc">#polar             (purple)</FONT>
<FONT COLOR="#33cc00">Pp     </FONT>  =>  <FONT COLOR="#33cc00">#33cc00        </FONT> <FONT COLOR="#33cc00">#hydrophobic       (bright green)</FONT>
<FONT COLOR="#6600cc">Qq     </FONT>  =>  <FONT COLOR="#6600cc">#6600cc        </FONT> <FONT COLOR="#6600cc">#polar             (purple)</FONT>
<FONT COLOR="#cc0000">Rr     </FONT>  =>  <FONT COLOR="#cc0000">#cc0000        </FONT> <FONT COLOR="#cc0000">#positive charge   (bright red)</FONT>
<FONT COLOR="#0099ff">Ss     </FONT>  =>  <FONT COLOR="#0099ff">#0099ff        </FONT> <FONT COLOR="#0099ff">#small alcohol     (dull blue)</FONT>
<FONT COLOR="#0099ff">Tt     </FONT>  =>  <FONT COLOR="#0099ff">#0099ff        </FONT> <FONT COLOR="#0099ff">#small alcohol     (dull blue)</FONT>
<FONT COLOR="#33cc00">Vv     </FONT>  =>  <FONT COLOR="#33cc00">#33cc00        </FONT> <FONT COLOR="#33cc00">#hydrophobic       (bright green)</FONT>
<FONT COLOR="#009900">Ww     </FONT>  =>  <FONT COLOR="#009900">#009900        </FONT> <FONT COLOR="#009900">#large hydrophobic (dark green)</FONT>
<FONT COLOR="#666666">Xx     </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#any               (dark grey)</FONT>
<FONT COLOR="#009900">Yy     </FONT>  =>  <FONT COLOR="#009900">#009900        </FONT> <FONT COLOR="#009900">#large hydrophobic (dark green)</FONT>
<FONT COLOR="#666666">Zz     </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#E or Q            (dark grey)</FONT>

[D1]
&#35;DNA: highlight purine versus pyrimidine
&#35;symbols =>  color           [#comment]
<FONT COLOR="#666666">*      </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#mismatch           (dark grey)</FONT>
<FONT COLOR="#999999">?      </FONT>  =>  <FONT COLOR="#999999">#999999        </FONT> <FONT COLOR="#999999">#unknown            (light grey)</FONT>
<FONT COLOR="#0033ff">Aa     </FONT>  =>  <FONT COLOR="#0033ff">#0033ff        </FONT> <FONT COLOR="#0033ff">#purine             (bright blue)</FONT>
<FONT COLOR="#666666">Bb     </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#C or G or T; not A (dark grey)</FONT>
<FONT COLOR="#0099ff">Cc     </FONT>  =>  <FONT COLOR="#0099ff">#0099ff        </FONT> <FONT COLOR="#0099ff">#pyrimidine         (dull blue)</FONT>
<FONT COLOR="#666666">Dd     </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#A or G or T; not C (dark grey)</FONT>
<FONT COLOR="#0033ff">Gg     </FONT>  =>  <FONT COLOR="#0033ff">#0033ff        </FONT> <FONT COLOR="#0033ff">#purine             (bright blue)</FONT>
<FONT COLOR="#666666">Hh     </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#A or C or T; not G (dark grey)</FONT>
<FONT COLOR="#666666">Kk     </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#G or T             (dark grey)</FONT>
<FONT COLOR="#666666">Mm     </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#A or C             (dark grey)</FONT>
<FONT COLOR="#666666">Nn     </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#A or C or G or T   (dark grey)</FONT>
<FONT COLOR="#666666">Rr     </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#A or G             (dark grey)</FONT>
<FONT COLOR="#666666">Ss     </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#C or G             (dark grey)</FONT>
<FONT COLOR="#0099ff">Tt     </FONT>  =>  <FONT COLOR="#0099ff">#0099ff        </FONT> <FONT COLOR="#0099ff">#pyrimidine         (dull blue)</FONT>
<FONT COLOR="#0099ff">Uu     </FONT>  =>  <FONT COLOR="#0099ff">#0099ff        </FONT> <FONT COLOR="#0099ff">#pyrimidine         (dull blue)</FONT>
<FONT COLOR="#666666">Vv     </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#A or C or G; not T (dark grey)</FONT>
<FONT COLOR="#666666">Ww     </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#A or T             (dark grey)</FONT>
<FONT COLOR="#666666">Xx     </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#any                (dark grey)</FONT>
<FONT COLOR="#666666">Yy     </FONT>  =>  <FONT COLOR="#666666">#666666        </FONT> <FONT COLOR="#666666">#C or T             (dark grey)</FONT>
</small></PRE>

In these examples, both lower and uppercase versions of each residue are
given with their associated colour to ensure that either case is coloured
the same.

<P>The arrow separating the symbols from the colour codes can be double
<CODE>=></CODE> or single <CODE>-></CODE>. When <A HREF="#css">style
sheets</A> have been selected <CODE>-css on</CODE>, a double arrow means
that the colour should be applied to the background of the symbol while a
single arrow means that only the letter should be coloured. When Style
Sheets are off, only letters can be coloured anyway and the arrows are
equivalent.


<P><A NAME="consensus_patterns">
<DT><FONT COLOR="orange">Consensus patterns</FONT></A>
<DD>

A block of consensus lines can be added beneath the alignment using
<CODE>-consensus on</CODE>. By default, this adds 4 extra lines giving
consensus patterns computed at thresholds of 100,90,80,70%. 

<P>Consensus patterns are based on residue equivalence classes, that is,
sets of residues that share some physicochemical property. There are two
default consensus group definitions for protein <CODE>P1</CODE> and
nucleotide <CODE>D1</CODE> alignments, the latter being selected with the
<CODE>-dna</CODE> option.

<P>At a given percentage threshold, the most discriminating equivalence
class is chosen to represent the residues in a given column and an
associated symbol is displayed. For example, the default protein and
nucleotide consensus groups define the following symbols and equivalence
class mappings:

<PRE><small>
[P1]
&#35;protein consensus: report conserved physicochemical classes, derived from
&#35;the Venn diagrams of:
&#35; Taylor W. R. (1986). The classification of amino acid conservation.
&#35; J. Theor. Biol. 119:205-218.
&#35;as used in:
&#35; Bork, P., Brown, N.P., Hegyi, H., Schultz, J. (1996). The protein
&#35; phosphatase 2C (PP2C) superfamily: Detection of bacterial homologues.
&#35; Protein Science. 5:1421-1425.
&#35;description     =>  symbol  members
*                =>  .
A                =>  A       { A }
C                =>  C       { C }
D                =>  D       { D }
E                =>  E       { E }
F                =>  F       { F }
G                =>  G       { G }
H                =>  H       { H }
I                =>  I       { I }
K                =>  K       { K }
L                =>  L       { L }
M                =>  M       { M }
N                =>  N       { N }
P                =>  P       { P }
Q                =>  Q       { Q }
R                =>  R       { R }
S                =>  S       { S }
T                =>  T       { T }
V                =>  V       { V }
W                =>  W       { W }
Y                =>  Y       { Y }
alcohol          =>  o       { S, T }
aliphatic        =>  l       { I, L, V }
aromatic         =>  a       { F, H, W, Y }
charged          =>  c       { D, E, H, K, R }
hydrophobic      =>  h       { A, C, F, G, H, I, K, L, M, R, T, V, W, Y }
negative         =>  -       { D, E }
polar            =>  p       { C, D, E, H, K, N, Q, R, S, T }
positive         =>  +       { H, K, R }
small            =>  s       { A, C, D, G, N, P, S, T, V }
tiny             =>  u       { A, G, S }
turnlike         =>  t       { A, C, D, E, G, H, K, N, Q, R, S, T }

[D1]
&#35;DNA consensus: report conserved ring types
&#35;description     =>  symbol  members
*                =>  .
A                =>  A       { A }
C                =>  C       { C }
G                =>  G       { G }
T                =>  T       { T }
U                =>  U       { U }
purine           =>  r       { A, G }
pyrimidine       =>  y       { C, T, U }
</small></PRE>

Alternative equivalence classes can be selected using
<CODE>-con_groupmap</CODE>, the available list of built-ins can be seen
with <CODE>-listgroups</CODE>, and new groups can be defined in the same
format and read in from a file using <CODE>-groupfile</CODE>.

<P>Alternative thresholds to be displayed can be specified as a
comma-separated list using the <CODE>-con_threshold</CODE> option.

<P>Tip: A useful capability is to control whether only consensus properties
(<CODE>-con_ignore singleton</CODE>) or just the conserved residues
themselves (<CODE>-con_ignore class</CODE>) are displayed in consensus
lines. The default is to show both using whichever equivalence class is
most specific.

<P>By default, the consensus computation counts gap characters, so that
sections of the alignment may have gaps as the consensus. Setting
<CODE>-con_gaps off</CODE> prevents this, producing consensi based only on
sequence.

<P>You can specify a colour scheme for the consensus lines using
<CODE>-con_coloring</CODE> and <CODE>-con_colormap</CODE> to change the
default palette (<CODE>PC1</CODE> for protein or <CODE>DC1</CODE> for
nucleotide). These options are analogous to those for controlling the
alignment colouring and follow the same naming scheme.


<P><A NAME="alignment_consensus_colouring">
<DT><FONT COLOR="orange">Alignment consensus colouring</FONT></A>
<DD>

This section assumes an understanding of <A HREF="#palettes">palettes</A>
and <A HREF="#consensus_patterns">consensus patterns</A>.

<P>Colouring of an alignment by consensus determines which residues to
colour and the colours to use based on (1) the consensus threshold chosen
for the colouring operation (covered in the section on <A
HREF="#alignment_colouring_modes">alignment colouring modes</A>), (2) a
consideration of the common physicochemical properties of the residues in
that column, and (3) the chosen colour scheme:

<P><em>Given the most specific equivalence class describing the column
using the prevailing consensus equivalence classes, any residues in the
column belonging to that class will be coloured using the prevailing
palette.</em>

<P>In practice, for the default situation of a protein alignment and no
special selection of palettes or consensus groups from the command line,
then the <CODE>P1</CODE> (<CODE>D1</CODE>) equivalence classes and the
<CODE>P1</CODE> (<CODE>D1</CODE>) colour palette will be used (option
<CODE>-dna</CODE>).

<P>Tip: If you want to see only the conserved residues above the threshold
(ie., only one type of conserved residue per column), add the option
<CODE>-ignore class</CODE>.

<P>Alternative consensus classes and palettes can be specified using
<CODE>-groupmap</CODE> and <CODE>-colormap</CODE>. Note that these are
distinct from any settings used to control displayed consensus lines,
although the option naming is similar.


<P><A NAME="reference">
<DT><FONT COLOR="orange">Sequence numbering or ranking</FONT></A>
<DD>

One can colour and compute identities with respect to a sequence other
than the first/query sequence using the <CODE>-reference</CODE>
option. This takes either the sequence identifier or an integer argument
corresponding to the ranking or ordering of a sequence. For multiple
alignment input formats, sequences are numbered from 1, while for searches
the <em>hits</em> are numbered from 1, but the query itself is 0, so
beware.


<P><A NAME="labelling_pagination">
<DT><FONT COLOR="orange">Labelling and pagination</FONT></A>
<DD>

The labelling information can be too broad, so you can switch some
off. Labels at the left of the alignment are in blocks numbered from zero
(0) rank, (1) identifier, (2) description, (3) score block, (4) percent
identities. Each of these can be switched off with an option like
<CODE>-label2</CODE> to remove descriptions.

<P>The default layout is a single unbroken horizontal band of alignment -
fine if scrolling inside Netscape. However, you may prefer to break the
alignment into vertically stacked panes. For panes, for example, 80 columns
wide, set <CODE>-width 80</CODE>. Widths refer to the alignment, not to the
descriptor information at left.

<P>It is possible to narrow (or expand!) the displayed sequence range, for
example, <CODE>-range 10:78</CODE> would select only that column range of
the alignment using the numbering scheme reported when <CODE>-ruler
on</CODE> is set (see <A HREF="#rulers">Rulers</A>). The order of the
numbers is unimportant making it simpler to state interest in a region of
the alignment that might actually be reversed in the output (eg., a BLASTN
search hit matching the reverse complement of the query strand).


<P><A NAME="filtering">
<DT><FONT COLOR="orange">Filtering alignments</FONT></A>
<IMG SRC="gifs/updated.gif" ALT="[updated]">
<DD>

Usually, specifying a limited number of hits to view from a long search
alignment speeds things up a lot as there's less parsing and less
formatting to be generated, so to get the best 10 hits, use the option
<CODE>-top 10</CODE>.

<P>You also can squeeze more out of a deep alignment and get a less biased
view if a threshold on the pairwise sequence identity is set using
<CODE>-maxident N</CODE>, where N is some value between 0 and 100.

<P>Other filters specific to BLASTP, FASTA, etc., input formats allow
cutoffs on scores or p-values, etc. In particular, it is possible to apply
some control over the <A HREF="#blaststack">selection of HSPs</A> used in
building the <strong>MView</strong> alignment using the <CODE>-hsp</CODE>
filtering option.

<P>Of interest to anyone using PSI-BLAST, you can display alignments for
any/all iterations of a PSI-BLAST run using, say:

<PRE>    mview -in blast -cycle 1,5,10,20 mydata  </PRE>

to get just those iterations. The default is to display only the last
iteration. If you want all output, use <CODE>-cycle '*'</CODE>.

<P>Rows can be dropped explicitly using the <CODE>-disc</CODE> option. This
can be supplied a comma separated list of row identifiers, rank numbers
(see above for an explanation of sequence rank numbers), rank number
ranges, or regular expressions (case insensitive, enclosed between  
characters <IMG SRC="gifs/new.gif" ALT="[new]">) to
match against row identifiers.

<P>Likewise, the <CODE>-keep</CODE> option specifies a list of rows to keep
in the alignment. The <CODE>-keep</CODE> option overrides
<CODE>-disc</CODE> whenever a row is common to both <IMG
SRC="gifs/updated.gif" ALT="[updated]">.

<P>For example, the options
<PRE>
    -disc '/.&#42;&#47;'  -keep '2,3,6..10,/^pdb/'
</PRE>
would discard everything except rows 2, 3, 6 through 10 inclusive, and any
hits beginning with the string 'pdb'.

<P>Note: the currently set <A HREF="#reference">reference</A> row is still
used for percent identity and colouring operations, even though the row may
have been dropped from display by the <CODE>-disc</CODE> list. <IMG
SRC="gifs/new.gif" ALT="[new]">.

<P>Another control option can be used to prevent <STRONG>MView</STRONG>
from using rows for colouring or for calculation of percent identities
although these rows will still be displayed. Use <CODE>-nop</CODE> to
specify a list (comma separated as usual) of id's or row numbers to flag
for 'NO Processing'. This is useful for displaying non-alignment data (eg.,
secondary structure predictions) alongside the alignment. 


<P><A NAME="SRS">
<DT><FONT COLOR="orange">SRS (Sequence Retrieval System) links</FONT></A>
<DD>

If HTML markup is produced it is possible to embed SRS links in sequence
identifiers if they conform to the following patterns:
<PRE>
    database|accession|identifier 
    database:identifier
</PRE> as produced by some BLAST and FASTA servers. Such links will be to
the EBI and EMBL SRS services and will only be constructed if the database
names are listed in the SRS.pm library with this software. This library can
be modified for your site if you know some Perl and a little SRS syntax.


<P><A NAME="css">
<DT><FONT COLOR="orange">Using Cascading Style Sheets</FONT></A>
<DD>

<P>Release 1.40 added cascading style sheets allowing more specific control
of HTML elements. In particular, this enables selective colouring of text
fore/backgrounds allowing alignments to use coloured blocks instead of just
coloured lettering.

<P>This is enabled with the <CODE>-css on</CODE> option in combination with
the <CODE>-html</CODE> option to switch HTML processing on generally. It is
disabled with <CODE>-css off</CODE>. You can refer to an external style
file with <CODE>-css URL</CODE> where the URL give a valid path for the Web
server to find the file (ie., <CODE>file:/some/path</CODE> or
<CODE>http:&#47;&#47;server/path</CODE>).

<P>Having loaded your own colour schemes into <strong>MView</strong> with
the <CODE>-colorfile</CODE></CODE> option, you can dump these as a style
file with <CODE>-html css</CODE> which just dumps the style sheet to
standard output for redirection to a file.

<P>Controlling coloured fore/backgrounds for alignment lettering is handled
in the <A HREF="#palettes">colour scheme</A> definition mechanism.


</DL>


<P><A NAME="printing">
<DT><FONT COLOR="green">How can I print?</FONT></A>
<DD>

With difficulty from Netscape. First my UNIX Netscape (4.03) won't produce
colour postscript, but a Mac version does. On the Mac one can zoom and
preview the image. To produce something that fits on A4 one must set
<CODE>-width 60</CODE> or similar and turn off some of the leading text,
eg., <CODE>-label2 -label3</CODE>.


<P><A NAME="percents">
<DT><FONT COLOR="green">What do the percent identities mean?</FONT></A>
<DD>

Percent identities reported in each alignment row are calculated with
respect to the <A HREF="#reference">reference</A> sequence (usually the
query or first row), as follows:
<PRE>
            number of identical residues
           ------------------------------  x 100
            length of ungapped reference
            sequence over aligned region
</PRE>
Still, in the case of BLAST <STRONG>MView</STRONG> output, minor deviations
from the percentages reported by BLAST are due to 1) different rounding,
and 2) the way <strong>MView</strong> assembles a single <A
HREF="#funnyseqs">pseudo-sequence</A> for a hit composed of multiple HSPs,
giving an averaged percent identity. 


<P><A NAME="nohtml">
<DT><FONT COLOR="green">Can I switch off HTML markup?</FONT></A>
<DD>

Yes - the program defaults to plain ASCII output unless the
<CODE>-html</CODE> option is set. Also, selecting the relational RDB output
format will switch off HTML.


<P><A NAME="funnyseqs">
<DT><FONT COLOR="green">Some sequences are incomplete or look corrupted?</FONT></A>
<DD>

There are three kinds of input to <STRONG>MView</STRONG>: (i) a
preformatted multiple alignment, (ii) an <STRONG>ungapped</STRONG> search
(eg., BLAST 1.4.x), or (iii) a <STRONG>gapped</STRONG> search (eg., FASTA,
BLAST2). Multiple alignments require minimal parsing and are subjected only
to formatting stages. Searches are processed according to whether they are
are ungapped or gapped, and this can lead to some apparent inconsistencies in 
the output (discussed below). 

<DL>

<P><A NAME="incompletequery">
<DT><FONT COLOR="orange">Why is the BLAST query sequence incomplete?</FONT></A>
<DD>

The query sequence is recovered from the input search results. If, like in
BLAST, portions of the query were unmatched, they will not appear on the
output. Nevertheless <STRONG>MView</STRONG> will pad the missing sequence
with 'X' characters based on the numeric match ranges - the worst that can
happen then is that the trailing end (ie., the C-terminus) of the alignment
is missing. Occasionally, you may see a '?' character - this means that a
non-standard residue was seen on input.


<P><A NAME="blaststack">
<DT><FONT COLOR="orange">How are overlapping BLAST HSPs processed?</FONT></A>
<DD>

<STRONG>Ungapped</STRONG> input is processed to produce a stack of hit
sequence strings aligned against a contiguous query sequence.  The query
sequence acts as a template for each hit sequence onto which hit fragments
are overlayed in the query positions. 

<P>In outline the default method of processing of HSPs is as follows:

<P>For BLAST (series 1), as of <STRONG>MView</STRONG> version 1.37, only
the HSPs contributing to the ranked hit contribute to this overlay
process. A sorting scheme ensures that the best of these fragments are
overlayed last and are not obscured by weaker ones, for example, BLAST hits
are sorted by score and length. Differences of ordering of fragments along
query and hit naturally result in a <em>patchwork</em> that may not
correspond exactly to the real hit sequences. Nevertheless, the resulting
alignment stack is very informative, and the user can always run and view a
gapped search if that is preferred.

<P>For BLAST (series 2), only the single gapped hit reported in the ranking
is used and the patchwork problem does not arise.

<P>More detailed descriptions of the <a href="blast/index.html">rules for
HSP selection and tiling</a> are available. Some control over the choice of
HSPs is available through the <code>-hsp mode</code> option described
therein which allows a) only ranked HSPs (the default) to be tiled; b) all
HSPs to be tiled, or c) all HSPs to be extracted separately.


<P><A NAME="fastastack">
<DT><FONT COLOR="orange">What happens to sequence gaps?</FONT></A>
<DD>

<STRONG>Gapped</STRONG> input (eg., FASTA) is stacked in a slightly
different way. The query sequence again acts as a template, but gapped
regions introduced into the query are <em>excised</em> from both query and
hit to ensure a contiguous query string. In the affected hit sequence, the
position of the excision is marked by <em>lowercasing</em> the boundary
symbols. Again, the stacked alignments produced are very informative since
only unmatched regions of hits are lost from the display.

</DL>


<P><A NAME="memory">
<DT><FONT COLOR="green">Memory usage</FONT></A>
<IMG SRC="gifs/new.gif" ALT="[new]">
<DD>

Use of memory by <STRONG>MView</STRONG> can be very great, particularly if
you try to process complete sets of PSI-BLAST cycles each containing 1000s
of hits all at once. Use of most filtering options should reduce memory
requirements by cutting down the number of internal data structures
created. Likewise, processing each alignment separately will save memory or
you can use the option <CODE>-register off</CODE><IMG
SRC="gifs/new.gif" ALT="[new]"> to cause
each alignment to be output when ready (by default all alignments are saved
until the end so they can be printed with fields in register). Finally, the
choice of malloc library compiled into your perl may affect memory use.


<P><A NAME="system">
<DT><FONT COLOR="green">System requirements</FONT></A>
<DD>

<STRONG>MView</STRONG> and its underlying class libraries are implemented
in <A HREF="http://language.perl.com/">Perl</A>, version 5, for UNIX, and
should be easily portable to other systems. 

<P>As of <STRONG>MView</STRONG> release 1.40, the code requires a
<strong>minimum</strong> of perl version 5.004 and has been tested with
5.004_03/04 and 5.005_02. However, if you only have perl 5.003, you can run
older versions of <STRONG>MView</STRONG>, also available from the ftp site.

<P>Formatting and colouring of HTML alignments requires a fixed-width font
(eg., Courier) and support for the &lt;FONT&gt; tag, so a recent version of
a browser such as <A HREF="http://home.netscape.com/">Netscape</A> is
recommended. In particular, use of style sheets as of
<STRONG>MView</STRONG> release 1.40 requires that your browser supports
HTML 4.0.


<P><A NAME="getting">
<DT><FONT COLOR="green">Obtaining the software</FONT></A>
<DD>

The latest version of the software is <A
HREF="ftp://mathbio.nimr.mrc.ac.uk/pub/nige/MView/">available</A> free as a
UNIX gzipped tar archive. Be aware of the <A HREF="#copying">Copyright</A>
restrictions.


<P><A NAME="installing">
<DT><FONT COLOR="green">Installation</FONT></A>
<DD>

Installation on UNIX is easy, but assumes you already have <A
HREF="http://language.perl.com/">Perl5</A>.

<P>Save the archive to your software area, eg., <CODE>/usr/local</CODE>. 
<OL>

<LI>Gunzip it and extract it through tar, eg., 
<PRE>    gunzip < mview-1.24.tar.gz | tar xvof -</PRE>
This would create the subdirectory <CODE>mview-1.24</CODE>.

<LI>Change to this directory and load <CODE>bin/mview</CODE> into an
editor.

<LI>Set a perl interpreter path valid for your machine after the
'<CODE>#!</CODE>' magic number.

<LI>Change the "<CODE>use lib 'some stuff';</CODE>" line to, in our example,
<PRE>    use lib '/usr/local/mview-1.24/lib';</PRE>

<LI>Finally, copy <CODE>mview</CODE> to somewhere on your <CODE>PATH</CODE>
and rehash or login again.

</OL>

Ask your system manager or a Perl guru for help if this looks weird.


<P><A NAME="contact">
<DT><FONT COLOR="green">Contacts and acknowledgments</FONT></A>
<DD>

<TABLE CELLPADDING=0 CELLSPACING=0>

    <TR>
    <TD><A HREF="http://mathbio.nimr.mrc.ac.uk/~nbrown/">Nigel P. Brown</A></TD>
    </TR>
    
    <TR>
    <TD>National Institute for Medical Research,</TD>
    <TD><strong>Tel:</strong></TD> 
    <TD>+44 (0)181 959 3666</TD>
    </TR>
    
    <TR>
    <TD>The Ridgeway, Mill Hill,</TD>
    <TD><strong>Fax:</strong></TD>
    <TD>+44 (0)181 913 8545</TD>
    </TR>
    
    <TR>
    <TD>London NW7 1AA, U.K.</TD>
    <TD><strong>Email:</strong></TD> 
    <TD><A HREF="MAILTO:nbrown@nimr.mrc.ac.uk">nbrown@nimr.mrc.ac.uk</A></strong></TD>
    </TR>
    
</TABLE>

<P>People who have contributed include C. Leroy (prototype FASTA parser,
BLAST2 (WashU) modifications to BLAST 1.4 parser, prototype PSI-BLAST
parser). Useful suggestions came from R. Lopez, and my former colleagues in
the defunct <A HREF="http://www.sander.ebi.ac.uk">Sander group</A>.

<P>This project is unrelated to the <A
HREF="http://www.bioperl.org">Bioperl</A>
project, but probably should be...


<P><A NAME="copying">
<DT><FONT COLOR="green">Copyright and Licensing information</FONT></A>
<DD>

<STRONG>MView</STRONG> and associated libraries are free software, subject
to restrictions concerning commercial use. All users must adhere to the <A
HREF="COPYING.html">licensing terms</A>, acceptance of which is implicit
when the software is downloaded.

</DL>


<HR>
 <A HREF="http://www.nimr.mrc.ac.uk">[NIMR]</A>
 <A HREF="http://mathbio.nimr.mrc.ac.uk">[MathBio]</A>
<HR>
<ADDRESS><SMALL>Maintained by <A HREF="MAILTO:nbrown@nimr.mrc.ac.uk">Nigel Brown</A>.</SMALL> <SMALL>Last update Dec  1 1999
</SMALL>. <A HREF="http://mathbio.nimr.mrc.ac.uk/disclaimer.html"><SMALL>Disclaimer</SMALL></A></ADDRESS>
</BODY></HTML>
